{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Alignment from the same protein\n",
    "\n",
    "When dealing with real data, we are getting the case where two subsequences that have an overlap but from completely different parts of the protein. We need to fix this because we are generating sequences that don't exist"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Example issue\n",
    "\n",
    "Protein entry: `CELR3_MOUSE`\n",
    "\n",
    "Protein sequence:\n",
    "```\n",
    "MARRPLWWGLPGPSTPVLLLLLLSLFPFSREELGGGGDQDWDPGVATTTGPRAQIGSGAVALCPESPGVWEDGDPGLGVREPVFMRLRVGRQNARNGRGAPEQPNAEVVVQALGSREQEAGQGPGYLLCWHPEISSCGRTGPLRRGSLPLDALSPGDSDLRNSSPHPSELLAQPDGSRPVAFQRNARRSIRKRVETSRCCGKLWEPGHKGQGERSATSTVDRGPFRRDCLPGSLGSGLGEDSAPRAVRTAPTPGSAPRESRTAPGRMRSRGLFRRRFLFERPGPRPPGFPTGPEAKQILSTNQARPRRAANRHPQFPQYNYQTLVPENEAAGTSVLRVVAQDPDPGEAGRLIYSLAALMNSRSLELFSIDPQSGLIRTAAALDRESMERHYLRVTAQDHGSPRLSATTMVAVTVADRNDHAPVFEQAQYRETLRENVEEGYPILQLRATDGDAPPNANLRYRFVGSPAVRTAAAAAFEIDPRSGLISTSGRVDREHMESYELVVEASDQGQEPGPRSATVRVHITVLDENDNAPQFSEKRYVAQVREDVRPHTVVLRVTATDKDKDANGLVHYNIISGNSRGHFAIDSLTGEIQVMAPLDFEAEREYALRIRAQDAGRPPLSNNTGLASIQVVDINDHAPIFVSTPFQVSVLENAPLGHSVIHIQAVDADHGENSRLEYSLTGVASDTPFVINSATGWVSVSGPLDRESVEHYFFGVEARDHGSPPLSASASVTVTVLDVNDNRPEFTMKEYHLRLNEDAAVGTSVVSVTAVDRDANSAISYQITGGNTRNRFAISTQGGVGLVTLALPLDYKQERYFKLVLTASDRALHDHCYVHINITDANTHRPVFQSAHYSVSMNEDRPVGSTVVVISASDDDVGENARITYLLEDNLPQFRIDADSGAITLQAPLDYEDQVTYTLAITARDNGIPQKADTTYVEVMVNDVNDNAPQFVASHYTGLVSEDAPPFTSVLQISATDRDAHANGRVQYTFQNGEDGDGDFTIEPTSGIVRTVRRLDREAVPVYELTAYAVDRGVPPLRTPVSIQVTVQDVNDNAPVFPAEEFEVRVKENSIVGSVVAQITAVDPDDGPNAHIMYQIVEGNIPELFQMDIFSGELTALIDLDYEARQEYVIVVQATSAPLVSRATVHVRLVDQNDNSPVLNNFQILFNNYVSNRSDTFPSGIIGRIPAYDPDVSDHLFYSFERGNELQLLVVNRTSGELRLSRKLDNNRPLVASMLVTVTDGLHSVTAQCVLRVVIITEELLANSLTVRLENMWQERFLSPLLGHFLEGVAAVLATPTEDVFIFNIQNDTDVGGTVLNVSFSALAPRGAGAGAAGPWFSSEELQEQLYVRRAALAARSLLDVLPFDDNVCLREPCENYMKCVSVLRFDSSAPFLASTSTLFRPIQPIAGLRCRCPPGFTGDFCETELDLCYSNPCRNGGACARREGGYTCVCRPRFTDCELDTEAGRCVPGVCRNGGTCTNAPNGGFRCQCPAGGAFEGPRCEVAARSFPPSSFVMFRGLRQRFHLTLSLSFATVQPSGLLFYNGRLNEKHDFLALELVAGQVRLTYSTGESNTVVSPTVPGGLSDGQWHTVHLRYYNKPRTDALGGAQGPSKDKVAVLSVDDCNVAVALQFGAEIGNYSCAAAGVQTSSKKSLDLTGPLLLGGVPNLPENFPVSHKDFIGCMRDLHIDGRRMDMAAFVANNGTMAGCQAKSHFCASGPCKNNGFCSERWGGFSCDCPVGFGGKDCRLTMAHPYHFQGNGTLSWDFGNDMAVSVPWYLGLSFRTRATKGILMQVQLGPHSVLLCKLDRGLLSVTLNRASGHTVHLLLDQMTVSDGRWHDLRLELQEEPGGRRGHHIFMVSLDFTLFQDTMAMGGELQGLKVKQLHVGGLPPSSKEEGHQGLVGCIQGVWIGFTPFGSSALLPPSHRVNVEPGCTVTNPCASGPCPPHADCKDLWQTFSCTCRPGYYGPGCVDACLLNPCQNQGSCRHLQGAPHGYTCDCVSGYFGQHCEHRVDQQCPRGWWGSPTCGPCNCDVHKGFDPNCNKTNGQCHCKEFHYRPRGSDSCLPCDCYPVGSTSRSCAPHSGQCPCRPGALGRQCNSCDSPFAEVTASGCRVLYDACPKSLRSGVWWPQTKFGVLATVPCPRGALGAAVRLCDEDQGWLEPDLFNCTSPAFRELSLLLDGLELNKTALDTVEAKKLAQRLREVTGQTDHYFSQDVRVTARLLAYLLAFESHQQGFGLTATQDAHFNENLLWAGSALLAPETGHLWAALGQRAPGGSPGSAGLVQHLEEYAATLARNMELTYLNPVGLVTPNIMLSIDRMEHPSSTQGARRYPRYHSNLFRGQDAWDPHTHVLLPSQASQPSPSEVLPTSSNAENATASSVVSPPAPLEPESEPGISIVILLVYRALGGLLPAQFQAERRGARLPQNPVMNSPVVSVAVFHGRNFLRGVLVSPINLEFRLLQTANRSKAICVQWDPPGPTDQHGMWTARDCELVHRNGSHARCRCSRTGTFGVLMDASPRERLEGDLELLAVFTHVVVAVSVTALVLTAAVLLSLRSLKSNVRGIHANVAAALGVAELLFLLGIHRTHNQLLCTAVAILLHYFFLSTFAWLLVQGLHLYRMQVEPRNVDRGAMRFYHALGWGVPAVLLGLAVGLDPEGYGNPDFCWISIHEPLIWSFAGPIVLVIVMNGTMFLLAARTSCSTGQREAKKTSVLTLRSSFLLLLLVSASWLFGLLAVNHSILAFHYLHAGLCGLQGLAVLLLFCVLNADARAAWTPACLGKKAAPEETRPAPGPGSGAYNNTALFEESGLIRITLGASTVSSVSSARSGRAQDQDSQRGRSYLRDNVLVRHGSTAEHTERSLQAHAGPTDLDVAMFHRDAGADSDSDSDLSLEEERSLSIPSSESEDNGRTRGRFQRPLRRAAQSERLLAHPKDVDGNDLLSYWPALGECEAAPCALQAWGSERRLGLDSNKDAANNNQPELALTSGDETSLGRAQRQRKGILKNRLQYPLVPQSRGTPELSWCRAATLGHRAVPAASYGRIYAGGGTGSLSQPASRYSSREQLDLLLRRQLSKERLEEVPVPAPVLHPLSRPGSQERLDTAPARLEARDRGSTLPRRQPPRDYPGTMAGRFGSRDALDLGAPREWLSTLPPPRRNRDLDPQHPPLPLSPQRQLSRDPLLPSRPLDSLSRISNSREGLDQVPSRHPSREALGPAPQLLRAREDPASGPSHGPSTEQLDILSSILASFNSSALSSVQSSSTPSGPHTTATASALGPSTPRSATSHSISELSPDSEVPRSEGHS\n",
    "```\n",
    "b sequence: `PET`\n",
    "\n",
    "positions b sequence exists: `[2259]`\n",
    "\n",
    "y sequence: `TAA`\n",
    "\n",
    "positions y sequence exists: `[377, 470, 2547]`\n",
    "\n",
    "The aligned sequence is `PETAA` even though that sequence doesn't exist in the protein. We need a method to prevent this from happening"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Solution\n",
    "Make a function that looks for the disparity in the starting positions. Take the two closest positions possible and see if the gap between them is small enough. We must make sure that if one sequence doesn't completely overlap the other, that the y sequence is to the right of the b sequence"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import sys\n",
    "module_path = os.path.abspath(os.path.join('..'))\n",
    "if module_path not in sys.path:\n",
    "    sys.path.append(module_path)\n",
    "module_path = os.path.abspath(os.path.join('../..'))\n",
    "if module_path not in sys.path:\n",
    "    sys.path.append(module_path)\n",
    "\n",
    "from src.database import Database\n",
    "import re\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "def align_overlaps(seq1: str, seq2: str) -> str:\n",
    "    '''\n",
    "    Attempt to align two string sequences. It will look at the right side of seq1 and left side of seq2\n",
    "    to overlap the two strings. If no overlap is found, seq2 is appended to seq1\n",
    "\n",
    "    Example 1: strings with overlap\n",
    "\n",
    "        seq1: ABCD\n",
    "        seq2: CDEF\n",
    "\n",
    "        returns: ABCDEF\n",
    "\n",
    "    Example 2: strings with no overlap\n",
    "\n",
    "        seq1: ABCD\n",
    "        seq2: EFGH\n",
    "\n",
    "        return: ABCD-EFGH\n",
    "    \n",
    "    Inputs:\n",
    "        seq1:    (str) the left side sequence \n",
    "        seq2:    (str) the right side sequence\n",
    "    Outputs:\n",
    "        str the attempted alignments\n",
    "    '''\n",
    "    alignment = None\n",
    "    # if we have a perfect overlap, return it\n",
    "    if seq1 == seq2:\n",
    "        return seq1\n",
    "    \n",
    "    # if one is a full subsequence of another, return the larger one\n",
    "    elif seq1 in seq2:\n",
    "        return seq2\n",
    "    elif seq2 in seq1:\n",
    "        return seq1\n",
    "    \n",
    "    # try and find an alignment. seq2 should overlap as much of the right of seq1 as possible\n",
    "    # get the starting points. \n",
    "    # Starting points means we found the first character in seq2 in seq1\n",
    "    start_points = [i for i in range(len(seq1)) if seq2[0] == seq1[i]]\n",
    "    for sp in start_points:\n",
    "\n",
    "        # try and see if extending it makes it match\n",
    "        # a correct overlap should mean we run out of characters in \n",
    "        # seq1 before we hit the end of seq2\n",
    "        for i in range(sp, len(seq1)):\n",
    "\n",
    "            # if the next value i in seq1 does not equal the next \n",
    "            # characeter i-sp in seq2\n",
    "            if seq1[i] != seq2[i-sp]:\n",
    "                i -= 1\n",
    "                break\n",
    "\n",
    "        # if i hits the the end of seq1, we have an overlap\n",
    "        if i == len(seq1) - 1:\n",
    "            s2_start = len(seq1) - sp\n",
    "            right_seq = seq2[s2_start:] if s2_start < len(seq2) else ''\n",
    "            alignment = seq1 + right_seq\n",
    "            break\n",
    "  \n",
    "    # if no overlpa exists, just make append seq2 to seq1\n",
    "    if alignment is None:\n",
    "        alignment = seq1 + '-' + seq2\n",
    "\n",
    "    return alignment\n",
    "\n",
    "def hybrid_alignment(seq1: str, seq2: str) -> (str, str):\n",
    "    '''\n",
    "    Create a hybrid alignment from 2 sequences. If an overlap between these two sequences\n",
    "    is found, () are placed around the ambiguous section. If there is no overlap, then \n",
    "    seq2 is appended to seq1 with a - at the junction\n",
    "\n",
    "    Example 1: overlap of two strings \n",
    "\n",
    "        seq1: ABCDE\n",
    "        seq2: DEFGH\n",
    "\n",
    "        attempted alignment: ABC(DE)FGH\n",
    "        Output: (ABCDEFGH, ABC(DE)FGH)\n",
    "\n",
    "    Example 2: no overlap between the two strings\n",
    "\n",
    "        seq1: ABCD\n",
    "        seq2: EFGH\n",
    "\n",
    "        attempted alignment: ABCD-EFGH\n",
    "        Output: (ABCDEFGH, ABCD-EFGH)\n",
    "        \n",
    "    Inputs:\n",
    "        seq1:    (str) the left sequence\n",
    "        seq2:    (str) the right sequence\n",
    "    Outputs:\n",
    "        (str, str)  a tuple of strings. The first string is the sequence without any hybrid  \n",
    "                    identifying strings. The second is the sequence with the hybrid\n",
    "                    identifying strings -()\n",
    "    '''\n",
    "    alignment = ''\n",
    "    hybalignment = ''\n",
    "    attempted_overlap = align_overlaps(seq1, seq2)\n",
    "\n",
    "    # If an alignment was made (the shorter length means combined sequences is shorter than\n",
    "    # just appending them), identify the ambiguous area\n",
    "    if attempted_overlap is not None and 0 < len(attempted_overlap) < len(seq1) + len(seq2):\n",
    "        \n",
    "        # there is an overlap and some ambiguity\n",
    "        # get the starting point of seq2\n",
    "        rightstart = attempted_overlap.index(seq2)\n",
    "        leftend = len(seq1) - 1\n",
    "\n",
    "        # range between leftend and rigth start is ambiguous\n",
    "        middle_sec = attempted_overlap[rightstart:leftend + 1]\n",
    "        alignment = attempted_overlap\n",
    "        hybalignment = attempted_overlap[:rightstart] \\\n",
    "                        + '(' + middle_sec + ')' \\\n",
    "                        + attempted_overlap[leftend+1:]\n",
    "    \n",
    "    # therwise just append right seq to the left with a - to seperate them \n",
    "    else:\n",
    "        alignment = seq1 + seq2\n",
    "        hybalignment = seq1 + '-' + seq2\n",
    "        \n",
    "    return (alignment, hybalignment)\n",
    "\n",
    "\n",
    "\n",
    "def same_protein_alignment(seq1: str, seq2: str, parent_sequence: str) -> (str, str):\n",
    "    '''\n",
    "    Attempt to create a non-hybrid alignment from two sequences from the same protein. If the \n",
    "    two sequences do not directly overlap but have <= 1% (or 1 if very short) of the total number \n",
    "    of amino acids, make the alignment. If not, create a hybrid alignment from that. If one \n",
    "    compeletely overlaps the other, use that as the alignment. \n",
    "    \n",
    "    Example 1: Overlapped sequences\n",
    "        seq1: CDE        starting position: 2\n",
    "        seq2: ABCDEFG    starting position: 0\n",
    "        \n",
    "        Output: (ABCDEFG, None)\n",
    "        \n",
    "    Example 2: Partial overlaping sequences\n",
    "        seq1: ABCDE      starting position: 0\n",
    "        seq2: DEFGH      starting position: 3\n",
    "        \n",
    "        Outputs: (ABCDEFGH, None)\n",
    "        \n",
    "    Example 3: Non overlapping within the 1% (or 1)\n",
    "        seq1: ABCDE      starting position: 0\n",
    "        seq2: GHIJK      starting position: 6\n",
    "        \n",
    "        protein length: 52\n",
    "        grace length = max(1, 1% of 52) = 1\n",
    "        \n",
    "        Outputs: (ABCDEFGHIJK, None)\n",
    "        \n",
    "    Example 4: Theoretical overlap but outside of range\n",
    "        seq1: ABCDE      starting position: 0\n",
    "        seq2: EFGHI      starting position: 30\n",
    "        \n",
    "        outside of allowed range, make hybrid\n",
    "        \n",
    "        Outputs: (ABCDEFGHI, ABCD(E)FGHI)\n",
    "        \n",
    "    \n",
    "    Inputs:\n",
    "        seq1:            (str) left sequence to align\n",
    "        seq2:            (str) right sequence to align\n",
    "        parent_sequence: (str) sequence of the shared parent protein\n",
    "    Outputs:\n",
    "        tuple:   first entry is the seqence, second entry is \n",
    "                 the second entry is the hybrid sequence, if not hybrid, then its None\n",
    "                 \n",
    "                 \n",
    "                 \n",
    "                 ABCDE     DEFGH\n",
    "    ''' \n",
    "    # check to see if they are equal or one covers the entirety of the other\n",
    "    if seq1 == seq2:\n",
    "        return (seq1, None)\n",
    "    \n",
    "    if seq1 in seq2:\n",
    "        return (seq2, None)\n",
    "    \n",
    "    if seq2 in seq1: \n",
    "        return (seq1, None)\n",
    "    \n",
    "    # get the number of gap amino acids allowed\n",
    "    gap_aa = max(1, len(parent_sequence) // 100)\n",
    "    \n",
    "    # get the positions of the left sequence from the protein\n",
    "    left_start = [m.start() for m in re.finditer(seq1, parent_sequence)]\n",
    "    \n",
    "    # get the positions of the right sequence from the protein\n",
    "    right_start = [m.start() for m in re.finditer(seq2, parent_sequence)]\n",
    "    \n",
    "    # if all of the right starts are the the left of the left starts, make a hybrid alignment\n",
    "    if all([r < l for r in right_start for l in left_start]):\n",
    "        return hybrid_alignment(seq1, seq2)\n",
    "    \n",
    "    # check to see if any of the points are within the gap_aa limit\n",
    "    nonhybrid_alignments = []\n",
    "\n",
    "    for l in left_start:\n",
    "        for r in right_start:\n",
    "            \n",
    "            # if the right is to the left of left, continue\n",
    "            if r < l:\n",
    "                continue\n",
    "                \n",
    "            # if the right start - left start plus the length of the subsequence \n",
    "            # is less than the gap, just take the starting position of l and ending position \n",
    "            # of seq2 to make the full alignment\n",
    "            if r - (l + len(seq1)) <= gap_aa:\n",
    "                overlapped = parent_sequence[l: r + len(seq2)]\n",
    "                nonhybrid_alignments.append(overlapped)\n",
    "        \n",
    "    # if no nonhybrids could be made, return a hybrid alignment\n",
    "    if len(nonhybrid_alignments) == 0:\n",
    "        return hybrid_alignment(seq1, seq2)\n",
    "    \n",
    "    # we have at least one good one. Return the shortest one\n",
    "    nonhybrid_alignments.sort(key=lambda x: len(x))\n",
    "    return nonhybrid_alignments[0]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "('PETAA', 'PE(T)AA')\n"
     ]
    }
   ],
   "source": [
    "parent = 'MARRPLWWGLPGPSTPVLLLLLLSLFPFSREELGGGGDQDWDPGVATTTGPRAQIGSGAVALCPESPGVWEDGDPGLGVREPVFMRLRVGRQNARNGRGAPEQPNAEVVVQALGSREQEAGQGPGYLLCWHPEISSCGRTGPLRRGSLPLDALSPGDSDLRNSSPHPSELLAQPDGSRPVAFQRNARRSIRKRVETSRCCGKLWEPGHKGQGERSATSTVDRGPFRRDCLPGSLGSGLGEDSAPRAVRTAPTPGSAPRESRTAPGRMRSRGLFRRRFLFERPGPRPPGFPTGPEAKQILSTNQARPRRAANRHPQFPQYNYQTLVPENEAAGTSVLRVVAQDPDPGEAGRLIYSLAALMNSRSLELFSIDPQSGLIRTAAALDRESMERHYLRVTAQDHGSPRLSATTMVAVTVADRNDHAPVFEQAQYRETLRENVEEGYPILQLRATDGDAPPNANLRYRFVGSPAVRTAAAAAFEIDPRSGLISTSGRVDREHMESYELVVEASDQGQEPGPRSATVRVHITVLDENDNAPQFSEKRYVAQVREDVRPHTVVLRVTATDKDKDANGLVHYNIISGNSRGHFAIDSLTGEIQVMAPLDFEAEREYALRIRAQDAGRPPLSNNTGLASIQVVDINDHAPIFVSTPFQVSVLENAPLGHSVIHIQAVDADHGENSRLEYSLTGVASDTPFVINSATGWVSVSGPLDRESVEHYFFGVEARDHGSPPLSASASVTVTVLDVNDNRPEFTMKEYHLRLNEDAAVGTSVVSVTAVDRDANSAISYQITGGNTRNRFAISTQGGVGLVTLALPLDYKQERYFKLVLTASDRALHDHCYVHINITDANTHRPVFQSAHYSVSMNEDRPVGSTVVVISASDDDVGENARITYLLEDNLPQFRIDADSGAITLQAPLDYEDQVTYTLAITARDNGIPQKADTTYVEVMVNDVNDNAPQFVASHYTGLVSEDAPPFTSVLQISATDRDAHANGRVQYTFQNGEDGDGDFTIEPTSGIVRTVRRLDREAVPVYELTAYAVDRGVPPLRTPVSIQVTVQDVNDNAPVFPAEEFEVRVKENSIVGSVVAQITAVDPDDGPNAHIMYQIVEGNIPELFQMDIFSGELTALIDLDYEARQEYVIVVQATSAPLVSRATVHVRLVDQNDNSPVLNNFQILFNNYVSNRSDTFPSGIIGRIPAYDPDVSDHLFYSFERGNELQLLVVNRTSGELRLSRKLDNNRPLVASMLVTVTDGLHSVTAQCVLRVVIITEELLANSLTVRLENMWQERFLSPLLGHFLEGVAAVLATPTEDVFIFNIQNDTDVGGTVLNVSFSALAPRGAGAGAAGPWFSSEELQEQLYVRRAALAARSLLDVLPFDDNVCLREPCENYMKCVSVLRFDSSAPFLASTSTLFRPIQPIAGLRCRCPPGFTGDFCETELDLCYSNPCRNGGACARREGGYTCVCRPRFTDCELDTEAGRCVPGVCRNGGTCTNAPNGGFRCQCPAGGAFEGPRCEVAARSFPPSSFVMFRGLRQRFHLTLSLSFATVQPSGLLFYNGRLNEKHDFLALELVAGQVRLTYSTGESNTVVSPTVPGGLSDGQWHTVHLRYYNKPRTDALGGAQGPSKDKVAVLSVDDCNVAVALQFGAEIGNYSCAAAGVQTSSKKSLDLTGPLLLGGVPNLPENFPVSHKDFIGCMRDLHIDGRRMDMAAFVANNGTMAGCQAKSHFCASGPCKNNGFCSERWGGFSCDCPVGFGGKDCRLTMAHPYHFQGNGTLSWDFGNDMAVSVPWYLGLSFRTRATKGILMQVQLGPHSVLLCKLDRGLLSVTLNRASGHTVHLLLDQMTVSDGRWHDLRLELQEEPGGRRGHHIFMVSLDFTLFQDTMAMGGELQGLKVKQLHVGGLPPSSKEEGHQGLVGCIQGVWIGFTPFGSSALLPPSHRVNVEPGCTVTNPCASGPCPPHADCKDLWQTFSCTCRPGYYGPGCVDACLLNPCQNQGSCRHLQGAPHGYTCDCVSGYFGQHCEHRVDQQCPRGWWGSPTCGPCNCDVHKGFDPNCNKTNGQCHCKEFHYRPRGSDSCLPCDCYPVGSTSRSCAPHSGQCPCRPGALGRQCNSCDSPFAEVTASGCRVLYDACPKSLRSGVWWPQTKFGVLATVPCPRGALGAAVRLCDEDQGWLEPDLFNCTSPAFRELSLLLDGLELNKTALDTVEAKKLAQRLREVTGQTDHYFSQDVRVTARLLAYLLAFESHQQGFGLTATQDAHFNENLLWAGSALLAPETGHLWAALGQRAPGGSPGSAGLVQHLEEYAATLARNMELTYLNPVGLVTPNIMLSIDRMEHPSSTQGARRYPRYHSNLFRGQDAWDPHTHVLLPSQASQPSPSEVLPTSSNAENATASSVVSPPAPLEPESEPGISIVILLVYRALGGLLPAQFQAERRGARLPQNPVMNSPVVSVAVFHGRNFLRGVLVSPINLEFRLLQTANRSKAICVQWDPPGPTDQHGMWTARDCELVHRNGSHARCRCSRTGTFGVLMDASPRERLEGDLELLAVFTHVVVAVSVTALVLTAAVLLSLRSLKSNVRGIHANVAAALGVAELLFLLGIHRTHNQLLCTAVAILLHYFFLSTFAWLLVQGLHLYRMQVEPRNVDRGAMRFYHALGWGVPAVLLGLAVGLDPEGYGNPDFCWISIHEPLIWSFAGPIVLVIVMNGTMFLLAARTSCSTGQREAKKTSVLTLRSSFLLLLLVSASWLFGLLAVNHSILAFHYLHAGLCGLQGLAVLLLFCVLNADARAAWTPACLGKKAAPEETRPAPGPGSGAYNNTALFEESGLIRITLGASTVSSVSSARSGRAQDQDSQRGRSYLRDNVLVRHGSTAEHTERSLQAHAGPTDLDVAMFHRDAGADSDSDSDLSLEEERSLSIPSSESEDNGRTRGRFQRPLRRAAQSERLLAHPKDVDGNDLLSYWPALGECEAAPCALQAWGSERRLGLDSNKDAANNNQPELALTSGDETSLGRAQRQRKGILKNRLQYPLVPQSRGTPELSWCRAATLGHRAVPAASYGRIYAGGGTGSLSQPASRYSSREQLDLLLRRQLSKERLEEVPVPAPVLHPLSRPGSQERLDTAPARLEARDRGSTLPRRQPPRDYPGTMAGRFGSRDALDLGAPREWLSTLPPPRRNRDLDPQHPPLPLSPQRQLSRDPLLPSRPLDSLSRISNSREGLDQVPSRHPSREALGPAPQLLRAREDPASGPSHGPSTEQLDILSSILASFNSSALSSVQSSSTPSGPHTTATASALGPSTPRSATSHSISELSPDSEVPRSEGHS'\n",
    "\n",
    "print(same_protein_alignment('PET', 'TAA', parent))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
