{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Debug the real spectra\n",
    "\n",
    "So it doesn't work well with real data, so having taken out one example of real data, the sequence is\n",
    "\n",
    "Non hybrid: \n",
    "```\n",
    "DAAIVGYK\n",
    "```\n",
    "\n",
    "Non hybrid:\n",
    "```\n",
    "DEAPNFEANTTIGRIRFH\n",
    "```\n",
    "\n",
    "Non hybrid:\n",
    "``` \n",
    "DPQVEQLEL\n",
    "```\n",
    "\n",
    "Hybrid 1: \n",
    "```\n",
    "DLQTLALWSRM\n",
    "```\n",
    "\n",
    "Non hybrid:\n",
    "```\n",
    "DLTEYLSRFGEVV\n",
    "```\n",
    "\n",
    "Hybrid 2:\n",
    "```\n",
    "DLQTLEVE\n",
    "```\n",
    "\n",
    "Non Hybrid:\n",
    "```\n",
    "DLSSASAIMGNAKVKAHGKKVITAFN\n",
    "```\n",
    "\n",
    "Non hybrid:\n",
    "```\n",
    "EEILKAFKLF\n",
    "```\n",
    "\n",
    "so toy with the tool until we find out whats wrong"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Load params and run"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import sys\n",
    "module_path = os.path.abspath(os.path.join('..'))\n",
    "if module_path not in sys.path:\n",
    "    sys.path.append(module_path)\n",
    "module_path = os.path.abspath(os.path.join('../..'))\n",
    "if module_path not in sys.path:\n",
    "    sys.path.append(module_path)\n",
    "    \n",
    "from src import runner\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "specPath = '/Users/zacharymcgrath/Desktop/nod2 data/single/'\n",
    "fastaPath = '/Users/zacharymcgrath/Desktop/nod2 data/filteredNOD2.fasta'\n",
    "outputDir = '/Users/zacharymcgrath/Desktop/Experiment output/singleReal/'\n",
    "minPep = 3\n",
    "maxPep = 30\n",
    "tolerance = 20\n",
    "relative_abundance_filter = 0.0\n",
    "precursor_tolerance = 3\n",
    "peak_filter = 25\n",
    "verbose = True\n",
    "scoringAlg = 'ion'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "params = {\n",
    "    'spectra_folder': specPath,\n",
    "    'database_file': fastaPath,\n",
    "    'output_dir': outputDir,\n",
    "    'min_peptide_len': minPep,\n",
    "    'max_peptide_len': maxPep,\n",
    "    'tolerance': tolerance,\n",
    "    'precursor_tolerance': precursor_tolerance,\n",
    "    'peak_filter': peak_filter, \n",
    "    'relative_abundance_filter': relative_abundance_filter,\n",
    "    'verbose': verbose, \n",
    "    'scoring_alg': scoringAlg, \n",
    "    'DEBUG': True\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Loading database...\n",
      "Adding protein 279/279 to tree\n",
      "Done.\n",
      "Building hashes for kmers...\n",
      "Looking at protein 279/279\n",
      "Done.\n",
      "Analyzing spectra file 1/1[0%]\n",
      "\n",
      "Analyzing spectrum 1/8[0%]\n",
      "Filtering time took 0.1446380615234375 resulting in 1 b sequences and 7 y sequences\n",
      "First alignment round took 0.0006330013275146484 time resulting in 15 alignments\n",
      "Getting rid of ambiguous time took 0.006187915802001953\n",
      "Filling in precursor took 0.006826877593994141 for 15 sequences\n",
      "Time to make into objects took 0.00325775146484375\n",
      "Analyzing spectrum 2/8[12%]\n",
      "Filtering time took 0.377946138381958 resulting in 10 b sequences and 3 y sequences\n",
      "First alignment round took 0.004295825958251953 time resulting in 43 alignments\n",
      "Getting rid of ambiguous time took 0.05206012725830078\n",
      "Filling in precursor took 0.0056459903717041016 for 43 sequences\n",
      "Time to make into objects took 0.005527019500732422\n",
      "Analyzing spectrum 3/8[25%]\n",
      "Filtering time took 0.3484349250793457 resulting in 1 b sequences and 6 y sequences\n",
      "First alignment round took 0.0006549358367919922 time resulting in 13 alignments\n",
      "Getting rid of ambiguous time took 0.0017888545989990234\n",
      "Filling in precursor took 0.0030820369720458984 for 13 sequences\n",
      "Time to make into objects took 0.003390789031982422\n",
      "Analyzing spectrum 4/8[37%]\n",
      "Filtering time took 0.7787652015686035 resulting in 4 b sequences and 38 y sequences\n",
      "First alignment round took 0.020107030868530273 time resulting in 194 alignments\n",
      "Getting rid of ambiguous time took 1.3378968238830566\n",
      "Filling in precursor took 0.026715755462646484 for 194 sequences\n",
      "Time to make into objects took 0.021257877349853516\n",
      "Analyzing spectrum 5/8[50%]\n",
      "Filtering time took 0.5669651031494141 resulting in 6 b sequences and 2 y sequences\n",
      "First alignment round took 0.0017409324645996094 time resulting in 20 alignments\n",
      "Getting rid of ambiguous time took 0.010967016220092773\n",
      "Filling in precursor took 0.0029189586639404297 for 20 sequences\n",
      "Time to make into objects took 0.0054569244384765625\n",
      "Analyzing spectrum 6/8[62%]\n",
      "Filtering time took 0.3483881950378418 resulting in 3 b sequences and 2 y sequences\n",
      "First alignment round took 0.00092315673828125 time resulting in 11 alignments\n",
      "Getting rid of ambiguous time took 0.005083799362182617\n",
      "Filling in precursor took 0.007498025894165039 for 11 sequences\n",
      "Time to make into objects took 0.01840996742248535\n",
      "Analyzing spectrum 7/8[75%]\n",
      "Filtering time took 0.06113290786743164 resulting in 4 b sequences and 1 y sequences\n",
      "First alignment round took 0.0026619434356689453 time resulting in 9 alignments\n",
      "Getting rid of ambiguous time took 0.0024869441986083984\n",
      "Filling in precursor took 0.27611804008483887 for 9 sequences\n",
      "Time to make into objects took 1.0304038524627686\n",
      "Analyzing spectrum 8/8[87%]\n",
      "Filtering time took 0.28052496910095215 resulting in 10 b sequences and 6 y sequences\n",
      "First alignment round took 0.007096767425537109 time resulting in 74 alignments\n",
      "Getting rid of ambiguous time took 0.1065669059753418\n",
      "Filling in precursor took 0.02784895896911621 for 74 sequences\n",
      "Time to make into objects took 0.07814979553222656\n",
      "\n",
      "Finished search. Writting results to /Users/zacharymcgrath/Desktop/Experiment output/singleReal/...\n",
      "Could not make an alignment for 0/8 spectra (0%)\n"
     ]
    }
   ],
   "source": [
    "import time\n",
    "st = time.time()\n",
    "runner.run(params)\n",
    "ttime = time.time() - st"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "TIME: 75.63827514648438s\n"
     ]
    }
   ],
   "source": [
    "print(f'TIME: {ttime}s')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Load and read results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "\n",
    "outputFile = '/Users/zacharymcgrath/Desktop/Experiment output/singleReal/summary.tsv'\n",
    "hybridOutputFile = '/Users/zacharymcgrath/Desktop/Experiment output/singleReal/hybrid_summary.tsv'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "df = pd.read_csv(outputFile, sep='\\t')\n",
    "hdf = pd.read_csv(hybridOutputFile, sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th>left_proteins</th>\n",
       "      <th>right_proteins</th>\n",
       "      <th>sequence</th>\n",
       "      <th>hybrid_sequence</th>\n",
       "      <th>b_score</th>\n",
       "      <th>y_score</th>\n",
       "      <th>total_score</th>\n",
       "      <th>precursor_distance</th>\n",
       "      <th>entry name</th>\n",
       "      <th>id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>['INS1_MOUSE Insulin-1', 'INS2_MOUSE Insulin-2']</td>\n",
       "      <td>['CMGA_MOUSE Chromogranin-A', 'ANKS6_MOUSE Ank...</td>\n",
       "      <td>DLQTLALWSRM</td>\n",
       "      <td>DLQTLAL-WSRM</td>\n",
       "      <td>4</td>\n",
       "      <td>8</td>\n",
       "      <td>13</td>\n",
       "      <td>0.000192</td>\n",
       "      <td>/Users/zacharymcgrath/Desktop/nod2 data/single...</td>\n",
       "      <td>scan=10130</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>['NFU1_MOUSE NFU1 iron-sulfur cluster scaffold...</td>\n",
       "      <td>['CELR3_MOUSE Cadherin EGF LAG seven-pass G-ty...</td>\n",
       "      <td>SSIITLKGVREPVFMRLRVGRQN</td>\n",
       "      <td>SSIITLK-GVREPVFMRLRVGRQN</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>7</td>\n",
       "      <td>2.474539</td>\n",
       "      <td>/Users/zacharymcgrath/Desktop/nod2 data/single...</td>\n",
       "      <td>NOD2_E3.7258.7258.5.pkl</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   Unnamed: 0                                      left_proteins  \\\n",
       "0           0   ['INS1_MOUSE Insulin-1', 'INS2_MOUSE Insulin-2']   \n",
       "1           1  ['NFU1_MOUSE NFU1 iron-sulfur cluster scaffold...   \n",
       "\n",
       "                                      right_proteins                 sequence  \\\n",
       "0  ['CMGA_MOUSE Chromogranin-A', 'ANKS6_MOUSE Ank...              DLQTLALWSRM   \n",
       "1  ['CELR3_MOUSE Cadherin EGF LAG seven-pass G-ty...  SSIITLKGVREPVFMRLRVGRQN   \n",
       "\n",
       "            hybrid_sequence  b_score  y_score  total_score  \\\n",
       "0              DLQTLAL-WSRM        4        8           13   \n",
       "1  SSIITLK-GVREPVFMRLRVGRQN        3        3            7   \n",
       "\n",
       "   precursor_distance                                         entry name  \\\n",
       "0            0.000192  /Users/zacharymcgrath/Desktop/nod2 data/single...   \n",
       "1            2.474539  /Users/zacharymcgrath/Desktop/nod2 data/single...   \n",
       "\n",
       "                        id  \n",
       "0               scan=10130  \n",
       "1  NOD2_E3.7258.7258.5.pkl  "
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "hdf.head(5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th>proteins</th>\n",
       "      <th>sequence</th>\n",
       "      <th>b_score</th>\n",
       "      <th>y_score</th>\n",
       "      <th>total_score</th>\n",
       "      <th>precursor_distance</th>\n",
       "      <th>entry name</th>\n",
       "      <th>id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>['PROF1_MOUSE Profilin-1']</td>\n",
       "      <td>DAAIVGYK</td>\n",
       "      <td>3</td>\n",
       "      <td>6</td>\n",
       "      <td>9</td>\n",
       "      <td>0.001222</td>\n",
       "      <td>/Users/zacharymcgrath/Desktop/nod2 data/single...</td>\n",
       "      <td>scan=2470</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>['PRDX6_MOUSE Peroxiredoxin-6']</td>\n",
       "      <td>DEAPNFEANTTIGRIRFH</td>\n",
       "      <td>2</td>\n",
       "      <td>5</td>\n",
       "      <td>7</td>\n",
       "      <td>1.010352</td>\n",
       "      <td>/Users/zacharymcgrath/Desktop/nod2 data/single...</td>\n",
       "      <td>NOD2_E3.9670.9702.4.pkl</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>['INS1_MOUSE Insulin-1']</td>\n",
       "      <td>DPQVEQLEL</td>\n",
       "      <td>6</td>\n",
       "      <td>2</td>\n",
       "      <td>9</td>\n",
       "      <td>0.000673</td>\n",
       "      <td>/Users/zacharymcgrath/Desktop/nod2 data/single...</td>\n",
       "      <td>NOD2_E3.13446.13477.2.pkl</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>['HNRDL_MOUSE Heterogeneous nuclear ribonucleo...</td>\n",
       "      <td>DLTEYLSRFGEVV</td>\n",
       "      <td>5</td>\n",
       "      <td>5</td>\n",
       "      <td>10</td>\n",
       "      <td>0.000568</td>\n",
       "      <td>/Users/zacharymcgrath/Desktop/nod2 data/single...</td>\n",
       "      <td>scan=11410</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td>['CO3_MOUSE Complement C3']</td>\n",
       "      <td>ELTNIELL</td>\n",
       "      <td>5</td>\n",
       "      <td>1</td>\n",
       "      <td>6</td>\n",
       "      <td>0.971732</td>\n",
       "      <td>/Users/zacharymcgrath/Desktop/nod2 data/single...</td>\n",
       "      <td>NOD2_E3.11427.11516.2.pkl</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>5</td>\n",
       "      <td>5</td>\n",
       "      <td>['CETN2_MOUSE Centrin-2']</td>\n",
       "      <td>EEILKAFKLF</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>7</td>\n",
       "      <td>0.000765</td>\n",
       "      <td>/Users/zacharymcgrath/Desktop/nod2 data/single...</td>\n",
       "      <td>NOD2_E3.13697.13697.2.pkl</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   Unnamed: 0                                           proteins  \\\n",
       "0           0                         ['PROF1_MOUSE Profilin-1']   \n",
       "1           1                    ['PRDX6_MOUSE Peroxiredoxin-6']   \n",
       "2           2                           ['INS1_MOUSE Insulin-1']   \n",
       "3           3  ['HNRDL_MOUSE Heterogeneous nuclear ribonucleo...   \n",
       "4           4                        ['CO3_MOUSE Complement C3']   \n",
       "5           5                          ['CETN2_MOUSE Centrin-2']   \n",
       "\n",
       "             sequence  b_score  y_score  total_score  precursor_distance  \\\n",
       "0            DAAIVGYK        3        6            9            0.001222   \n",
       "1  DEAPNFEANTTIGRIRFH        2        5            7            1.010352   \n",
       "2           DPQVEQLEL        6        2            9            0.000673   \n",
       "3       DLTEYLSRFGEVV        5        5           10            0.000568   \n",
       "4            ELTNIELL        5        1            6            0.971732   \n",
       "5          EEILKAFKLF        2        2            7            0.000765   \n",
       "\n",
       "                                          entry name  \\\n",
       "0  /Users/zacharymcgrath/Desktop/nod2 data/single...   \n",
       "1  /Users/zacharymcgrath/Desktop/nod2 data/single...   \n",
       "2  /Users/zacharymcgrath/Desktop/nod2 data/single...   \n",
       "3  /Users/zacharymcgrath/Desktop/nod2 data/single...   \n",
       "4  /Users/zacharymcgrath/Desktop/nod2 data/single...   \n",
       "5  /Users/zacharymcgrath/Desktop/nod2 data/single...   \n",
       "\n",
       "                          id  \n",
       "0                  scan=2470  \n",
       "1    NOD2_E3.9670.9702.4.pkl  \n",
       "2  NOD2_E3.13446.13477.2.pkl  \n",
       "3                 scan=11410  \n",
       "4  NOD2_E3.11427.11516.2.pkl  \n",
       "5  NOD2_E3.13697.13697.2.pkl  "
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.head(6)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "import json\n",
    "\n",
    "s = json.load(open('/Users/zacharymcgrath/Desktop/Experiment output/singleReal/summary.json', 'r'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "DAAIVGYK: 9, 3, 6\n",
      "AAIVGYKD: 0, 0, 0\n",
      "LMMKKTS: 0, 0, 0\n",
      "DEAPNFEANTTIGRIRFH: 7, 2, 5\n",
      "TNFYFSLHIDGKIFQEK: 5, 1, 4\n",
      "EAPNFEANTTIGRIRFHD: 0, 0, 0\n",
      "DPQVEQLEL: 9, 6, 2\n",
      "EEIPHSELD: 4, 0, 3\n",
      "EQLESYGLE: 4, 0, 3\n",
      "DLQTLALWSRM: 13, 4, 8\n",
      "DLQTLALEFGLGG: 12, 4, 7\n",
      "DLQTLALGQIFD: 11, 4, 6\n",
      "DLTEYLSRFGEVV: 10, 5, 5\n",
      "EVTESIYLSRFGE: 4, 4, 0\n",
      "LTEYLSRFGEVVD: 0, 0, 0\n",
      "ELTNIELL: 6, 5, 1\n",
      "ELDVEELP: 6, 5, 0\n",
      "ELTNIEND: 5, 5, 0\n",
      "SSIITLKGVREPVFMRLRVGRQN: 7, 3, 3\n",
      "SSILLDLGVREPVFMRLRVGRQN: 7, 3, 3\n",
      "SSLIETLGVREPVFMRLRVGRQN: 7, 3, 3\n",
      "EEILKAFKLF: 7, 2, 2\n",
      "SVLNAAAQQGIAP: 7, 3, 2\n",
      "SVLNANINLPLA: 6, 3, 0\n"
     ]
    }
   ],
   "source": [
    "for k, v in s.items():\n",
    "    for a in v['alignments']:\n",
    "        print(f'{a[\"sequence\"]}: {a[\"total_score\"]}, {a[\"b_score\"]}, {a[\"y_score\"]}')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
