{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Debug the real spectra\n",
    "\n",
    "So it doesn't work well with real data, so having taken out one example of real data, the sequence is\n",
    "\n",
    "Non hybrid:\n",
    "```\n",
    "DEAPNFEANTTIGRIRFH\n",
    "```\n",
    "\n",
    "Non hybrid:\n",
    "``` \n",
    "DPQVEQLEL\n",
    "```\n",
    "\n",
    "Hybrid 1: \n",
    "```\n",
    "DLQTLALWSRM\n",
    "```\n",
    "\n",
    "Non hybrid:\n",
    "```\n",
    "DLTEYLSRFGEVV\n",
    "```\n",
    "\n",
    "Hybrid 2:\n",
    "```\n",
    "DLQTLEVE\n",
    "```\n",
    "\n",
    "Non Hybrid:\n",
    "```\n",
    "DLSSASAIMGNAKVKAHGKKVITAFN\n",
    "```\n",
    "\n",
    "Non hybrid:\n",
    "```\n",
    "EEILKAFKLF\n",
    "```\n",
    "\n",
    "so toy with the tool until we find out whats wrong"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Load params and run"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import sys\n",
    "module_path = os.path.abspath(os.path.join('..'))\n",
    "if module_path not in sys.path:\n",
    "    sys.path.append(module_path)\n",
    "module_path = os.path.abspath(os.path.join('../..'))\n",
    "if module_path not in sys.path:\n",
    "    sys.path.append(module_path)\n",
    "    \n",
    "from src import runner\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "specPath = '/Users/zacharymcgrath/Desktop/nod2 data/single/'\n",
    "fastaPath = '/Users/zacharymcgrath/Desktop/nod2 data/filteredNOD2.fasta'\n",
    "outputDir = '/Users/zacharymcgrath/Desktop/Experiment output/singleReal/'\n",
    "minPep = 3\n",
    "maxPep = 30\n",
    "tolerance = 20\n",
    "relative_abundance_filter = 0.0\n",
    "peak_filter = 25\n",
    "verbose = True\n",
    "scoringAlg = 'ion'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "params = {\n",
    "    'spectra_folder': specPath,\n",
    "    'database_file': fastaPath,\n",
    "    'output_dir': outputDir,\n",
    "    'min_peptide_len': minPep,\n",
    "    'max_peptide_len': maxPep,\n",
    "    'tolerance': tolerance,\n",
    "    'peak_filter': peak_filter, \n",
    "    'relative_abundance_filter': relative_abundance_filter,\n",
    "    'verbose': verbose, \n",
    "    'scoring_alg': scoringAlg, \n",
    "    'DEBUG': True\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Loading database...\n",
      "Adding protein 279/279 to tree\n",
      "Done.\n",
      "Building hashes for kmers...\n",
      "Looking at protein 279/279\n",
      "Done.\n",
      "Analyzing spectra file 1/1[0%]\n",
      "\n",
      "Analyzing spectrum 1/7[0%]\n",
      "max score: 1\n",
      "score pairs: \n",
      "[('EEI', 0), ('EEIRS', 1), ('EEISR', 1)]\n",
      "Reduced seq: EEI\n",
      "\n",
      "max score: 4\n",
      "score pairs: \n",
      "[('EEI', 0), ('NAKRFQILWEEEI', 4), ('AKRFQILWEEEI', 1)]\n",
      "Reduced seq: NAKRFQILWEEEI\n",
      "\n",
      "Filtering time took 0.387113094329834 resulting in 10 b sequences and 3 y sequences\n",
      "First alignment round took 0.003983736038208008 time resulting in 43 alignments\n",
      "Getting rid of ambiguous time took 0.06332015991210938\n",
      "Filling in precursor took 0.008602142333984375 for 43 sequences\n",
      "Time to make into objects took 0.005703926086425781\n",
      "Analyzing spectrum 2/7[14%]\n",
      "max score: 2\n",
      "score pairs: \n",
      "[('EEIL', 2), ('EEISRRGAGTPGQPGQ', 1), ('EEIPHSELDLET', 1), ('EEII', 2), ('EEI', 2)]\n",
      "Reduced seq: EEI\n",
      "\n",
      "Filtering time took 0.5254321098327637 resulting in 1 b sequences and 6 y sequences\n",
      "First alignment round took 0.0006546974182128906 time resulting in 13 alignments\n",
      "Getting rid of ambiguous time took 0.0018160343170166016\n",
      "Filling in precursor took 0.0031058788299560547 for 13 sequences\n",
      "Time to make into objects took 0.0035369396209716797\n",
      "Analyzing spectrum 3/7[28%]\n",
      "max score: 1\n",
      "score pairs: \n",
      "[('EEIAQVATISANGDK', 1), ('EEIDAQLQAEKQ', 1), ('EEIQTLSQVLAA', 1), ('EEIAKDKALKDLYVHTV', 1), ('EEITSLW', 1), ('EEILHY', 1), ('EEI', 0), ('EEIEKEEQAA', 1)]\n",
      "Reduced seq: EEI\n",
      "\n",
      "max score: 4\n",
      "score pairs: \n",
      "[('SRM', 2), ('YLHLSVSRM', 3), ('WSRM', 4), ('ILQESRM', 3)]\n",
      "Reduced seq: WSRM\n",
      "\n",
      "Filtering time took 0.7637341022491455 resulting in 4 b sequences and 38 y sequences\n",
      "First alignment round took 0.01929783821105957 time resulting in 194 alignments\n",
      "Getting rid of ambiguous time took 1.4996380805969238\n",
      "Filling in precursor took 0.026439905166625977 for 194 sequences\n",
      "Time to make into objects took 0.0221099853515625\n",
      "Analyzing spectrum 4/7[42%]\n",
      "max score: 2\n",
      "score pairs: \n",
      "[('EEIKKSEQP', 1), ('EEIS', 1), ('EEIRKCGMKNFR', 1), ('EEIPHSELDLETV', 2), ('EEI', 0)]\n",
      "Reduced seq: EEIPHSELDLETV\n",
      "\n",
      "max score: 2\n",
      "score pairs: \n",
      "[('DKRWSRM', 1), ('SRM', 0), ('WEDKRWSRM', 2)]\n",
      "Reduced seq: WEDKRWSRM\n",
      "\n",
      "max score: 1\n",
      "score pairs: \n",
      "[('EEI', 0), ('ELPLVMWLQGGPGGSSTGFGNFEEI', 1), ('HFDQEEI', 1)]\n",
      "Reduced seq: EEI\n",
      "\n",
      "Filtering time took 0.5869512557983398 resulting in 6 b sequences and 2 y sequences\n",
      "First alignment round took 0.0017161369323730469 time resulting in 20 alignments\n",
      "Getting rid of ambiguous time took 0.010381937026977539\n",
      "Filling in precursor took 0.0029380321502685547 for 20 sequences\n",
      "Time to make into objects took 0.004921913146972656\n",
      "Analyzing spectrum 5/7[57%]\n",
      "max score: 4\n",
      "score pairs: \n",
      "[('EEIAQVATISANGD', 3), ('EEIL', 4), ('EEIAK', 3), ('EEII', 4), ('EEI', 3), ('EEIQTLSQVL', 4)]\n",
      "Reduced seq: EEI\n",
      "\n",
      "Filtering time took 0.3465261459350586 resulting in 3 b sequences and 2 y sequences\n",
      "First alignment round took 0.0009441375732421875 time resulting in 11 alignments\n",
      "Getting rid of ambiguous time took 0.007074117660522461\n",
      "Filling in precursor took 0.004148960113525391 for 11 sequences\n",
      "Time to make into objects took 0.013200044631958008\n",
      "Analyzing spectrum 6/7[71%]\n",
      "Filtering time took 0.06443190574645996 resulting in 4 b sequences and 1 y sequences\n",
      "First alignment round took 0.002187013626098633 time resulting in 9 alignments\n",
      "Getting rid of ambiguous time took 0.0019388198852539062\n",
      "Filling in precursor took 0.26909923553466797 for 9 sequences\n",
      "Time to make into objects took 0.0076711177825927734\n",
      "Analyzing spectrum 7/7[85%]\n",
      "max score: 7\n",
      "score pairs: \n",
      "[('EEIL', 3), ('EEILKAFKLF', 7), ('EEII', 3), ('EEI', 1), ('EEILKVEQ', 4)]\n",
      "Reduced seq: EEILKAFKLF\n",
      "\n",
      "Filtering time took 0.2781641483306885 resulting in 10 b sequences and 6 y sequences\n",
      "First alignment round took 0.009705781936645508 time resulting in 74 alignments\n",
      "Getting rid of ambiguous time took 0.14099931716918945\n",
      "Filling in precursor took 0.040461063385009766 for 74 sequences\n",
      "Time to make into objects took 0.07154703140258789\n",
      "\n",
      "Finished search. Writting results to /Users/zacharymcgrath/Desktop/Experiment output/singleReal/...\n",
      "Could not make an alignment for 0/7 spectra (0%)\n"
     ]
    }
   ],
   "source": [
    "import time\n",
    "st = time.time()\n",
    "runner.run(params)\n",
    "ttime = time.time() - st"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "TIME: 93.50325894355774s\n"
     ]
    }
   ],
   "source": [
    "print(f'TIME: {ttime}s')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Load and read results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "\n",
    "outputFile = '/Users/zacharymcgrath/Desktop/Experiment output/singleReal/summary.tsv'\n",
    "hybridOutputFile = '/Users/zacharymcgrath/Desktop/Experiment output/singleReal/hybrid_summary.tsv'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "df = pd.read_csv(outputFile, sep='\\t')\n",
    "hdf = pd.read_csv(hybridOutputFile, sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th>left_proteins</th>\n",
       "      <th>right_proteins</th>\n",
       "      <th>sequence</th>\n",
       "      <th>hybrid_sequence</th>\n",
       "      <th>b_score</th>\n",
       "      <th>y_score</th>\n",
       "      <th>total_score</th>\n",
       "      <th>precursor_distance</th>\n",
       "      <th>entry name</th>\n",
       "      <th>id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>['INS1_MOUSE Insulin-1', 'INS2_MOUSE Insulin-2']</td>\n",
       "      <td>['CMGA_MOUSE Chromogranin-A', 'ANKS6_MOUSE Ank...</td>\n",
       "      <td>DLQTLALWSRM</td>\n",
       "      <td>DLQTLAL-WSRM</td>\n",
       "      <td>4</td>\n",
       "      <td>8</td>\n",
       "      <td>13</td>\n",
       "      <td>0.000192</td>\n",
       "      <td>/Users/zacharymcgrath/Desktop/nod2 data/single...</td>\n",
       "      <td>scan=10130</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>['VATF_MOUSE V-type proton ATPase subunit F']</td>\n",
       "      <td>['LEG1_MOUSE Galectin-1', 'CATB_MOUSE Cathepsi...</td>\n",
       "      <td>TINEIEIN</td>\n",
       "      <td>TINEIE-IN</td>\n",
       "      <td>5</td>\n",
       "      <td>0</td>\n",
       "      <td>5</td>\n",
       "      <td>0.492301</td>\n",
       "      <td>/Users/zacharymcgrath/Desktop/nod2 data/single...</td>\n",
       "      <td>NOD2_E3.11427.11516.2.pkl</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>['MRP9_MOUSE Multidrug resistance-associated p...</td>\n",
       "      <td>['CELR3_MOUSE Cadherin EGF LAG seven-pass G-ty...</td>\n",
       "      <td>LSFSSILGVREPVFMRLRVGRQN</td>\n",
       "      <td>LSFSSI-LGVREPVFMRLRVGRQN</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "      <td>4</td>\n",
       "      <td>0.004361</td>\n",
       "      <td>/Users/zacharymcgrath/Desktop/nod2 data/single...</td>\n",
       "      <td>NOD2_E3.7258.7258.5.pkl</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   Unnamed: 0                                      left_proteins  \\\n",
       "0           0   ['INS1_MOUSE Insulin-1', 'INS2_MOUSE Insulin-2']   \n",
       "1           1      ['VATF_MOUSE V-type proton ATPase subunit F']   \n",
       "2           2  ['MRP9_MOUSE Multidrug resistance-associated p...   \n",
       "\n",
       "                                      right_proteins                 sequence  \\\n",
       "0  ['CMGA_MOUSE Chromogranin-A', 'ANKS6_MOUSE Ank...              DLQTLALWSRM   \n",
       "1  ['LEG1_MOUSE Galectin-1', 'CATB_MOUSE Cathepsi...                 TINEIEIN   \n",
       "2  ['CELR3_MOUSE Cadherin EGF LAG seven-pass G-ty...  LSFSSILGVREPVFMRLRVGRQN   \n",
       "\n",
       "            hybrid_sequence  b_score  y_score  total_score  \\\n",
       "0              DLQTLAL-WSRM        4        8           13   \n",
       "1                 TINEIE-IN        5        0            5   \n",
       "2  LSFSSI-LGVREPVFMRLRVGRQN        1        3            4   \n",
       "\n",
       "   precursor_distance                                         entry name  \\\n",
       "0            0.000192  /Users/zacharymcgrath/Desktop/nod2 data/single...   \n",
       "1            0.492301  /Users/zacharymcgrath/Desktop/nod2 data/single...   \n",
       "2            0.004361  /Users/zacharymcgrath/Desktop/nod2 data/single...   \n",
       "\n",
       "                          id  \n",
       "0                 scan=10130  \n",
       "1  NOD2_E3.11427.11516.2.pkl  \n",
       "2    NOD2_E3.7258.7258.5.pkl  "
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "hdf.head(5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th>proteins</th>\n",
       "      <th>sequence</th>\n",
       "      <th>b_score</th>\n",
       "      <th>y_score</th>\n",
       "      <th>total_score</th>\n",
       "      <th>precursor_distance</th>\n",
       "      <th>entry name</th>\n",
       "      <th>id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>['PRDX6_MOUSE Peroxiredoxin-6']</td>\n",
       "      <td>DEAPNFEANTTIGRIRFH</td>\n",
       "      <td>2</td>\n",
       "      <td>5</td>\n",
       "      <td>7</td>\n",
       "      <td>1.010352</td>\n",
       "      <td>/Users/zacharymcgrath/Desktop/nod2 data/single...</td>\n",
       "      <td>NOD2_E3.9670.9702.4.pkl</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>['INS1_MOUSE Insulin-1']</td>\n",
       "      <td>DPQVEQLEL</td>\n",
       "      <td>6</td>\n",
       "      <td>2</td>\n",
       "      <td>9</td>\n",
       "      <td>0.000673</td>\n",
       "      <td>/Users/zacharymcgrath/Desktop/nod2 data/single...</td>\n",
       "      <td>NOD2_E3.13446.13477.2.pkl</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>['HNRDL_MOUSE Heterogeneous nuclear ribonucleo...</td>\n",
       "      <td>DLTEYLSRFGEVV</td>\n",
       "      <td>5</td>\n",
       "      <td>5</td>\n",
       "      <td>10</td>\n",
       "      <td>0.000568</td>\n",
       "      <td>/Users/zacharymcgrath/Desktop/nod2 data/single...</td>\n",
       "      <td>scan=11410</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>['CETN2_MOUSE Centrin-2']</td>\n",
       "      <td>EEILKAFKLF</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>7</td>\n",
       "      <td>0.000765</td>\n",
       "      <td>/Users/zacharymcgrath/Desktop/nod2 data/single...</td>\n",
       "      <td>NOD2_E3.13697.13697.2.pkl</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   Unnamed: 0                                           proteins  \\\n",
       "0           0                    ['PRDX6_MOUSE Peroxiredoxin-6']   \n",
       "1           1                           ['INS1_MOUSE Insulin-1']   \n",
       "2           2  ['HNRDL_MOUSE Heterogeneous nuclear ribonucleo...   \n",
       "3           3                          ['CETN2_MOUSE Centrin-2']   \n",
       "\n",
       "             sequence  b_score  y_score  total_score  precursor_distance  \\\n",
       "0  DEAPNFEANTTIGRIRFH        2        5            7            1.010352   \n",
       "1           DPQVEQLEL        6        2            9            0.000673   \n",
       "2       DLTEYLSRFGEVV        5        5           10            0.000568   \n",
       "3          EEILKAFKLF        2        2            7            0.000765   \n",
       "\n",
       "                                          entry name  \\\n",
       "0  /Users/zacharymcgrath/Desktop/nod2 data/single...   \n",
       "1  /Users/zacharymcgrath/Desktop/nod2 data/single...   \n",
       "2  /Users/zacharymcgrath/Desktop/nod2 data/single...   \n",
       "3  /Users/zacharymcgrath/Desktop/nod2 data/single...   \n",
       "\n",
       "                          id  \n",
       "0    NOD2_E3.9670.9702.4.pkl  \n",
       "1  NOD2_E3.13446.13477.2.pkl  \n",
       "2                 scan=11410  \n",
       "3  NOD2_E3.13697.13697.2.pkl  "
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.head(5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "import json\n",
    "\n",
    "s = json.load(open('/Users/zacharymcgrath/Desktop/Experiment output/singleReal/summary.json', 'r'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "DEAPNFEANTTIGRIRFH: 7, 2, 5\n",
      "TNFYFSLHIDGKIFQEK: 5, 1, 4\n",
      "EAPNFEANTTIGRIRFHD: 0, 0, 0\n",
      "DPQVEQLEL: 9, 6, 2\n",
      "EEIPHSELD: 4, 0, 3\n",
      "EQLESYGLE: 4, 0, 3\n",
      "DLQTLALWSRM: 13, 4, 8\n",
      "DLQTLALEFGLGG: 12, 4, 7\n",
      "DLQTLALGQIFD: 11, 4, 6\n",
      "DLTEYLSRFGEVV: 10, 5, 5\n",
      "EVTESIYLSRFGE: 4, 4, 0\n",
      "TINEIEIN: 5, 5, 0\n",
      "ELDVEEPN: 5, 5, 0\n",
      "ELDVEEPD: 5, 5, 0\n",
      "LSFSSILGVREPVFMRLRVGRQN: 4, 1, 3\n",
      "IYSSLLGVREPVFMRLRVGRQNA: 0, 0, 0\n",
      "EEILKAFKLF: 7, 2, 2\n",
      "SVLNAAAQQGIAP: 7, 3, 2\n",
      "SVLNANINLPLA: 6, 3, 0\n"
     ]
    }
   ],
   "source": [
    "for k, v in s.items():\n",
    "    for a in v['alignments']:\n",
    "        print(f'{a[\"sequence\"]}: {a[\"total_score\"]}, {a[\"b_score\"]}, {a[\"y_score\"]}')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
