{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def align_overlaps(seq1: str, seq2: str) -> str:\n",
    "    '''\n",
    "    Attempt to align two string sequences. It will look at the right side of seq1 and left side of seq2\n",
    "    to overlap the two strings. If no overlap is found, seq2 is appended to seq1\n",
    "\n",
    "    Example 1: strings with overlap\n",
    "\n",
    "        seq1: ABCD\n",
    "        seq2: CDEF\n",
    "\n",
    "        returns: ABCDEF\n",
    "\n",
    "    Example 2: strings with no overlap\n",
    "\n",
    "        seq1: ABCD\n",
    "        seq2: EFGH\n",
    "\n",
    "        return: ABCD-EFGH\n",
    "    \n",
    "    Inputs:\n",
    "        seq1:    (str) the left side sequence \n",
    "        seq2:    (str) the right side sequence\n",
    "    Outputs:\n",
    "        str the attempted alignments\n",
    "    '''\n",
    "    alignment = None\n",
    "    # if we have a perfect overlap, return it\n",
    "    if seq1 == seq2:\n",
    "        return seq1\n",
    "    \n",
    "    # if one is a full subsequence of another, return the larger one\n",
    "    elif seq1 in seq2:\n",
    "        return seq2\n",
    "    elif seq2 in seq1:\n",
    "        return seq1\n",
    "    \n",
    "    # try and find an alignment. seq2 should overlap as much of the right of seq1 as possible\n",
    "    # get the starting points. \n",
    "    # Starting points means we found the first character in seq2 in seq1\n",
    "    start_points = [i for i in range(len(seq1)) if seq2[0] == seq1[i]]\n",
    "    for sp in start_points:\n",
    "\n",
    "        # try and see if extending it makes it match\n",
    "        # a correct overlap should mean we run out of characters in \n",
    "        # seq1 before we hit the end of seq2\n",
    "        for i in range(sp, len(seq1)):\n",
    "\n",
    "            # if the next value i in seq1 does not equal the next \n",
    "            # characeter i-sp in seq2\n",
    "            if seq1[i] != seq2[i-sp]:\n",
    "                i -= 1\n",
    "                break\n",
    "\n",
    "        # if i hits the the end of seq1, we have an overlap\n",
    "        if i == len(seq1) - 1:\n",
    "            s2_start = len(seq1) - sp\n",
    "            right_seq = seq2[s2_start:] if s2_start < len(seq2) else ''\n",
    "            alignment = seq1 + right_seq\n",
    "            break\n",
    "  \n",
    "    # if no overlpa exists, just make append seq2 to seq1\n",
    "    if alignment is None:\n",
    "        alignment = seq1 + '-' + seq2\n",
    "\n",
    "    return alignment"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def __split_hybrid(sequence: str) -> (str, str):\n",
    "    '''\n",
    "    Split a hybrid sequence into it's left and right components\n",
    "    \n",
    "    Inputs:\n",
    "        sequence:    (str) the hybrid string sequnce\n",
    "    Outputs: \n",
    "        (str, str) the left and right sequences respectively\n",
    "    '''\n",
    "    if '-' in sequence:\n",
    "        return (sequence.split('-')[0], sequence.split('-')[1])\n",
    "    \n",
    "    else:\n",
    "        left = sequence.split(')')[0].replace('(', '')\n",
    "        right = sequence.split('(')[1].replace(')', '')\n",
    "        return (left, right)\n",
    "\n",
    "def __remove_amino_acids(spectrum: Spectrum, sequence: str, gap=3, tolerance=1) -> list:\n",
    "    '''\n",
    "    Remove up to gap number of amino acids to try and match precursor mass\n",
    "\n",
    "    Inputs:\n",
    "        spectrum:   (Spectrum) the aligned spectrum\n",
    "        sequence:   (str) the attempted string alignment\n",
    "    kwargs:\n",
    "        gap:        (int) the total number of free amino acids to try. Default=3\n",
    "        tolerance:  (float) the number (in Da) to allow as the tolerance for acceptable \n",
    "                            precursor masses. Default=1\n",
    "    Outputs:\n",
    "        (list) the sequence(s) with the closest precursor mass\n",
    "    '''\n",
    "    attempted = []\n",
    "    \n",
    "    # treat hybrids different than non-hybrids\n",
    "    if '-' in sequence or '(' in sequence or ')' in sequence:\n",
    "        \n",
    "        # get the left and right seperately\n",
    "        left_seq, right_seq = __split_hybrid(sequence)\n",
    "        \n",
    "        # since this is a hybrid, we assume that the left is ~correct and the right is ~correct\n",
    "        # so we only want to remove amino acids from the middle section\n",
    "        for i in range(gap):\n",
    "            for j in range(gap - i):\n",
    "                \n",
    "                # take left to :-i and right j:\n",
    "                new_left = left[:-i] if i > 0 else left\n",
    "                new_right = right[j:]\n",
    "                \n",
    "                # create a new hybrid\n",
    "                new_seq = align_overlaps(new_left, new_right)\n",
    "                \n",
    "                # get the new precursor\n",
    "                new_prec = gen_spectra.get_precursor(new_seq.replace('-', '').replace('(', '').replace(')', ''))\n",
    "                \n",
    "                # get the new precursor distance\n",
    "                pd = scoring.precursor_distance(spectrum.precursor_mass, new_prec)\n",
    "                \n",
    "                # if the precursor distance is within our tolerance, append it\n",
    "                if pd <= tolerance:\n",
    "                    attempted.append(new_seq)\n",
    "\n",
    "      # otherwise, just take up to gap off from the left and the right\n",
    "    else:\n",
    "        for i in range(1, gap):\n",
    "            new_seq1 = sequence[i:]\n",
    "            new_seq2 = sequence[:-i]\n",
    "            \n",
    "            # cacluate the new precurosrs and add to attempted if within the tolerance\n",
    "            new_prec1 = gen_sectra.get_precursor(new_seq1)\n",
    "            new_prec2 = gen_sectra.get_precursor(new_seq2)\n",
    "            \n",
    "            pd1 = scoring.precursor_distance(spectrum.precursor_mass, new_prec1)\n",
    "            pd2 = scoring.precursor_distance(spectrum.precursor_mass, new_prec2)\n",
    "            \n",
    "            if pd1 <= tolerance:\n",
    "                attempted.append(new_seq1)\n",
    "            \n",
    "            if pd2 <= tolerance:\n",
    "                attempted.append(new_seq2)\n",
    "\n",
    "    return list(set(attempted))\n",
    "\n",
    "def __add_amino_acids(spectrum: Spectrum, sequence: str, db: Database, gap=3, tolerance=1) -> list:\n",
    "    '''\n",
    "    Try and add amino acids to get the closest precursor mass\n",
    "\n",
    "    Inputs:\n",
    "        spectrum:   (Spectrum) the observed precursor mass\n",
    "        seqeunce:   (str) the attempted alignment\n",
    "        db:         (Database) holds the protein sequences\n",
    "    kwargs:\n",
    "        gap:        (int) the number of additions allowed. Default=3\n",
    "        tolerance:  (float) the mass (in Da) tolerance to accept in a precursor distance. Default=1\n",
    "    Outputs:\n",
    "        (list) the sequence(s) with the closest precursor mass\n",
    "    '''\n",
    "    filled_in  = []\n",
    "\n",
    "    # get the parents of the sequence(s) and add or subtract amino acids\n",
    "    parents = get_parents(sequence, db)\n",
    "\n",
    "    # if its hybrid, we should fill it in in all possible ways\n",
    "    if HYBRID_ALIGNMENT_PATTERN.findall(sequence):\n",
    "\n",
    "        # go through each set of parents\n",
    "        for l_p in parents[0]:\n",
    "            for r_p in parents[1]:\n",
    "                \n",
    "                # since its a hybrid, we only want to add to the middle. We know that\n",
    "                # the most left and most right SHOULD be the best scoring bits, so adding \n",
    "                # to the left of the left will only hurt, and adding to the right of the right\n",
    "                # will only hurt\n",
    "                left_seq, right_seq = __split_hybrid(sequence)\n",
    "                \n",
    "                # get the left and right protein sequences\n",
    "                left_prot = database.get_entry_by_name(db, l_p).sequence\n",
    "                right_prot = database.get_entry_by_name(db, r_p).sequence\n",
    "                \n",
    "                # get the aas to the right of the left sequence\n",
    "                left_append = [x[1] for x in __get_surrounding_amino_acids(left_prot, left_seq, gap)]\n",
    "                \n",
    "                # get the aas to the left of the right sequence\n",
    "                right_prepend = [x[0] for x in __get_surrounding_amino_acids(right_prot, right_seq, gap)]\n",
    "                \n",
    "                # slowly add each set of amino acids\n",
    "                for to_append in left_append:\n",
    "                    for to_preprend in right_prepend:\n",
    "                        \n",
    "                        # slowly add each\n",
    "                        for i in range(len(to_append)):\n",
    "                            for j in range(len(to_prepend)):\n",
    "                                \n",
    "                                new_left = left_seq + to_append[:i]\n",
    "                                new_right = ('' if j == 0 else to_prepend[-j:]) + right_seq\n",
    "                                \n",
    "                                # create the new sequence and get the new precursor mass\n",
    "                                new_seq = align_overlaps(new_left, new_right)\n",
    "                                \n",
    "                                new_prec = gen_spectra.get_precurosr(new_seq.replace('(', '').replace(')', '').replace('-', ''))\n",
    "                                \n",
    "                                # find the precursor distance, and if its close enough, keep it \n",
    "                                pd = scoring.precursor_distance(spectrum.precursor_mass, new_prec)\n",
    "                                \n",
    "                                if pd <= tolerance:\n",
    "                                    filled_in.append(new_seq)\n",
    "        \n",
    "    # if its nonhybrid, try left and right side\n",
    "    else:\n",
    "        for p in parents[0]:\n",
    "\n",
    "            # get the parent sequence\n",
    "            p_seq = database.get_entry_by_name(db, p).sequence\n",
    "\n",
    "            # get the flanking amino acid pairs\n",
    "            for flanking_pair in __get_surrounding_amino_acids(p_seq, sequence, gap):\n",
    "\n",
    "                # go through all possible combinations of the flanking pairs\n",
    "                for i in range(gap + 1):\n",
    "                    for j in range(gap - i + 1):\n",
    "\n",
    "                        # get the new sequence and its precursor mass. Add it to filled in list\n",
    "                        new_seq = flanking_pair[0][gap-i:] + sequence + flanking_pair[1][:j]\n",
    "\n",
    "                        new_prec = gen_spectra.get_precursor(new_seq)\n",
    "\n",
    "                        # get the precursor distance, and if it is close enough, keep it\n",
    "                        p_d = scoring.precursor_distance(spectrum.precursor_mass, new_prec)\n",
    "\n",
    "                        if p_d <= tolerance:\n",
    "                            filled_in.append(new_seq)\n",
    "    return filled_in"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'ABCD'"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "s = 'ABCD'\n",
    "s[0:]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
