{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Run a test of hypedsearch with generated data\n",
    "The following steps describe how the test works\n",
    "1. Load a fasta database\n",
    "2. Generate\n",
    "    1. Hybrid proteins\n",
    "    2. Peptides\n",
    "    3. Hybrid peptides from the hybrid proteins\n",
    "3. Generate spectra for all the peptides created\n",
    "4. Run hypedsearch with the .fasta file (no hybrid proteins included) and the spectra files\n",
    "5. Load the summary.json file created\n",
    "6. Determine what number of alignments were correct"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 1. Load fasta database"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import sys\n",
    "module_path = os.path.abspath(os.path.join('..'))\n",
    "if module_path not in sys.path:\n",
    "    sys.path.append(module_path)\n",
    "module_path = os.path.abspath(os.path.join('../..'))\n",
    "if module_path not in sys.path:\n",
    "    sys.path.append(module_path)\n",
    "    \n",
    "from src.file_io import fasta\n",
    "\n",
    "fasta_file = '../data/databases/100prots.fasta'\n",
    "database = fasta.read(fasta_file, True)\n",
    "\n",
    "database = {x['name']: x for x in database}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 2.  Generate the peptides, hybrid proteins and peptides"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Generating hybrid protein 0/5[0%]\r",
      "Generating hybrid protein 1/5[20%]\r",
      "Generating hybrid protein 2/5[40%]\r",
      "Generating hybrid protein 3/5[60%]\r",
      "Generating hybrid protein 4/5[80%]\r\n",
      "Finished generating hybrid proteins\n"
     ]
    }
   ],
   "source": [
    "from sequence_generation import proteins, peptides\n",
    "test_directory = '../data/testing_output/'\n",
    "\n",
    "num_hybs = 5\n",
    "min_length= 5\n",
    "max_length = 35\n",
    "num_peptides = 100\n",
    "min_cont = 3 #min contribution for each side of a hybrid\n",
    "\n",
    "# make hybrid proteins\n",
    "hyb_prots = proteins.generate_hybrids([x for _, x in database.items()], num_hybs, min_contribution=max_length)\n",
    "# create peptides\n",
    "non_hybrid_peps = peptides.gen_peptides([x for _, x in database.items()], num_peptides, min_length=min_length, max_length=max_length, digest='random', dist='beta')\n",
    "# create hybrid peptides\n",
    "hyb_peps = peptides.gen_peptides(hyb_prots, num_hybs, min_length=min_length, max_length=max_length, digest='random', min_contribution=min_cont, hybrid_list=True)\n",
    "\n",
    "all_proteins_raw = [x for _,x in database.items()] + hyb_prots\n",
    "all_peptides_raw = non_hybrid_peps + hyb_peps\n",
    "\n",
    "peptides = {}\n",
    "for i, pep in enumerate(all_peptides_raw):\n",
    "    peptides[i] = pep\n",
    "    peptides[i]['scan_no'] = i"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 2.1 Save this info so that I can analyze it later from Neo-Fusion"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "import json\n",
    "experiment_info_file_name = 'experiment_info.json'\n",
    "\n",
    "exp = {'database': fasta_file, 'peptides': peptides}\n",
    "with open(test_directory + experiment_info_file_name, 'w') as o:\n",
    "    json.dump(exp, o)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 2.2 Load data if available instead of creating it"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "# import json\n",
    "\n",
    "# expfile = '../data/testing_output/experiment_info.json'\n",
    "# exp = json.load(open(expfile, 'r'))\n",
    "# peptides = exp['peptides']"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 3. Generate spectra"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Determination of memory status is not supported on this \n",
      " platform, measuring for memoryleaks will never fail\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "'../data/testing_output/testSpectraFile.mzML'"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from src.spectra import gen_spectra\n",
    "from src.utils import utils\n",
    "from sequence_generation import write_spectra\n",
    "\n",
    "utils.make_dir(test_directory)\n",
    "\n",
    "spectra = []\n",
    "sorted_keys = [int(c) for c in peptides.keys()]\n",
    "sorted_keys.sort()\n",
    "for k in sorted_keys:\n",
    "    pep = peptides[k]\n",
    "    cont = gen_spectra.gen_spectrum(pep['sequence'])\n",
    "    spec = cont['spectrum']\n",
    "    pm = cont['precursor_mass']\n",
    "    spectra.append({'spectrum': spec, 'precursor_mass': pm})\n",
    "write_spectra.write_mzml('testSpectraFile', spectra, output_dir=test_directory)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 4. Run hypedsearch"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Loading database...\n",
      "Done. Indexing database...\n",
      "Done.\n",
      "Number of 3-mers found in the database: 7487\n",
      "Analyzing spectra file 1/1[0%]\n",
      "\n",
      "Analyzing spectrum 5/345[1%]\r"
     ]
    },
    {
     "ename": "KeyboardInterrupt",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m                         Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-6-79db44868069>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m     22\u001b[0m }\n\u001b[1;32m     23\u001b[0m \u001b[0mst\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mtime\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 24\u001b[0;31m \u001b[0mrunner\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mrun\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0margs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     25\u001b[0m \u001b[0mprint\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'\\nTotal runtime: {} seconds'\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mformat\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mtime\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;34m-\u001b[0m \u001b[0mst\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     26\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/Documents/Layer_Research/hypedsearch/src/runner.py\u001b[0m in \u001b[0;36mrun\u001b[0;34m(args)\u001b[0m\n\u001b[1;32m     35\u001b[0m         \u001b[0;32mbreak\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     36\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 37\u001b[0;31m     \u001b[0mmatched_spectra\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mid_spectra\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mid_spectra\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mspectra_files\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0margs\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'database_file'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mmin_peptide_len\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0margs\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'min_peptide_len'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mmax_peptide_len\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0margs\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'max_peptide_len'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     38\u001b[0m     \u001b[0mprint\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'\\nFinished search. Writting results to {}...'\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mformat\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0margs\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'output_dir'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     39\u001b[0m     \u001b[0msummary\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mgenerate\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mmatched_spectra\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0margs\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'output_dir'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/Documents/Layer_Research/hypedsearch/src/identfication/id_spectra.py\u001b[0m in \u001b[0;36mid_spectra\u001b[0;34m(spectra_files, database_file, verbose, min_peptide_len, max_peptide_len, ppm_tolerance)\u001b[0m\n\u001b[1;32m    126\u001b[0m             \u001b[0;31m# get the good kmers\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    127\u001b[0m             \u001b[0minteresting_kmers\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0msearch_base_kmers\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mspectrum\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mbase_k_mer_masses\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mppm_tolerance\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 128\u001b[0;31m             \u001b[0maligned_spectrum\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mid_spectrum\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mspectrum\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mdatabase\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0minteresting_kmers\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mmin_peptide_len\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mmin_peptide_len\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mmax_peptide_len\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mmax_peptide_len\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    129\u001b[0m             \u001b[0mentry_name\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m'{}_{}'\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mformat\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mspectrum_file\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mspectrum\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mscan_number\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    130\u001b[0m             \u001b[0mresults\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mentry_name\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0maligned_spectrum\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/Documents/Layer_Research/hypedsearch/src/identfication/id_spectra.py\u001b[0m in \u001b[0;36mid_spectrum\u001b[0;34m(spectrum, database, base_kmers, n, min_peptide_len, max_peptide_len)\u001b[0m\n\u001b[1;32m     81\u001b[0m     '''\n\u001b[1;32m     82\u001b[0m     \u001b[0;31m# get the top b and top y scores from the database\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 83\u001b[0;31m     \u001b[0mtop_b\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mtop_y\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0msearch\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0msearch_database\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mspectrum\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mdatabase\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mbase_kmers\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mTOP_N\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     84\u001b[0m     \u001b[0;31m# top b and y are dictionaries with rank as key and values are Scores objects\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     85\u001b[0m     \u001b[0;31m# try and find some alignment from these scores\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/Documents/Layer_Research/hypedsearch/src/identfication/search.py\u001b[0m in \u001b[0;36msearch_database\u001b[0;34m(spectrum, database, base_k_mers, n, ppm_tolerance)\u001b[0m\n\u001b[1;32m     81\u001b[0m     \u001b[0;31m# we have the interesting proteins and associated starting/ending positions\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     82\u001b[0m     \u001b[0;31m# try and extend these as much as possible\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 83\u001b[0;31m     \u001b[0mbest_b\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mextend_spots_of_interest\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mspectrum\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mdatabase\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mb_spots_of_interest\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'b'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mn\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     84\u001b[0m     \u001b[0mbest_y\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mextend_spots_of_interest\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mspectrum\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mdatabase\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0my_spots_of_interest\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'y'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mn\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     85\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/Documents/Layer_Research/hypedsearch/src/identfication/search.py\u001b[0m in \u001b[0;36mextend_spots_of_interest\u001b[0;34m(spectrum, database, spots_of_interest, ion_type, n)\u001b[0m\n\u001b[1;32m     54\u001b[0m         \u001b[0mb_score\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0my_score\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mscoring\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mscore_subsequence\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mspectrum\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mspectrum\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0msoi\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0msequence\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     55\u001b[0m         \u001b[0msk\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mScoredKmer\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mb_score\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0my_score\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0msoi\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 56\u001b[0;31m         \u001b[0mextended_kmer\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mkmer_extension\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mextend_kmer\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mspectrum\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mlocal_entries\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0msoi\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mprotein\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0msk\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mion_type\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mSTALL_LENGTH\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     57\u001b[0m         \u001b[0mextended\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mappend\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mextended_kmer\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     58\u001b[0m     \u001b[0mextended\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0msort\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mkey\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;32mlambda\u001b[0m \u001b[0mx\u001b[0m\u001b[0;34m:\u001b[0m \u001b[0mgetattr\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mx\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0msort_by_key\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mreverse\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;32mTrue\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/Documents/Layer_Research/hypedsearch/src/identfication/kmer_extension.py\u001b[0m in \u001b[0;36mextend_kmer\u001b[0;34m(spectrum, protein, kmer, ion, stall_length)\u001b[0m\n\u001b[1;32m     55\u001b[0m     \u001b[0;31m# keep going until we run out of extension\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     56\u001b[0m     \u001b[0;32mwhile\u001b[0m \u001b[0mstall_length\u001b[0m \u001b[0;34m>\u001b[0m \u001b[0;36m0\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 57\u001b[0;31m         \u001b[0mupdated\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mnew_entry\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mkmer\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mprotein\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mspectrum\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mion\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mion\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     58\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0mgetattr\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mupdated\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mscore_key\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;34m>\u001b[0m \u001b[0mgetattr\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mkmer\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mscore_key\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;32mand\u001b[0m \u001b[0mgetattr\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mupdated\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mscore_key\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;34m>\u001b[0m \u001b[0;36m0\u001b[0m \u001b[0;32mand\u001b[0m \u001b[0mupdated\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mkmer\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mk\u001b[0m \u001b[0;34m!=\u001b[0m \u001b[0mkmer\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mkmer\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mk\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     59\u001b[0m             \u001b[0mlast_maintenance\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mupdated\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/Documents/Layer_Research/hypedsearch/src/identfication/kmer_extension.py\u001b[0m in \u001b[0;36mnew_entry\u001b[0;34m(old_entry, protein, spectrum, ion)\u001b[0m\n\u001b[1;32m     28\u001b[0m     \u001b[0mmer_seq\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mprotein\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0msequence\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mstarting_pos\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0mending_pos\u001b[0m\u001b[0;34m+\u001b[0m\u001b[0;36m1\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     29\u001b[0m     \u001b[0mmer_spec_b\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mgen_spectra\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mgen_spectrum\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mmer_seq\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mion\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m'b'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'spectrum'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 30\u001b[0;31m     \u001b[0mmer_spec_y\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mgen_spectra\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mgen_spectrum\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mmer_seq\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mion\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m'y'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'spectrum'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     31\u001b[0m     \u001b[0mlonger_kmer\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mKmer\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mold_entry\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mkmer\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mk\u001b[0m \u001b[0;34m+\u001b[0m \u001b[0;36m1\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mmer_seq\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mprotein\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mname\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mstarting_pos\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mending_pos\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     32\u001b[0m     \u001b[0mnew_sk\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mScoredKmer\u001b[0m\u001b[0;34m(\u001b[0m \u001b[0mmass_comparisons\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0moptimized_compare_masses\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mspectrum\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mspectrum\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mmer_spec_b\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mmass_comparisons\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0moptimized_compare_masses\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mspectrum\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mspectrum\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mmer_spec_y\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mlonger_kmer\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/Documents/Layer_Research/hypedsearch/src/spectra/gen_spectra.py\u001b[0m in \u001b[0;36mgen_spectrum\u001b[0;34m(sequence, charge, ion)\u001b[0m\n\u001b[1;32m    146\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    147\u001b[0m     \u001b[0mthis_entry\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m{\u001b[0m\u001b[0;34m}\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 148\u001b[0;31m     \u001b[0mmasses\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mpre_mz\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mcalc_masses\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0msequence\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mcharge\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mcharge\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mion\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mion\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    149\u001b[0m     \u001b[0mmasses\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0msort\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    150\u001b[0m     \u001b[0mthis_entry\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'spectrum'\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mmasses\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/Documents/Layer_Research/hypedsearch/src/spectra/gen_spectra.py\u001b[0m in \u001b[0;36mcalc_masses\u001b[0;34m(sequence, charge, ion)\u001b[0m\n\u001b[1;32m    114\u001b[0m     \u001b[0mlength\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mlen\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0msequence\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    115\u001b[0m     \u001b[0mtotal\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mWATER_MASS\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 116\u001b[0;31m     \u001b[0;32mfor\u001b[0m \u001b[0mi\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mrange\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mlength\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    117\u001b[0m         \u001b[0mtotal\u001b[0m \u001b[0;34m+=\u001b[0m  \u001b[0mAMINO_ACIDS\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0msequence\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mi\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    118\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m: "
     ]
    }
   ],
   "source": [
    "import os\n",
    "import sys\n",
    "module_path = os.path.abspath(os.path.join('..'))\n",
    "if module_path not in sys.path:\n",
    "    sys.path.append(module_path)\n",
    "module_path = os.path.abspath(os.path.join('../..'))\n",
    "if module_path not in sys.path:\n",
    "    sys.path.append(module_path)\n",
    "\n",
    "from src import runner\n",
    "from time import time\n",
    "\n",
    "test_directory = '../data/testing_output/'\n",
    "fasta_file = '../data/databases/100prots.fasta'\n",
    "\n",
    "args = {\n",
    "    'spectra_folder': test_directory,\n",
    "    'database_file': fasta_file,\n",
    "    'output_dir': test_directory,\n",
    "    'min_peptide_len': 3,\n",
    "    'max_peptide_len': 35,\n",
    "}\n",
    "st = time()\n",
    "runner.run(args)\n",
    "print('\\nTotal runtime: {} seconds'.format(time() - st))\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 5. Load the summary json"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "../data/testing_output/\n"
     ]
    }
   ],
   "source": [
    "import json\n",
    "test_directory = '../data/testing_output/'\n",
    "\n",
    "summary = json.load(open(test_directory + 'summary.json', 'r'))\n",
    "print(test_directory)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### summary format\n",
    "Each entry is of the form\n",
    "```python\n",
    "{\n",
    "    '<file-name>_<scan_number>':{\n",
    "        'spectrum': {...},\n",
    "        'alignments': [{...}],\n",
    "        'b_scores': [{...}],\n",
    "        'y_score': [{...}]\n",
    "    }\n",
    "}\n",
    "```\n",
    "The only attribute we're really interested in is the alignments attribute which has the form\n",
    "```python\n",
    "{\n",
    "    'alignments': [{\n",
    "        'b_alignment': {...},\n",
    "        'y_alignment': {...}, \n",
    "        'spectrum': [...],\n",
    "        'protein': str,\n",
    "        'alignment_score': float,\n",
    "        'sequence': str,\n",
    "        'hybrid': bool,\n",
    "        'hybrid_sequence': str\n",
    "    }, ...\n",
    "    ]\n",
    "}\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 6. Determine which number of alignments were correct\n",
    "This needs to be broken down into hybrid and non hybrid peptides to get some stats on how well its doing"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "n = 5\n",
    "non_hyb_stats = {i: {\n",
    "        'correct': 0,\n",
    "        'correct_parent': 0,\n",
    "        'correct_length': 0,\n",
    "        'correct_start': 0,\n",
    "        'correct_end': 0\n",
    "    } for i in range(n)}\n",
    "non_hyb_stats['count'] = 0\n",
    "\n",
    "hyb_stats ={i: {\n",
    "        'left_correct_parent': 0,\n",
    "        'right_correct_parent': 0, \n",
    "        'correct_start': 0,\n",
    "        'correct_end': 0,\n",
    "        'correct_length': 0,\n",
    "        'correct_sequence': 0,\n",
    "        'correct': 0\n",
    "    } for i in range(n)}\n",
    "hyb_stats['count'] = 0\n",
    "\n",
    "wrong_hybrid_alignemnts = []"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "def hyb_calc(result, real_pep):\n",
    "    hyb_stats['count'] += 1\n",
    "    is_correct = False\n",
    "    for i in range(min(n, len(result))):\n",
    "        res = result[i]\n",
    "        result_hybrid = bool(res['hybrid'])\n",
    "        if not result_hybrid: \n",
    "            continue\n",
    "        \n",
    "        left_corrparent = real_pep['left_parent_name'] == res['protein'].split('~')[0]\n",
    "        right_corrparent = real_pep['right_parent_name'] == res['protein'].split('~')[1]\n",
    "        corr_start = real_pep['left_parent_starting_position'] == res['b_alignment']['kmer']['start_position']\n",
    "        corr_end = real_pep['right_parent_ending_position'] == res['y_alignment']['kmer']['end_position']\n",
    "        corr_len = len(real_pep['sequence']) == len(res['sequence'])\n",
    "        corr_seq = real_pep['sequence'] == res['sequence']\n",
    "        \n",
    "        if not corr_seq: \n",
    "            print('real pep seq: {} \\t result seq: {}'.format(real_pep['sequence'], res['sequence']))\n",
    "            continue\n",
    "        hyb_stats[i]['left_correct_parent'] += 1 if left_corrparent else 0\n",
    "        hyb_stats[i]['right_correct_parent'] += 1 if right_corrparent else 0\n",
    "        hyb_stats[i]['correct_start'] += 1 if corr_start else 0\n",
    "        hyb_stats[i]['correct_end'] += 1 if corr_end else 0\n",
    "        hyb_stats[i]['correct_length'] += 1 if corr_len else 0\n",
    "        hyb_stats[i]['correct_sequence'] += 1 if corr_seq else 0\n",
    "        hyb_stats[i]['correct'] += 1 if left_corrparent and right_corrparent and corr_start and corr_end and corr_len else 0\n",
    "        \n",
    "        is_correct = corr_seq\n",
    "            \n",
    "    if not is_correct:\n",
    "        wrong_hybrid_alignemnts.append((result, real_pep))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "def non_hyb_calc(result, real_pep):\n",
    "    non_hyb_stats['count'] += 1\n",
    "    iterrange = min(n, len(result))\n",
    "    for i in range(iterrange):\n",
    "        if result[i]['protein'] != real_pep['parent_name']:\n",
    "            continue\n",
    "        corrlen = len(result[i]['sequence']) == len(real_pep['sequence'])\n",
    "        resstartpos = min(result[i]['b_alignment']['kmer']['start_position'], result[i]['y_alignment']['kmer']['start_position'])\n",
    "        resendpos = min(result[i]['b_alignment']['kmer']['end_position'], result[i]['y_alignment']['kmer']['end_position'])\n",
    "        corrstart = resstartpos == real_pep['starting_position']\n",
    "        corrend = resendpos == real_pep['ending_position']\n",
    "        \n",
    "        non_hyb_stats[i]['correct_parent'] += 1 \n",
    "        non_hyb_stats[i]['correct_length'] += 1 if corrlen else 0\n",
    "        non_hyb_stats[i]['correct_start'] += 1 if corrstart else 0\n",
    "        non_hyb_stats[i]['correct_end'] += 1 if corrend else 0\n",
    "        non_hyb_stats[i]['correct'] += 1 if (corrend and corrlen and corrstart) else 0\n",
    "        return\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "real pep seq: LMLATMAS \t result seq: LLMATMAS\n",
      "real pep seq: LMLATMASP \t result seq: LLMATMASP\n",
      "real pep seq: LMLATMASP \t result seq: LLMATMASP\n",
      "real pep seq: LMLATMASPA \t result seq: LLMATMASPA\n",
      "real pep seq: LMLATMASPA \t result seq: LLMATMASPA\n",
      "real pep seq: LMLATMASPAM \t result seq: LLMATMASPAM\n",
      "real pep seq: LMLATMASPAM \t result seq: LLMATMASPAM\n",
      "real pep seq: LMLATMASPAML \t result seq: LLMATMASPAML\n",
      "real pep seq: LMLATMASPAML \t result seq: LLMATMASPAML\n",
      "real pep seq: LMLATMASPAMLG \t result seq: LLMATMASPAMLG\n",
      "real pep seq: LMLATMASPAMLG \t result seq: LLMATMASPAMLG\n",
      "real pep seq: LLMLATMAS \t result seq: LLMLTAMAS\n",
      "real pep seq: LLMLATMASP \t result seq: LLMLTMASP\n",
      "real pep seq: LLMLATMASP \t result seq: LLMLTMASP\n",
      "real pep seq: LLMLATMASPA \t result seq: LLMLTMASPA\n",
      "real pep seq: LLMLATMASPA \t result seq: LLMLTMASPA\n",
      "real pep seq: LLMLATMASPAM \t result seq: LLMLTMASPAM\n",
      "real pep seq: LLMLATMASPAM \t result seq: LLMLTMASPAM\n",
      "real pep seq: LLMLATMASPAML \t result seq: LLMLTMASPAML\n",
      "real pep seq: LLMLATMASPAML \t result seq: LLMLTMASPAML\n",
      "real pep seq: LLMLATMASPAMLG \t result seq: LLMLTMASPAMLG\n",
      "real pep seq: LLMLATMASPAMLG \t result seq: LLMLTMASPAMLG\n",
      "real pep seq: LLLMLATMAS \t result seq: ILILMTAMAS\n",
      "real pep seq: LLLMLATMASP \t result seq: ILILMTMASP\n",
      "real pep seq: LLLMLATMASP \t result seq: LLLMLAMASP\n",
      "real pep seq: LLLMLATMASPA \t result seq: ILILMTMASPA\n",
      "real pep seq: LLLMLATMASPA \t result seq: ILILMASPA\n",
      "real pep seq: LLLMLATMASPAM \t result seq: ILILMTMASPAM\n",
      "real pep seq: LLLMLATMASPAM \t result seq: LLLLTMASPAM\n",
      "real pep seq: LLLMLATMASPAML \t result seq: ILILMTMASPAML\n",
      "real pep seq: LLLMLATMASPAML \t result seq: LLLLTMASPAML\n",
      "real pep seq: LLLMLATMASPAMLG \t result seq: ILILMTMASPAMLG\n",
      "real pep seq: LLLMLATMASPAMLG \t result seq: LLLLTMASPAMLG\n",
      "real pep seq: KLLLMLATMAS \t result seq: KLLLMLATAMAS\n",
      "real pep seq: KLLLMLATMASP \t result seq: KLLLMLAMASP\n",
      "real pep seq: KLLLMLATMASP \t result seq: KLLLMLAMASP\n",
      "real pep seq: KLLLMLATMASPA \t result seq: KLLLMLAASPA\n",
      "real pep seq: KLLLMLATMASPA \t result seq: KLLLMLASPA\n",
      "real pep seq: KLLLMLATMASPAM \t result seq: KILITMASPAM\n",
      "real pep seq: KLLLMLATMASPAM \t result seq: KLLITMASPAM\n",
      "real pep seq: KLLLMLATMASPAML \t result seq: KILITMASPAML\n",
      "real pep seq: KLLLMLATMASPAML \t result seq: KLLITMASPAML\n",
      "real pep seq: KLLLMLATMASPAMLG \t result seq: KILITMASPAMLG\n",
      "real pep seq: KLLLMLATMASPAMLG \t result seq: KLLITMASPAMLG\n",
      "real pep seq: PKLLLMLATMAS \t result seq: PKLLLMLATAMAS\n",
      "real pep seq: PKLLLMLATMASP \t result seq: PKLLLMLAMASP\n",
      "real pep seq: PKLLLMLATMASP \t result seq: PKLLLMLAMASP\n",
      "real pep seq: PKLLLMLATMASPA \t result seq: PKLLLMLAASPA\n",
      "real pep seq: PKLLLMLATMASPA \t result seq: PKLLLMLASPA\n",
      "real pep seq: PKLLLMLATMASPAM \t result seq: PKLLTMASPAM\n",
      "real pep seq: PKLLLMLATMASPAM \t result seq: PKILMDSTMASPAM\n",
      "real pep seq: PKLLLMLATMASPAML \t result seq: PKILMDSTMASPAML\n",
      "real pep seq: PKLLLMLATMASPAML \t result seq: PKLLTMASPAML\n",
      "real pep seq: PKLLLMLATMASPAMLG \t result seq: PKLLTMASPAMLG\n",
      "real pep seq: PKLLLMLATMASPAMLG \t result seq: PKILMDSTMASPAMLG\n",
      "real pep seq: DPKLLLMLATMAS \t result seq: DPKLLLMLATAMAS\n",
      "real pep seq: DPKLLLMLATMASP \t result seq: DPKLLLMLAMASP\n",
      "real pep seq: DPKLLLMLATMASP \t result seq: DPKLLLMLAMASP\n",
      "real pep seq: DPKLLLMLATMASPA \t result seq: DPKLLLMLAASPA\n",
      "real pep seq: DPKLLLMLATMASPA \t result seq: DPKLLLMLASPA\n",
      "real pep seq: DPKLLLMLATMASPAM \t result seq: DPKLLLMLAPAM\n",
      "real pep seq: DPKLLLMLATMASPAM \t result seq: DPKLLLMLAPAM\n",
      "real pep seq: DPKLLLMLATMASPAML \t result seq: DPKTFTMASPAML\n",
      "real pep seq: DPKLLLMLATMASPAML \t result seq: DPKLLLMLAPAML\n",
      "real pep seq: DPKLLLMLATMASPAMLG \t result seq: DPKLLLMLAAMLG\n",
      "real pep seq: DPKLLLMLATMASPAMLG \t result seq: DPKLLLMLAMLG\n",
      "real pep seq: TLWVFW \t result seq: TIWVFW\n",
      "real pep seq: TLWVFW \t result seq: TIWVFW\n",
      "real pep seq: TLWVFW \t result seq: TIWVFW\n",
      "real pep seq: TLWVFWT \t result seq: TIWVFWT\n",
      "real pep seq: TLWVFWT \t result seq: TIWVFWT\n",
      "real pep seq: TLWVFWTI \t result seq: TIWVFWTI\n",
      "real pep seq: TLWVFWTI \t result seq: TIWVFWTI\n",
      "real pep seq: TLWVFWTIY \t result seq: TIWVFWTIY\n",
      "real pep seq: TLWVFWTIY \t result seq: TIWVFWTIY\n",
      "real pep seq: TLWVFWTIYA \t result seq: TIWVFWTIYA\n",
      "real pep seq: TLWVFWTIYA \t result seq: TIWVFWTIYA\n",
      "real pep seq: TLWVFWTIYAY \t result seq: TIWVFWTIYAY\n",
      "real pep seq: TLWVFWTIYAY \t result seq: TIWVFWTIYAY\n",
      "real pep seq: TLWVFWTIYAYD \t result seq: TIWVFWTIYAYD\n",
      "real pep seq: TLWVFWTIYAYD \t result seq: TIWVFWTIYAYD\n",
      "real pep seq: VTLWVFWT \t result seq: VTLSVFWT\n",
      "real pep seq: VTLWVFWTI \t result seq: VTLSVFWTI\n",
      "real pep seq: VTLWVFWTIY \t result seq: VTLSVFWTIY\n",
      "real pep seq: VTLWVFWTIYA \t result seq: VTLSVFWTIYA\n",
      "real pep seq: VTLWVFWTIYAY \t result seq: VTLSVFWTIYAY\n",
      "real pep seq: VTLWVFWTIYAYD \t result seq: VTLSVFWTIYAYD\n",
      "real pep seq: VVTLWVFWT \t result seq: VVTIVFWT\n",
      "real pep seq: VVTLWVFWTI \t result seq: VVTIVFWTI\n",
      "real pep seq: VVTLWVFWTIY \t result seq: VVTIVFWTIY\n",
      "real pep seq: VVTLWVFWTIYA \t result seq: VVTIVFWTIYA\n",
      "real pep seq: VVTLWVFWTIYAY \t result seq: VVTIVFWTIYAY\n",
      "real pep seq: VVTLWVFWTIYAYD \t result seq: VVTIVFWTIYAYD\n",
      "real pep seq: EVVTLWVFWT \t result seq: EVVTLWEWT\n",
      "real pep seq: EVVTLWVFWT \t result seq: EVVTLWEWT\n",
      "real pep seq: EVVTLWVFWTI \t result seq: EVVTLWWTL\n",
      "real pep seq: EVVTLWVFWTI \t result seq: EVVTLWWTL\n",
      "real pep seq: EVVTLWVFWTIY \t result seq: EVVTLWTLY\n",
      "real pep seq: EVVTLWVFWTIY \t result seq: EVVTLWTLY\n",
      "real pep seq: EVVTLWVFWTIYA \t result seq: EVVEVFWTIYA\n",
      "real pep seq: EVVTLWVFWTIYA \t result seq: EVVVFWTIYA\n",
      "real pep seq: EVVTLWVFWTIYAY \t result seq: EVVEVFWTIYAY\n",
      "real pep seq: EVVTLWVFWTIYAY \t result seq: EVVVFWTIYAY\n",
      "real pep seq: EVVTLWVFWTIYAYD \t result seq: EVVEVFWTIYAYD\n",
      "real pep seq: EVVTLWVFWTIYAYD \t result seq: EVVVFWTIYAYD\n",
      "real pep seq: HEVVTLWVFWT \t result seq: HEVVTLWDLAIEPAH\n",
      "real pep seq: HEVVTLWVFWT \t result seq: HEVVTLWLGELFT\n",
      "real pep seq: HEVVTLWVFWTI \t result seq: HEVVTLWWTL\n",
      "real pep seq: HEVVTLWVFWTI \t result seq: HEVVTLWWTL\n",
      "real pep seq: HEVVTLWVFWTIY \t result seq: HEVVTLWTLY\n",
      "real pep seq: HEVVTLWVFWTIY \t result seq: HEVVTLWTLY\n",
      "real pep seq: HEVVTLWVFWTIYA \t result seq: HEVVFWTIYA\n",
      "real pep seq: HEVVTLWVFWTIYA \t result seq: HEVVFWTIYA\n",
      "real pep seq: HEVVTLWVFWTIYAY \t result seq: HEVVFWTIYAY\n",
      "real pep seq: HEVVTLWVFWTIYAY \t result seq: HEVVFWTIYAY\n",
      "real pep seq: HEVVTLWVFWTIYAYD \t result seq: HEVVFWTIYAYD\n",
      "real pep seq: HEVVTLWVFWTIYAYD \t result seq: HEVVFWTIYAYD\n",
      "real pep seq: THEVVTLWVFWT \t result seq: THEVVTLWVVWY\n",
      "real pep seq: THEVVTLWVFWT \t result seq: THEVVTLWISVIPHD\n",
      "real pep seq: THEVVTLWVFWTI \t result seq: THEVVTLWVVWVY\n",
      "real pep seq: THEVVTLWVFWTI \t result seq: THEVVTLWVVWY\n",
      "real pep seq: THEVVTLWVFWTIY \t result seq: THEVVTLWVVWY\n",
      "real pep seq: THEVVTLWVFWTIY \t result seq: THEVVTLWTLY\n",
      "real pep seq: THEVVTLWVFWTIYA \t result seq: THEVVTLWLYA\n",
      "real pep seq: THEVVTLWVFWTIYA \t result seq: THEVVTLWLYA\n",
      "real pep seq: THEVVTLWVFWTIYAY \t result seq: THEVVTLWVVWY\n",
      "real pep seq: THEVVTLWVFWTIYAY \t result seq: THEVVTLWVWY\n",
      "real pep seq: THEVVTLWVFWTIYAYD \t result seq: THEVFWTIYAYD\n",
      "real pep seq: THEVVTLWVFWTIYAYD \t result seq: THEVFWTIYAYD\n",
      "real pep seq: YTHEVVTLWVFWT \t result seq: YTHEVVTLWLDPKWT\n",
      "real pep seq: YTHEVVTLWVFWT \t result seq: YTHEVVTLWEWT\n",
      "real pep seq: YTHEVVTLWVFWTI \t result seq: YTHEVVTLWDHMLTL\n",
      "real pep seq: YTHEVVTLWVFWTI \t result seq: YTHEVVTLWFTWL\n",
      "real pep seq: YTHEVVTLWVFWTIY \t result seq: YTHEVVTLWTLY\n",
      "real pep seq: YTHEVVTLWVFWTIY \t result seq: YTHEVVTLWTLY\n",
      "real pep seq: YTHEVVTLWVFWTIYA \t result seq: YTHEVVTLWLYA\n",
      "real pep seq: YTHEVVTLWVFWTIYA \t result seq: YTHEVVTLWLYA\n",
      "real pep seq: YTHEVVTLWVFWTIYAY \t result seq: YTHEVVTLWYAY\n",
      "real pep seq: YTHEVVTLWVFWTIYAY \t result seq: YTHEVVTLWAITPPQQPY\n",
      "real pep seq: YTHEVVTLWVFWTIYAYD \t result seq: YHTEIVFWTIYAYD\n",
      "real pep seq: YTHEVVTLWVFWTIYAYD \t result seq: YTHEVVTLWTLGENPD\n",
      "real pep seq: LPQVSL \t result seq: IPQVSL\n",
      "real pep seq: LPQVSL \t result seq: IPQVSL\n",
      "real pep seq: LPQVSL \t result seq: IPQVSL\n",
      "real pep seq: LPQVSLL \t result seq: IPQVSLL\n",
      "real pep seq: LPQVSLL \t result seq: IPQVSLL\n",
      "real pep seq: LPQVSLL \t result seq: IPQVSLL\n",
      "real pep seq: LPQVSLLP \t result seq: IPQVSLLP\n",
      "real pep seq: LPQVSLLP \t result seq: IPQVSLLP\n",
      "real pep seq: LPQVSLLP \t result seq: IPQVSLLP\n",
      "real pep seq: LPQVSLLPS \t result seq: TITISVSLLPS\n",
      "real pep seq: LPQVSLLPS \t result seq: TTLLVSLLPS\n",
      "real pep seq: LPQVSLLPS \t result seq: TTLLVSLLPS\n",
      "real pep seq: LPQVSLLPSS \t result seq: IPQVSLLPSS\n",
      "real pep seq: LPQVSLLPSS \t result seq: IPQVSLLPSS\n",
      "real pep seq: LPQVSLLPSS \t result seq: IPQVSLLPSS\n",
      "real pep seq: LPQVSLLPSSQ \t result seq: IPQVSLLPSSQ\n",
      "real pep seq: LPQVSLLPSSQ \t result seq: IPQVSLLPSSQ\n",
      "real pep seq: LPQVSLLPSSQ \t result seq: IPQVSLLPSSQ\n",
      "real pep seq: LPQVSLLPSSQT \t result seq: IPQVSLLPSSQT\n",
      "real pep seq: LPQVSLLPSSQT \t result seq: IPQVSLLPSSQT\n",
      "real pep seq: LPQVSLLPSSQT \t result seq: IPQVSLLPSSQT\n",
      "real pep seq: HLPQVSLLP \t result seq: HLPQSIIP\n",
      "real pep seq: HLPQVSLLP \t result seq: HLPQSLLP\n",
      "real pep seq: HLPQVSLLPS \t result seq: TITISVSLLPS\n",
      "real pep seq: HLPQVSLLPS \t result seq: TTLLVSLLPS\n",
      "real pep seq: HLPQVSLLPSS \t result seq: HHKLSVSLLPSS\n",
      "real pep seq: HLPQVSLLPSS \t result seq: HHKLSSILPSS\n",
      "real pep seq: HLPQVSLLPSSQ \t result seq: HIPVSLLPSSQ\n",
      "real pep seq: HLPQVSLLPSSQ \t result seq: HIPVSLLPSSQ\n",
      "real pep seq: HLPQVSLLPSSQT \t result seq: HIPVSLLPSSQT\n",
      "real pep seq: HLPQVSLLPSSQT \t result seq: HIPVSLLPSSQT\n",
      "real pep seq: VHLPQVSLLP \t result seq: VHLPQSIIP\n",
      "real pep seq: VHLPQVSLLP \t result seq: VHLPQSLLP\n",
      "real pep seq: VHLPQVSLLPS \t result seq: VHLPQSILPS\n",
      "real pep seq: VHLPQVSLLPS \t result seq: TITISVSLLPS\n",
      "real pep seq: VHLPQVSLLPSS \t result seq: VHLPQSILPSS\n",
      "real pep seq: VHLPQVSLLPSS \t result seq: VHIPVSLLPSS\n",
      "real pep seq: VHLPQVSLLPSSQ \t result seq: VHIPVSLLPSSQ\n",
      "real pep seq: VHLPQVSLLPSSQ \t result seq: VHLVSLLPSSQ\n",
      "real pep seq: VHLPQVSLLPSSQT \t result seq: VHIPVSLLPSSQT\n",
      "real pep seq: VHLPQVSLLPSSQT \t result seq: VHLREVSLLPSSQT\n",
      "real pep seq: MVHLPQVSLLP \t result seq: MVHLPQSIIP\n",
      "real pep seq: MVHLPQVSLLP \t result seq: MVHLPQSLLP\n",
      "real pep seq: MVHLPQVSLLPS \t result seq: MVHLPQSILPS\n",
      "real pep seq: MVHLPQVSLLPS \t result seq: TITISVSLLPS\n",
      "real pep seq: MVHLPQVSLLPSS \t result seq: MVHLPQSILPSS\n",
      "real pep seq: MVHLPQVSLLPSS \t result seq: MVHVSLLPSS\n",
      "real pep seq: MVHLPQVSLLPSSQ \t result seq: MVHVSLLPSSQ\n",
      "real pep seq: MVHLPQVSLLPSSQ \t result seq: MVHVSLLPSSQ\n",
      "real pep seq: MVHLPQVSLLPSSQT \t result seq: MMKAGIIVSLLPSSQT\n",
      "real pep seq: MVHLPQVSLLPSSQT \t result seq: MVHVSLLPSSQT\n",
      "real pep seq: AMVHLPQVSLLP \t result seq: AMVHLPQSIIP\n",
      "real pep seq: AMVHLPQVSLLP \t result seq: AMVHLPQSLLP\n",
      "real pep seq: AMVHLPQVSLLPS \t result seq: TITISFNVSLLPS\n",
      "real pep seq: AMVHLPQVSLLPS \t result seq: AMVHLPQFNVSLLPS\n",
      "real pep seq: AMVHLPQVSLLPS \t result seq: AMVHLPQSILPS\n",
      "real pep seq: AMVHLPQVSLLPSS \t result seq: AMVHLPQFNVSLLPSS\n",
      "real pep seq: AMVHLPQVSLLPSS \t result seq: AMVHLPQSILPSS\n",
      "real pep seq: AMVHLPQVSLLPSS \t result seq: AMVCFNVSLLPSS\n",
      "real pep seq: AMVHLPQVSLLPSSQ \t result seq: AMVCVSLLPSSQ\n",
      "real pep seq: AMVHLPQVSLLPSSQ \t result seq: AMVHLPQPSSQ\n",
      "real pep seq: AMVHLPQVSLLPSSQT \t result seq: AMVCVSLLPSSQT\n",
      "real pep seq: AMVHLPQVSLLPSSQT \t result seq: AMVVSLLPSSQT\n",
      "real pep seq: LAMVHLPQVSLLP \t result seq: LAMVHLPQLHWLP\n",
      "real pep seq: LAMVHLPQVSLLP \t result seq: LAMVHLPQSIIP\n",
      "real pep seq: LAMVHLPQVSLLPS \t result seq: LAMVHLPQSILPS\n",
      "real pep seq: LAMVHLPQVSLLPS \t result seq: LAMVHLPQLLPS\n",
      "real pep seq: LAMVHLPQVSLLPSS \t result seq: LAMVHLPQSILPSS\n",
      "real pep seq: LAMVHLPQVSLLPSS \t result seq: LAMVHLPQLPSS\n",
      "real pep seq: LAMVHLPQVSLLPSSQ \t result seq: LAMVHLPQGGGVSLLPSSQ\n",
      "real pep seq: LAMVHLPQVSLLPSSQ \t result seq: LAMVHLPQGGGPSSQ\n",
      "real pep seq: LAMVHLPQVSLLPSSQ \t result seq: LAMVVSLLPSSQ\n",
      "real pep seq: LAMVHLPQVSLLPSSQT \t result seq: IAMVSLLPSSQT\n",
      "real pep seq: LAMVHLPQVSLLPSSQT \t result seq: IAMVSLLPSSQT\n",
      "real pep seq: ALAMVHLPQVSL \t result seq: ALAMVHLPQQVSL\n",
      "real pep seq: ALAMVHLPQVSLLP \t result seq: ALAMVHLPQSIIP\n",
      "real pep seq: ALAMVHLPQVSLLP \t result seq: ALAMVHLPQSLLP\n",
      "real pep seq: ALAMVHLPQVSLLPS \t result seq: ALAMVHLPQSILPS\n",
      "real pep seq: ALAMVHLPQVSLLPS \t result seq: ALAMVHLPQLLPS\n",
      "real pep seq: ALAMVHLPQVSLLPSS \t result seq: ALAMVHLPQSILPSS\n",
      "real pep seq: ALAMVHLPQVSLLPSS \t result seq: ALAMVHLPQLPSS\n",
      "real pep seq: ALAMVHLPQVSLLPSSQ \t result seq: ALAMVHLPQPSSQ\n",
      "real pep seq: ALAMVHLPQVSLLPSSQ \t result seq: ALAMVVSLLPSSQ\n",
      "real pep seq: ALAMVHLPQVSLLPSSQT \t result seq: ALAMVVSLLPSSQT\n",
      "real pep seq: ALAMVHLPQVSLLPSSQT \t result seq: ALAMVHLPQSSQT\n",
      "real pep seq: LLGSSS \t result seq: LLGSSGSSS\n",
      "real pep seq: LLGSSS \t result seq: LLGSSGSSS\n",
      "real pep seq: LLGSSS \t result seq: LLGSSGSSS\n",
      "real pep seq: LLGSSSGRF \t result seq: LLGSSSSSGRF\n",
      "real pep seq: LLGSSSGRF \t result seq: LLGSSSSGRF\n",
      "real pep seq: LLGSSSGRF \t result seq: LLGSSSSGRF\n",
      "real pep seq: LLGSSSGRFV \t result seq: LLGSSSSSGRFV\n",
      "real pep seq: LLGSSSGRFV \t result seq: LLGSSSSGRFV\n",
      "real pep seq: LLGSSSGRFV \t result seq: LLGSSSSGRFV\n",
      "real pep seq: LLGSSSGRFVR \t result seq: LLGSSSSSGRFVR\n",
      "real pep seq: LLGSSSGRFVR \t result seq: LLGSSSSGRFVR\n",
      "real pep seq: LLGSSSGRFVR \t result seq: LLGSSSSGRFVR\n",
      "real pep seq: LLGSSSGRFVRL \t result seq: LLGSSSSSGRFVRL\n",
      "real pep seq: LLGSSSGRFVRL \t result seq: LLGSSSSGRFVRL\n",
      "real pep seq: LLGSSSGRFVRL \t result seq: LLGSSSSGRFVRL\n",
      "real pep seq: QQLLGSSS \t result seq: QLQLGSSS\n",
      "real pep seq: QQLLGSSS \t result seq: QLQLGSSS\n",
      "real pep seq: QQLLGSSSGR \t result seq: QQLLGSSSRG\n",
      "real pep seq: QQLLGSSSGR \t result seq: QQLLQQSSSGR\n",
      "real pep seq: QQLLGSSSGRF \t result seq: QQLLQQSSSGRF\n",
      "real pep seq: QQLLGSSSGRF \t result seq: AGLQLSSSGRF\n",
      "real pep seq: QQLLGSSSGRFV \t result seq: QQLLQQSSSGRFV\n",
      "real pep seq: QQLLGSSSGRFV \t result seq: AGLQLSSSGRFV\n",
      "real pep seq: QQLLGSSSGRFVR \t result seq: QQLLQQSSSGRFVR\n",
      "real pep seq: QQLLGSSSGRFVR \t result seq: AGLQLSSSGRFVR\n",
      "real pep seq: QQLLGSSSGRFVRL \t result seq: QQLLQQSSSGRFVRL\n",
      "real pep seq: QQLLGSSSGRFVRL \t result seq: AGLQLSSSGRFVRL\n",
      "real pep seq: CQQLLGSSS \t result seq: CQQLLGGSSS\n",
      "real pep seq: CQQLLGSSS \t result seq: CQQLLGGSSS\n",
      "real pep seq: CQQLLGSSSGR \t result seq: CQQLLGSSSRG\n",
      "real pep seq: CQQLLGSSSGR \t result seq: CQQLLGSGR\n",
      "real pep seq: CQQLLGSSSGRF \t result seq: CQQLSSSGRF\n",
      "real pep seq: CQQLLGSSSGRF \t result seq: CQQSSSGRF\n",
      "real pep seq: CQQLLGSSSGRFV \t result seq: CQQLSSSGRFV\n",
      "real pep seq: CQQLLGSSSGRFV \t result seq: CQQLLGRFV\n",
      "real pep seq: CQQLLGSSSGRFVR \t result seq: CQQLSSSGRFVR\n",
      "real pep seq: CQQLLGSSSGRFVR \t result seq: CQQLLGRFVR\n",
      "real pep seq: CQQLLGSSSGRFVRL \t result seq: CQQLSSSGRFVRL\n",
      "real pep seq: CQQLLGSSSGRFVRL \t result seq: CQGALSSSGRFVRL\n",
      "real pep seq: SCQQLLGSSS \t result seq: SCQQLLGGSSS\n",
      "real pep seq: SCQQLLGSSS \t result seq: SCQQLLGGSSS\n",
      "real pep seq: SCQQLLGSSSGR \t result seq: SCQQLLGSSSRG\n",
      "real pep seq: SCQQLLGSSSGR \t result seq: SCQQLLGSGR\n",
      "real pep seq: SCQQLLGSSSGRF \t result seq: SCQQLLGGRF\n",
      "real pep seq: SCQQLLGSSSGRF \t result seq: SCQQLLGGRF\n",
      "real pep seq: SCQQLLGSSSGRFV \t result seq: SCQQLLGRFV\n",
      "real pep seq: SCQQLLGSSSGRFV \t result seq: SCQQLLGRFV\n",
      "real pep seq: SCQQLLGSSSGRFVR \t result seq: SCQQLLGRFVR\n",
      "real pep seq: SCQQLLGSSSGRFVR \t result seq: SCQSSSGRFVR\n",
      "real pep seq: SCQQLLGSSSGRFVRL \t result seq: SCQSSSGRFVRL\n",
      "real pep seq: SCQQLLGSSSGRFVRL \t result seq: SCQSSSGRFVRL\n",
      "real pep seq: RSCQQLLGSSS \t result seq: RSCQQLLGGSSS\n",
      "real pep seq: RSCQQLLGSSS \t result seq: RSCQQLLGGSSS\n",
      "real pep seq: RSCQQLLGSSSGR \t result seq: RSCQQLLGSSSRG\n",
      "real pep seq: RSCQQLLGSSSGR \t result seq: RSCQQLLGSGR\n",
      "real pep seq: RSCQQLLGSSSGRF \t result seq: RSCQQLLGGRF\n",
      "real pep seq: RSCQQLLGSSSGRF \t result seq: RSCQQLLGGRF\n",
      "real pep seq: RSCQQLLGSSSGRFV \t result seq: RSCQQLLGRFV\n",
      "real pep seq: RSCQQLLGSSSGRFV \t result seq: RSCQQLLGRFV\n",
      "real pep seq: RSCQQLLGSSSGRFVR \t result seq: RSCQQLLGRFVR\n",
      "real pep seq: RSCQQLLGSSSGRFVR \t result seq: RSCQQLLGFVR\n",
      "real pep seq: RSCQQLLGSSSGRFVRL \t result seq: RSCQQLLGFVRI\n",
      "real pep seq: RSCQQLLGSSSGRFVRL \t result seq: RSCSSSGRFVRL\n",
      "real pep seq: QRSCQQLLGSSS \t result seq: QRSCQQLLGGSSS\n",
      "real pep seq: QRSCQQLLGSSS \t result seq: QRSCQQLLGGSSS\n",
      "real pep seq: QRSCQQLLGSSSGR \t result seq: QRSCQQLLGSSSRG\n",
      "real pep seq: QRSCQQLLGSSSGR \t result seq: QRSCQQLLGSGR\n",
      "real pep seq: QRSCQQLLGSSSGRF \t result seq: QRSCQQLLGGRF\n",
      "real pep seq: QRSCQQLLGSSSGRF \t result seq: QRSCQQLLGGRF\n",
      "real pep seq: QRSCQQLLGSSSGRFV \t result seq: QRSCQQLLGRFV\n",
      "real pep seq: QRSCQQLLGSSSGRFV \t result seq: QRSCQQLLGRFV\n",
      "real pep seq: QRSCQQLLGSSSGRFVR \t result seq: QRSCQQLLGRFVR\n",
      "real pep seq: QRSCQQLLGSSSGRFVR \t result seq: QRSCQQLLGFVR\n",
      "real pep seq: QRSCQQLLGSSSGRFVRL \t result seq: QRSCQQLLGFVRI\n",
      "real pep seq: QRSCQQLLGSSSGRFVRL \t result seq: QRSCSSSGRFVRL\n",
      "real pep seq: NSMIDRS \t result seq: NSMLDRS\n",
      "real pep seq: NSMIDRS \t result seq: NSMLDRS\n",
      "real pep seq: NSMIDRS \t result seq: NSMLDRS\n",
      "real pep seq: NSMIDRSN \t result seq: GGSMIDRSN\n",
      "real pep seq: NSMIDRSNSM \t result seq: NSMLIDRSNSM\n",
      "real pep seq: NSMIDRSNSMS \t result seq: NSMLIDRSNSMS\n",
      "real pep seq: NSMIDRSNSMSK \t result seq: NSMLIDRSNSMSK\n",
      "real pep seq: SNSMIDRS \t result seq: SNSMLDRS\n",
      "real pep seq: SNSMIDRS \t result seq: SNSMLDRS\n",
      "real pep seq: SNSMIDRSN \t result seq: FDLDLMIDRSN\n",
      "real pep seq: SNSMIDRSN \t result seq: SNSNIDRSN\n",
      "real pep seq: SNSMIDRSNS \t result seq: SNSNIDRSNS\n",
      "real pep seq: SNSMIDRSNS \t result seq: SNSIDRSNS\n",
      "real pep seq: SNSMIDRSNSM \t result seq: SNSNIDRSNSM\n",
      "real pep seq: SNSMIDRSNSM \t result seq: SNSIDRSNSM\n",
      "real pep seq: SNSMIDRSNSM \t result seq: SNSIDRSNSM\n",
      "real pep seq: SNSMIDRSNSMS \t result seq: SNSNIDRSNSMS\n",
      "real pep seq: SNSMIDRSNSMS \t result seq: SNSIDRSNSMS\n",
      "real pep seq: SNSMIDRSNSMS \t result seq: SNSIDRSNSMS\n",
      "real pep seq: SNSMIDRSNSMSK \t result seq: SNSNIDRSNSMSK\n",
      "real pep seq: SNSMIDRSNSMSK \t result seq: SNSIDRSNSMSK\n",
      "real pep seq: SNSMIDRSNSMSK \t result seq: SNSIDRSNSMSK\n",
      "real pep seq: RSNSMIDRS \t result seq: RSNSMLDRS\n",
      "real pep seq: RSNSMIDRS \t result seq: RSNSMLDRS\n",
      "real pep seq: RSNSMIDRSN \t result seq: FDLDLMIDRSN\n",
      "real pep seq: RSNSMIDRSN \t result seq: RSNSIDRSN\n",
      "real pep seq: RSNSMIDRSNS \t result seq: RSNSIDRSNS\n",
      "real pep seq: RSNSMIDRSNS \t result seq: RSNSIDRSNS\n",
      "real pep seq: RSNSMIDRSNSM \t result seq: RSNSIDRSNSM\n",
      "real pep seq: RSNSMIDRSNSM \t result seq: RSNSIDRSNSM\n",
      "real pep seq: RSNSMIDRSNSM \t result seq: RSNSIDRSNSM\n",
      "real pep seq: RSNSMIDRSNSMS \t result seq: RSNSIDRSNSMS\n",
      "real pep seq: RSNSMIDRSNSMS \t result seq: RSNSIDRSNSMS\n",
      "real pep seq: RSNSMIDRSNSMS \t result seq: RSNSIDRSNSMS\n",
      "real pep seq: RSNSMIDRSNSMSK \t result seq: RSNSIDRSNSMSK\n",
      "real pep seq: RSNSMIDRSNSMSK \t result seq: RSNSIDRSNSMSK\n",
      "real pep seq: RSNSMIDRSNSMSK \t result seq: RSNSIDRSNSMSK\n",
      "real pep seq: DRSNSMIDRS \t result seq: DRSNSMLDRS\n",
      "real pep seq: DRSNSMIDRS \t result seq: DRSNSMLDRS\n",
      "real pep seq: DRSNSMIDRSN \t result seq: DRSNSMRSN\n",
      "real pep seq: DRSNSMIDRSN \t result seq: DRSNSMRSN\n",
      "real pep seq: DRSNSMIDRSNS \t result seq: DRSNSMRSNS\n",
      "real pep seq: DRSNSMIDRSNS \t result seq: DRSNSMRSNS\n",
      "real pep seq: DRSNSMIDRSNSM \t result seq: DRSNSM\n",
      "real pep seq: DRSNSMIDRSNSM \t result seq: DRSNSM\n",
      "real pep seq: DRSNSMIDRSNSM \t result seq: DRSNSM\n",
      "real pep seq: DRSNSMIDRSNSMS \t result seq: DRSIDRSNSMS\n",
      "real pep seq: DRSNSMIDRSNSMS \t result seq: DRSIDRSNSMS\n",
      "real pep seq: DRSNSMIDRSNSMS \t result seq: DRSIDRSNSMS\n",
      "real pep seq: DRSNSMIDRSNSMSK \t result seq: DRSIDRSNSMSK\n",
      "real pep seq: DRSNSMIDRSNSMSK \t result seq: DRSIDRSNSMSK\n",
      "real pep seq: DRSNSMIDRSNSMSK \t result seq: DRSIDRSNSMSK\n",
      "real pep seq: IDRSNSMIDRS \t result seq: IDRSNSMLDRS\n",
      "real pep seq: IDRSNSMIDRS \t result seq: IDRSNSMLDRS\n",
      "real pep seq: IDRSNSMIDRS \t result seq: IDRSNSMDRS\n",
      "real pep seq: IDRSNSMIDRSN \t result seq: IDRSNSMRSN\n",
      "real pep seq: IDRSNSMIDRSN \t result seq: IDRSNSMRSN\n",
      "real pep seq: IDRSNSMIDRSN \t result seq: IDRSNSMRSN\n",
      "real pep seq: IDRSNSMIDRSNS \t result seq: IDRSNS\n",
      "real pep seq: IDRSNSMIDRSNS \t result seq: IDRSNS\n",
      "real pep seq: IDRSNSMIDRSNS \t result seq: IDRSNSMRSNS\n",
      "real pep seq: IDRSNSMIDRSNSM \t result seq: LDRSNSM\n",
      "real pep seq: IDRSNSMIDRSNSM \t result seq: LDRSNSM\n",
      "real pep seq: IDRSNSMIDRSNSMS \t result seq: LDRSIDRSNSMS\n",
      "real pep seq: IDRSNSMIDRSNSMS \t result seq: LDRSIDRSNSMS\n",
      "real pep seq: IDRSNSMIDRSNSMSK \t result seq: LDRSIDRSNSMSK\n",
      "real pep seq: IDRSNSMIDRSNSMSK \t result seq: LDRSIDRSNSMSK\n",
      "real pep seq: LIDRSNSMIDRS \t result seq: LIDRSNSMLDRS\n",
      "real pep seq: LIDRSNSMIDRS \t result seq: LIDRSNSMLDRS\n",
      "real pep seq: LIDRSNSMIDRS \t result seq: LIDRSNSMDRS\n",
      "real pep seq: LIDRSNSMIDRSN \t result seq: LIDRSNSMRSN\n",
      "real pep seq: LIDRSNSMIDRSN \t result seq: LIDRSNSMRSN\n",
      "real pep seq: LIDRSNSMIDRSN \t result seq: LIDRSNSMRSN\n",
      "real pep seq: LIDRSNSMIDRSNS \t result seq: LLDRSNS\n",
      "real pep seq: LIDRSNSMIDRSNS \t result seq: LLDRSNS\n",
      "real pep seq: LIDRSNSMIDRSNS \t result seq: LIDRSNS\n",
      "real pep seq: LIDRSNSMIDRSNSM \t result seq: LIDRSNSMNSM\n",
      "real pep seq: LIDRSNSMIDRSNSM \t result seq: LIDRSNSMNSM\n",
      "real pep seq: LIDRSNSMIDRSNSMS \t result seq: LLDRIDRSNSMS\n",
      "real pep seq: LIDRSNSMIDRSNSMS \t result seq: LLDRIDRSNSMS\n",
      "real pep seq: LIDRSNSMIDRSNSMSK \t result seq: LLDRIDRSNSMSK\n",
      "real pep seq: LIDRSNSMIDRSNSMSK \t result seq: LLDRIDRSNSMSK\n",
      "real pep seq: FLIDRSNSMIDRS \t result seq: FLIDRSNSMLDRS\n",
      "real pep seq: FLIDRSNSMIDRS \t result seq: FLIDRSNSMLDRS\n",
      "real pep seq: FLIDRSNSMIDRS \t result seq: FLIDRSNSMRDLKN\n",
      "real pep seq: FLIDRSNSMIDRSN \t result seq: FLIDRSNSMDLKN\n",
      "real pep seq: FLIDRSNSMIDRSN \t result seq: FLIDRSNSMRDLKN\n",
      "real pep seq: FLIDRSNSMIDRSN \t result seq: FLIDRSNSMIFLIDRSN\n",
      "real pep seq: FLIDRSNSMIDRSNS \t result seq: FLLDRSNS\n",
      "real pep seq: FLIDRSNSMIDRSNS \t result seq: FLLDRSNS\n",
      "real pep seq: FLIDRSNSMIDRSNS \t result seq: FLLDRSNS\n",
      "real pep seq: FLIDRSNSMIDRSNSM \t result seq: FLIDRSNSMRDLKN\n",
      "real pep seq: FLIDRSNSMIDRSNSM \t result seq: FLLDRIDRSNSM\n",
      "real pep seq: FLIDRSNSMIDRSNSMS \t result seq: FLLDRIDRSNSMS\n",
      "real pep seq: FLIDRSNSMIDRSNSMS \t result seq: FLIDRSNSMRDLKN\n",
      "real pep seq: FLIDRSNSMIDRSNSMSK \t result seq: FLLDRIDRSNSMSK\n",
      "real pep seq: FLIDRSNSMIDRSNSMSK \t result seq: FLIDRSNSMDINK\n"
     ]
    }
   ],
   "source": [
    "expfile = '../data/testing_output/experiment_info.json'\n",
    "exp = json.load(open(expfile, 'r'))\n",
    "\n",
    "scan_no_keyed_results = {x['spectrum']['scan_number']: x for _, x in summary.items()}\n",
    "sorted_keys = [int(c) for c in exp['peptides'].keys()]\n",
    "\n",
    "for k in sorted_keys:\n",
    "    pep = exp['peptides'][str(k)]\n",
    "    if k not in scan_no_keyed_results:\n",
    "        continue\n",
    "    if 'hybrid' in pep['peptide_name'].lower():\n",
    "        hyb_calc(scan_no_keyed_results[k]['alignments'], pep)\n",
    "    else:\n",
    "        non_hyb_calc(scan_no_keyed_results[k]['alignments'], pep)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "NON HYBRID STATS\n",
      "------------------------------------------------------------\n",
      "number of peptides.......................................100\n",
      "Top alignment\n",
      "\n",
      "correct alignment.........................................96\n",
      "%.........................................................96\n",
      "correct protein...........................................97\n",
      "%.........................................................97\n",
      "correct starting position.................................96\n",
      "%.........................................................96\n",
      "correct ending position...................................96\n",
      "%.........................................................96\n",
      "correct length............................................96\n",
      "%.........................................................96\n",
      "\n",
      "2 to 5 alignment\n",
      "\n",
      "number of peptides.......................................100\n",
      "correct alignment..........................................0\n",
      "%..........................................................0\n",
      "correct protein............................................0\n",
      "%..........................................................0\n",
      "correct starting position..................................0\n",
      "%..........................................................0\n",
      "correct ending position....................................0\n",
      "%..........................................................0\n",
      "correct length.............................................0\n",
      "%..........................................................0\n",
      "\n",
      "============================================================\n",
      "\n",
      "HYBRID STATS\n",
      "------------------------------------------------------------\n",
      "number of peptides.......................................245\n",
      "Top alignment\n",
      "\n",
      "correct alignment........................................128\n",
      "%.........................................................52\n",
      "correct sequence.........................................192\n",
      "%.........................................................78\n",
      "correct left parent......................................170\n",
      "%.........................................................69\n",
      "correct right parent.....................................144\n",
      "%.........................................................58\n",
      "correct starting position................................170\n",
      "%.........................................................69\n",
      "correct ending position..................................144\n",
      "%.........................................................58\n",
      "correct length...........................................192\n",
      "%.........................................................78\n",
      "\n",
      "2 to 5 alignment\n",
      "\n",
      "correct alignment.........................................20\n",
      "%..........................................................8\n",
      "correct sequence.........................................136\n",
      "%.........................................................55\n",
      "correct left parent.......................................80\n",
      "%.........................................................32\n",
      "correct right parent......................................64\n",
      "%.........................................................26\n",
      "correct starting position.................................80\n",
      "%.........................................................32\n",
      "correct ending position...................................64\n",
      "%.........................................................26\n",
      "correct length.............................................0\n",
      "%.........................................................55\n",
      "\n"
     ]
    }
   ],
   "source": [
    "percent = lambda a, b: (a* 100 // b)\n",
    "\n",
    "printstat = lambda name, stat: '{}{}\\n'.format(name, str(stat).rjust(60-len(name), '.'))\n",
    "\n",
    "secbreak = ''.join(['=' for _ in range(60)])\n",
    "headbreak = ''.join(['-' for _ in range(60)])\n",
    "nhcount = non_hyb_stats['count']\n",
    "topalign = non_hyb_stats[0]\n",
    "otheralign = {}\n",
    "for i in range(1, n):\n",
    "    for stat in topalign.keys():\n",
    "        if stat not in otheralign:\n",
    "            otheralign[stat] = 0\n",
    "        otheralign[stat] += non_hyb_stats[i][stat]\n",
    "\n",
    "######################## NON HYBRID PRETTY PRINTING ############################\n",
    "\n",
    "nonhybsum = 'NON HYBRID STATS\\n' + headbreak + '\\n'\n",
    "nonhybsum += printstat('number of peptides', nhcount) \n",
    "nonhybsum += 'Top alignment\\n\\n'\n",
    "nonhybsum += printstat('correct alignment', topalign['correct'])\n",
    "nonhybsum += printstat('%', percent(topalign['correct'], nhcount)) \n",
    "nonhybsum += printstat('correct protein', topalign['correct_parent']) \n",
    "nonhybsum += printstat('%', percent(topalign['correct_parent'], nhcount))\n",
    "nonhybsum += printstat('correct starting position', topalign['correct_start'])\n",
    "nonhybsum += printstat('%', percent(topalign['correct_start'], nhcount))\n",
    "nonhybsum += printstat('correct ending position', topalign['correct_end'])\n",
    "nonhybsum += printstat('%', percent(topalign['correct_end'], nhcount))\n",
    "nonhybsum += printstat('correct length', topalign['correct_length'])\n",
    "nonhybsum += printstat('%', percent(topalign['correct_length'], nhcount))\n",
    "\n",
    "nonhybsum += '\\n2 to {} alignment\\n\\n'.format(n)\n",
    "nonhybsum += printstat('number of peptides', nhcount) \n",
    "nonhybsum += printstat('correct alignment', otheralign['correct'])\n",
    "nonhybsum += printstat('%', percent(otheralign['correct'], nhcount)) \n",
    "nonhybsum += printstat('correct protein', otheralign['correct_parent']) \n",
    "nonhybsum += printstat('%', percent(otheralign['correct_parent'], nhcount))\n",
    "nonhybsum += printstat('correct starting position', otheralign['correct_start'])\n",
    "nonhybsum += printstat('%', percent(otheralign['correct_start'], nhcount))\n",
    "nonhybsum += printstat('correct ending position', otheralign['correct_end'])\n",
    "nonhybsum += printstat('%', percent(otheralign['correct_end'], nhcount))\n",
    "nonhybsum += printstat('correct length', otheralign['correct_length'])\n",
    "nonhybsum += printstat('%', percent(otheralign['correct_length'], nhcount))\n",
    "nonhybsum += '\\n' + secbreak + '\\n\\n'\n",
    "\n",
    "############################ HYBRID PRETYY PRINTING ##############################\n",
    "hcount = hyb_stats['count']\n",
    "topalignh = hyb_stats[0]\n",
    "otheralignh = {}\n",
    "for i in range(1, n):\n",
    "    for stat in topalignh.keys():\n",
    "        if stat not in otheralignh:\n",
    "            otheralignh[stat] = 0\n",
    "        otheralignh[stat] += hyb_stats[i][stat]\n",
    "\n",
    "hybsum = 'HYBRID STATS\\n'+ headbreak + '\\n'\n",
    "hybsum += printstat('number of peptides', hcount)\n",
    "hybsum += 'Top alignment\\n\\n'\n",
    "hybsum += printstat('correct alignment', topalignh['correct'])\n",
    "hybsum += printstat('%', percent(topalignh['correct'], hcount))\n",
    "hybsum += printstat('correct sequence', topalignh['correct_sequence'])\n",
    "hybsum += printstat('%', percent(topalignh['correct_sequence'], hcount))\n",
    "hybsum += printstat('correct left parent', topalignh['left_correct_parent'])\n",
    "hybsum += printstat('%', percent(topalignh['left_correct_parent'], hcount))\n",
    "hybsum += printstat('correct right parent', topalignh['right_correct_parent'])\n",
    "hybsum += printstat('%', percent(topalignh['right_correct_parent'], hcount))\n",
    "hybsum += printstat('correct starting position', topalignh['correct_start'])\n",
    "hybsum += printstat('%', percent(topalignh['correct_start'], hcount))\n",
    "hybsum += printstat('correct ending position', topalignh['correct_end'])\n",
    "hybsum += printstat('%', percent(topalignh['correct_end'], hcount))\n",
    "hybsum += printstat('correct length', topalignh['correct_length'])\n",
    "hybsum += printstat('%', percent(topalignh['correct_length'], hcount))\n",
    "\n",
    "hybsum += '\\n2 to {} alignment\\n\\n'.format(n)\n",
    "hybsum += printstat('correct alignment', otheralignh['correct'])\n",
    "hybsum += printstat('%', percent(otheralignh['correct'], hcount))\n",
    "hybsum += printstat('correct sequence', otheralignh['correct_sequence'])\n",
    "hybsum += printstat('%', percent(otheralignh['correct_sequence'], hcount))\n",
    "hybsum += printstat('correct left parent', otheralignh['left_correct_parent'])\n",
    "hybsum += printstat('%', percent(otheralignh['left_correct_parent'], hcount))\n",
    "hybsum += printstat('correct right parent', otheralignh['right_correct_parent'])\n",
    "hybsum += printstat('%', percent(otheralignh['right_correct_parent'], hcount))\n",
    "hybsum += printstat('correct starting position', otheralignh['correct_start'])\n",
    "hybsum += printstat('%', percent(otheralignh['correct_start'], hcount))\n",
    "hybsum += printstat('correct ending position', otheralignh['correct_end'])\n",
    "hybsum += printstat('%', percent(otheralignh['correct_end'], hcount))\n",
    "hybsum += printstat('correct length', otheralign['correct_length'])\n",
    "hybsum += printstat('%', percent(otheralignh['correct_length'], hcount))\n",
    "print(nonhybsum + hybsum)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## For each of the hybrid missed alignments, see what was chosen instead"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Correct sequence: \t TLWVFW \t TLW-VFW\n",
      "Attempted alignments\n",
      "\tTIWVFW\tTIW-VFW\n",
      "\tTIWVFW\tTIW-VFW\n",
      "\tTIWVFW\tTIW-VFW\n",
      "\n",
      "Correct sequence: \t LPQVSL \t LPQ-VSL\n",
      "Attempted alignments\n",
      "\tIPQVSL\tIPQ-VSL\n",
      "\tIPQVSL\tIPQ-VSL\n",
      "\tIPQVSL\tIPQ-VSL\n",
      "\n",
      "Correct sequence: \t LPQVSLL \t LPQ-VSLL\n",
      "Attempted alignments\n",
      "\tIPQVSLL\tIPQ-VSLL\n",
      "\tIPQVSLL\tIPQ-VSLL\n",
      "\tIPQVSLL\tIPQ-VSLL\n",
      "\n",
      "Correct sequence: \t LPQVSLLP \t LPQ-VSLLP\n",
      "Attempted alignments\n",
      "\tIPQVSLLP\tIPQ-VSLLP\n",
      "\tIPQVSLLP\tIPQ-VSLLP\n",
      "\tIPQVSLLP\tIPQ-VSLLP\n",
      "\n",
      "Correct sequence: \t LPQVSLLPS \t LPQ-VSLLPS\n",
      "Attempted alignments\n",
      "\tTITISVSLLPS\tTITIS-VSLLPS\n",
      "\tTTLLVSLLPS\tTTLL-VSLLPS\n",
      "\tTTLLVSLLPS\tTTLL-VSLLPS\n",
      "\n",
      "Correct sequence: \t LPQVSLLPSS \t LPQ-VSLLPSS\n",
      "Attempted alignments\n",
      "\tIPQVSLLPSS\tIPQ-VSLLPSS\n",
      "\tIPQVSLLPSS\tIPQ-VSLLPSS\n",
      "\tIPQVSLLPSS\tIPQ-VSLLPSS\n",
      "\n",
      "Correct sequence: \t LPQVSLLPSSQ \t LPQ-VSLLPSSQ\n",
      "Attempted alignments\n",
      "\tIPQVSLLPSSQ\tIPQ-VSLLPSSQ\n",
      "\tIPQVSLLPSSQ\tIPQ-VSLLPSSQ\n",
      "\tIPQVSLLPSSQ\tIPQ-VSLLPSSQ\n",
      "\n",
      "Correct sequence: \t LPQVSLLPSSQT \t LPQ-VSLLPSSQT\n",
      "Attempted alignments\n",
      "\tIPQVSLLPSSQT\tIPQ-VSLLPSSQT\n",
      "\tIPQVSLLPSSQT\tIPQ-VSLLPSSQT\n",
      "\tIPQVSLLPSSQT\tIPQ-VSLLPSSQT\n",
      "\n",
      "Correct sequence: \t AMVHLPQVSLLPS \t AMVHLPQ-VSLLPS\n",
      "Attempted alignments\n",
      "\tTITISFNVSLLPS\tTITIS-FNVSLLPS\n",
      "\tAMVHLPQFNVSLLPS\tAMVHLPQ-FNVSLLPS\n",
      "\tAMVHLPQSILPS\tAMVHLPQ-SILPS\n",
      "\n",
      "Correct sequence: \t AMVHLPQVSLLPSS \t AMVHLPQ-VSLLPSS\n",
      "Attempted alignments\n",
      "\tAMVHLPQFNVSLLPSS\tAMVHLPQ-FNVSLLPSS\n",
      "\tAMVHLPQSILPSS\tAMVHLPQ-SILPSS\n",
      "\tAMVCFNVSLLPSS\tAMVC-FNVSLLPSS\n",
      "\n",
      "Correct sequence: \t LAMVHLPQVSLLPSSQ \t LAMVHLPQ-VSLLPSSQ\n",
      "Attempted alignments\n",
      "\tLAMVHLPQGGGVSLLPSSQ\tLAMVHLPQGGG-VSLLPSSQ\n",
      "\tLAMVHLPQGGGPSSQ\tLAMVHLPQGGG-PSSQ\n",
      "\tLAMVVSLLPSSQ\tLAMV-VSLLPSSQ\n",
      "\n",
      "Correct sequence: \t LLGSSS \t LLG-SSS\n",
      "Attempted alignments\n",
      "\tLLGSSGSSS\tLLGSS-GSSS\n",
      "\tLLGSSGSSS\tLLGSS-GSSS\n",
      "\tLLGSSGSSS\tLLGSS-GSSS\n",
      "\n",
      "Correct sequence: \t LLGSSSGRF \t LLG-SSSGRF\n",
      "Attempted alignments\n",
      "\tLLGSSSSSGRF\tLLGSS-SSSGRF\n",
      "\tLLGSSSSGRF\tLLGS-SSSGRF\n",
      "\tLLGSSSSGRF\tLLGS-SSSGRF\n",
      "\n",
      "Correct sequence: \t LLGSSSGRFV \t LLG-SSSGRFV\n",
      "Attempted alignments\n",
      "\tLLGSSSSSGRFV\tLLGSS-SSSGRFV\n",
      "\tLLGSSSSGRFV\tLLGS-SSSGRFV\n",
      "\tLLGSSSSGRFV\tLLGS-SSSGRFV\n",
      "\n",
      "Correct sequence: \t LLGSSSGRFVR \t LLG-SSSGRFVR\n",
      "Attempted alignments\n",
      "\tLLGSSSSSGRFVR\tLLGSS-SSSGRFVR\n",
      "\tLLGSSSSGRFVR\tLLGS-SSSGRFVR\n",
      "\tLLGSSSSGRFVR\tLLGS-SSSGRFVR\n",
      "\n",
      "Correct sequence: \t LLGSSSGRFVRL \t LLG-SSSGRFVRL\n",
      "Attempted alignments\n",
      "\tLLGSSSSSGRFVRL\tLLGSS-SSSGRFVRL\n",
      "\tLLGSSSSGRFVRL\tLLGS-SSSGRFVRL\n",
      "\tLLGSSSSGRFVRL\tLLGS-SSSGRFVRL\n",
      "\n",
      "Correct sequence: \t NSMIDRS \t NSM-IDRS\n",
      "Attempted alignments\n",
      "\tNSMLDRS\tNSML-DRS\n",
      "\tNSMLDRS\tNSML-DRS\n",
      "\tNSMLDRS\tNSM-LDRS\n",
      "\n",
      "Correct sequence: \t SNSMIDRSNSM \t SNSM-IDRSNSM\n",
      "Attempted alignments\n",
      "\tSNSNIDRSNSM\tSNSN-IDRSNSM\n",
      "\tSNSIDRSNSM\tSNS-IDRSNSM\n",
      "\tSNSIDRSNSM\tSNS-IDRSNSM\n",
      "\n",
      "Correct sequence: \t SNSMIDRSNSMS \t SNSM-IDRSNSMS\n",
      "Attempted alignments\n",
      "\tSNSNIDRSNSMS\tSNSN-IDRSNSMS\n",
      "\tSNSIDRSNSMS\tSNS-IDRSNSMS\n",
      "\tSNSIDRSNSMS\tSNS-IDRSNSMS\n",
      "\n",
      "Correct sequence: \t SNSMIDRSNSMSK \t SNSM-IDRSNSMSK\n",
      "Attempted alignments\n",
      "\tSNSNIDRSNSMSK\tSNSN-IDRSNSMSK\n",
      "\tSNSIDRSNSMSK\tSNS-IDRSNSMSK\n",
      "\tSNSIDRSNSMSK\tSNS-IDRSNSMSK\n",
      "\n",
      "Correct sequence: \t RSNSMIDRSNSM \t RSNSM-IDRSNSM\n",
      "Attempted alignments\n",
      "\tRSNSIDRSNSM\tRSNS-IDRSNSM\n",
      "\tRSNSIDRSNSM\tRSNS-IDRSNSM\n",
      "\tRSNSIDRSNSM\tRSNS-IDRSNSM\n",
      "\n",
      "Correct sequence: \t RSNSMIDRSNSMS \t RSNSM-IDRSNSMS\n",
      "Attempted alignments\n",
      "\tRSNSIDRSNSMS\tRSNS-IDRSNSMS\n",
      "\tRSNSIDRSNSMS\tRSNS-IDRSNSMS\n",
      "\tRSNSIDRSNSMS\tRSNS-IDRSNSMS\n",
      "\n",
      "Correct sequence: \t RSNSMIDRSNSMSK \t RSNSM-IDRSNSMSK\n",
      "Attempted alignments\n",
      "\tRSNSIDRSNSMSK\tRSNS-IDRSNSMSK\n",
      "\tRSNSIDRSNSMSK\tRSNS-IDRSNSMSK\n",
      "\tRSNSIDRSNSMSK\tRSNS-IDRSNSMSK\n",
      "\n",
      "Correct sequence: \t DRSNSMIDRSNSM \t DRSNSM-IDRSNSM\n",
      "Attempted alignments\n",
      "\tDRSNSM\tDRS-NSM\n",
      "\tDRSNSM\tDRS-NSM\n",
      "\tDRSNSM\tDRS-NSM\n",
      "\n",
      "Correct sequence: \t DRSNSMIDRSNSMS \t DRSNSM-IDRSNSMS\n",
      "Attempted alignments\n",
      "\tDRSIDRSNSMS\tDRS-IDRSNSMS\n",
      "\tDRSIDRSNSMS\tDRS-IDRSNSMS\n",
      "\tDRSIDRSNSMS\tDRS-IDRSNSMS\n",
      "\n",
      "Correct sequence: \t DRSNSMIDRSNSMSK \t DRSNSM-IDRSNSMSK\n",
      "Attempted alignments\n",
      "\tDRSIDRSNSMSK\tDRS-IDRSNSMSK\n",
      "\tDRSIDRSNSMSK\tDRS-IDRSNSMSK\n",
      "\tDRSIDRSNSMSK\tDRS-IDRSNSMSK\n",
      "\n",
      "Correct sequence: \t IDRSNSMIDRS \t IDRSNSM-IDRS\n",
      "Attempted alignments\n",
      "\tIDRSNSMLDRS\tIDRSNSM-LDRS\n",
      "\tIDRSNSMLDRS\tIDRSNSM-LDRS\n",
      "\tIDRSNSMDRS\tIDRSNSM-DRS\n",
      "\n",
      "Correct sequence: \t IDRSNSMIDRSN \t IDRSNSM-IDRSN\n",
      "Attempted alignments\n",
      "\tIDRSNSMRSN\tIDRSNSM-RSN\n",
      "\tIDRSNSMRSN\tIDRSNSM-RSN\n",
      "\tIDRSNSMRSN\tIDRSNSM-RSN\n",
      "\n",
      "Correct sequence: \t IDRSNSMIDRSNS \t IDRSNSM-IDRSNS\n",
      "Attempted alignments\n",
      "\tIDRSNS\tIDR-SNS\n",
      "\tIDRSNS\tIDR-SNS\n",
      "\tIDRSNSMRSNS\tIDRSNSM-RSNS\n",
      "\n",
      "Correct sequence: \t IDRSNSMIDRSNSM \t IDRSNSM-IDRSNSM\n",
      "Attempted alignments\n",
      "\tIDRSNSM\t\n",
      "\tLDRSNSM\tLDRS-NSM\n",
      "\tLDRSNSM\tLDRS-NSM\n",
      "\n",
      "Correct sequence: \t IDRSNSMIDRSNSMS \t IDRSNSM-IDRSNSMS\n",
      "Attempted alignments\n",
      "\tIDRSNSMS\t\n",
      "\tLDRSIDRSNSMS\tLDRS-IDRSNSMS\n",
      "\tLDRSIDRSNSMS\tLDRS-IDRSNSMS\n",
      "\n",
      "Correct sequence: \t IDRSNSMIDRSNSMSK \t IDRSNSM-IDRSNSMSK\n",
      "Attempted alignments\n",
      "\tIDRSNSMSK\t\n",
      "\tLDRSIDRSNSMSK\tLDRS-IDRSNSMSK\n",
      "\tLDRSIDRSNSMSK\tLDRS-IDRSNSMSK\n",
      "\n",
      "Correct sequence: \t LIDRSNSMIDRS \t LIDRSNSM-IDRS\n",
      "Attempted alignments\n",
      "\tLIDRSNSMLDRS\tLIDRSNSM-LDRS\n",
      "\tLIDRSNSMLDRS\tLIDRSNSM-LDRS\n",
      "\tLIDRSNSMDRS\tLIDRSNSM-DRS\n",
      "\n",
      "Correct sequence: \t LIDRSNSMIDRSN \t LIDRSNSM-IDRSN\n",
      "Attempted alignments\n",
      "\tLIDRSNSMRSN\tLIDRSNSM-RSN\n",
      "\tLIDRSNSMRSN\tLIDRSNSM-RSN\n",
      "\tLIDRSNSMRSN\tLIDRSNSM-RSN\n",
      "\n",
      "Correct sequence: \t LIDRSNSMIDRSNS \t LIDRSNSM-IDRSNS\n",
      "Attempted alignments\n",
      "\tLLDRSNS\tLLDR-SNS\n",
      "\tLLDRSNS\tLLDR-SNS\n",
      "\tLIDRSNS\tLIDR-SNS\n",
      "\n",
      "Correct sequence: \t LIDRSNSMIDRSNSM \t LIDRSNSM-IDRSNSM\n",
      "Attempted alignments\n",
      "\tLIDRSNSM\t\n",
      "\tLIDRSNSMNSM\tLIDRSNSM-NSM\n",
      "\tLIDRSNSMNSM\tLIDRSNSM-NSM\n",
      "\n",
      "Correct sequence: \t LIDRSNSMIDRSNSMS \t LIDRSNSM-IDRSNSMS\n",
      "Attempted alignments\n",
      "\tLIDRSNSMS\t\n",
      "\tLLDRIDRSNSMS\tLLDR-IDRSNSMS\n",
      "\tLLDRIDRSNSMS\tLLDR-IDRSNSMS\n",
      "\n",
      "Correct sequence: \t LIDRSNSMIDRSNSMSK \t LIDRSNSM-IDRSNSMSK\n",
      "Attempted alignments\n",
      "\tLIDRSNSMSK\t\n",
      "\tLLDRIDRSNSMSK\tLLDR-IDRSNSMSK\n",
      "\tLLDRIDRSNSMSK\tLLDR-IDRSNSMSK\n",
      "\n",
      "Correct sequence: \t FLIDRSNSMIDRS \t FLIDRSNSM-IDRS\n",
      "Attempted alignments\n",
      "\tFLIDRSNSMLDRS\tFLIDRSNSM-LDRS\n",
      "\tFLIDRSNSMLDRS\tFLIDRSNSM-LDRS\n",
      "\tFLIDRSNSMRDLKN\tFLIDRSNSM-RDLKN\n",
      "\n",
      "Correct sequence: \t FLIDRSNSMIDRSN \t FLIDRSNSM-IDRSN\n",
      "Attempted alignments\n",
      "\tFLIDRSNSMDLKN\tFLIDRSNSM-DLKN\n",
      "\tFLIDRSNSMRDLKN\tFLIDRSNSM-RDLKN\n",
      "\tFLIDRSNSMIFLIDRSN\tFLIDRSNSM-IFLIDRSN\n",
      "\n",
      "Correct sequence: \t FLIDRSNSMIDRSNS \t FLIDRSNSM-IDRSNS\n",
      "Attempted alignments\n",
      "\tFLLDRSNS\tFLLD-RSNS\n",
      "\tFLLDRSNS\tFLLD-RSNS\n",
      "\tFLLDRSNS\tFLLD-RSNS\n",
      "\n",
      "Correct sequence: \t FLIDRSNSMIDRSNSM \t FLIDRSNSM-IDRSNSM\n",
      "Attempted alignments\n",
      "\tFLIDRSNSM\t\n",
      "\tFLIDRSNSMRDLKN\tFLIDRSNSM-RDLKN\n",
      "\tFLLDRIDRSNSM\tFLLDR-IDRSNSM\n",
      "\n",
      "Correct sequence: \t FLIDRSNSMIDRSNSMS \t FLIDRSNSM-IDRSNSMS\n",
      "Attempted alignments\n",
      "\tFLIDRSNSMS\t\n",
      "\tFLLDRIDRSNSMS\tFLLDR-IDRSNSMS\n",
      "\tFLIDRSNSMRDLKN\tFLIDRSNSM-RDLKN\n",
      "\n",
      "Correct sequence: \t FLIDRSNSMIDRSNSMSK \t FLIDRSNSM-IDRSNSMSK\n",
      "Attempted alignments\n",
      "\tFLIDRSNSMSK\t\n",
      "\tFLLDRIDRSNSMSK\tFLLDR-IDRSNSMSK\n",
      "\tFLIDRSNSMDINK\tFLIDRSNSM-DINK\n"
     ]
    }
   ],
   "source": [
    "for i, bad in enumerate(wrong_hybrid_alignemnts):\n",
    "    print('\\nCorrect sequence: \\t {} \\t {}'.format(bad[1]['sequence'], bad[1]['hybrid_sequence']))\n",
    "    print('Attempted alignments')\n",
    "    for b in bad[0]:\n",
    "        print('\\t{}\\t{}'.format(b['sequence'], b['hybrid_sequence']))\n",
    "\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "119\n"
     ]
    }
   ],
   "source": [
    "print(len(wrong_hybrid_alignemnts))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
