{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Run a test of hypedsearch with generated data\n",
    "The following steps describe how the test works\n",
    "1. Load a fasta database\n",
    "2. Generate\n",
    "    1. Hybrid proteins\n",
    "    2. Peptides\n",
    "    3. Hybrid peptides from the hybrid proteins\n",
    "3. Generate spectra for all the peptides created\n",
    "4. Run hypedsearch with the .fasta file (no hybrid proteins included) and the spectra files\n",
    "5. Load the summary.json file created\n",
    "6. Determine what number of alignments were correct"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 1. Load fasta database"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import sys\n",
    "module_path = os.path.abspath(os.path.join('..'))\n",
    "if module_path not in sys.path:\n",
    "    sys.path.append(module_path)\n",
    "module_path = os.path.abspath(os.path.join('../..'))\n",
    "if module_path not in sys.path:\n",
    "    sys.path.append(module_path)\n",
    "    \n",
    "from src.file_io import fasta\n",
    "\n",
    "fasta_file = '../data/databases/100prots.fasta'\n",
    "database = fasta.read(fasta_file, True)\n",
    "\n",
    "database = {x['name']: x for x in database}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 2.  Generate the peptides, hybrid proteins and peptides"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Generating hybrid protein 0/5[0%]\r",
      "Generating hybrid protein 1/5[20%]\r",
      "Generating hybrid protein 2/5[40%]\r",
      "Generating hybrid protein 3/5[60%]\r",
      "Generating hybrid protein 4/5[80%]\r\n",
      "Finished generating hybrid proteins\n"
     ]
    }
   ],
   "source": [
    "from sequence_generation import proteins, peptides\n",
    "test_directory = '../data/testing_output/'\n",
    "\n",
    "num_hybs = 5\n",
    "min_length= 5\n",
    "max_length = 35\n",
    "num_peptides = 100\n",
    "min_cont = 3 #min contribution for each side of a hybrid\n",
    "\n",
    "# make hybrid proteins\n",
    "hyb_prots = proteins.generate_hybrids([x for _, x in database.items()], num_hybs, min_contribution=max_length)\n",
    "# create peptides\n",
    "non_hybrid_peps = peptides.gen_peptides([x for _, x in database.items()], num_peptides, min_length=min_length, max_length=max_length, digest='random', dist='beta')\n",
    "# create hybrid peptides\n",
    "hyb_peps = peptides.gen_peptides(hyb_prots, num_hybs, min_length=min_length, max_length=max_length, digest='random', min_contribution=min_cont, hybrid_list=True)\n",
    "\n",
    "all_proteins_raw = [x for _,x in database.items()] + hyb_prots\n",
    "all_peptides_raw = non_hybrid_peps + hyb_peps\n",
    "\n",
    "peptides = {}\n",
    "for i, pep in enumerate(all_peptides_raw):\n",
    "    peptides[i] = pep\n",
    "    peptides[i]['scan_no'] = i"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 2.1 Save this info so that I can analyze it later from Neo-Fusion"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "import json\n",
    "experiment_info_file_name = 'experiment_info.json'\n",
    "\n",
    "exp = {'database': fasta_file, 'peptides': peptides}\n",
    "with open(test_directory + experiment_info_file_name, 'w') as o:\n",
    "    json.dump(exp, o)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 2.2 Load data if available instead of creating it"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "# import json\n",
    "\n",
    "# expfile = '../data/testing_output/experiment_info.json'\n",
    "# exp = json.load(open(expfile, 'r'))\n",
    "# peptides = exp['peptides']"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 3. Generate spectra"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Determination of memory status is not supported on this \n",
      " platform, measuring for memoryleaks will never fail\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "'../data/testing_output/testSpectraFile.mzML'"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from src.spectra import gen_spectra\n",
    "from src.utils import utils\n",
    "from sequence_generation import write_spectra\n",
    "\n",
    "utils.make_dir(test_directory)\n",
    "\n",
    "spectra = []\n",
    "sorted_keys = [int(c) for c in peptides.keys()]\n",
    "sorted_keys.sort()\n",
    "for k in sorted_keys:\n",
    "    pep = peptides[k]\n",
    "    cont = gen_spectra.gen_spectrum(pep['sequence'])\n",
    "    spec = cont['spectrum']\n",
    "    pm = cont['precursor_mass']\n",
    "    spectra.append({'spectrum': spec, 'precursor_mass': pm})\n",
    "write_spectra.write_mzml('testSpectraFile', spectra, output_dir=test_directory)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 4. Run hypedsearch"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Loading database...\n",
      "Adding protein 42/100 to tree\r"
     ]
    },
    {
     "ename": "KeyboardInterrupt",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m                         Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-7-79db44868069>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m     22\u001b[0m }\n\u001b[1;32m     23\u001b[0m \u001b[0mst\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mtime\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 24\u001b[0;31m \u001b[0mrunner\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mrun\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0margs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     25\u001b[0m \u001b[0mprint\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'\\nTotal runtime: {} seconds'\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mformat\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mtime\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;34m-\u001b[0m \u001b[0mst\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     26\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/Documents/Layer_Research/hypedsearch/src/runner.py\u001b[0m in \u001b[0;36mrun\u001b[0;34m(args)\u001b[0m\n\u001b[1;32m     35\u001b[0m         \u001b[0;32mbreak\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     36\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 37\u001b[0;31m     \u001b[0mmatched_spectra\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mid_spectra\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mid_spectra\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mspectra_files\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0margs\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'database_file'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mmin_peptide_len\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0margs\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'min_peptide_len'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mmax_peptide_len\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0margs\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'max_peptide_len'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     38\u001b[0m     \u001b[0mprint\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'\\nFinished search. Writting results to {}...'\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mformat\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0margs\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'output_dir'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     39\u001b[0m     \u001b[0msummary\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mgenerate\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mmatched_spectra\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0margs\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'output_dir'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/Documents/Layer_Research/hypedsearch/src/identfication/id_spectra.py\u001b[0m in \u001b[0;36mid_spectra\u001b[0;34m(spectra_files, database_file, verbose, min_peptide_len, max_peptide_len, ppm_tolerance)\u001b[0m\n\u001b[1;32m    108\u001b[0m     '''\n\u001b[1;32m    109\u001b[0m     \u001b[0mverbose\u001b[0m \u001b[0;32mand\u001b[0m \u001b[0mprint\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'Loading database...'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 110\u001b[0;31m     \u001b[0mdatabase\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mDatabase\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mdatabase_file\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mis_uniprot\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;32mTrue\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mkmer_size\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mmin_peptide_len\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    111\u001b[0m     \u001b[0mverbose\u001b[0m \u001b[0;32mand\u001b[0m \u001b[0mprint\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'Done. Indexing database...'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    112\u001b[0m     \u001b[0mbase_k_mer_masses\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mmake_all_base_mers\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mdatabase\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mmin_peptide_len\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/Documents/Layer_Research/hypedsearch/src/types/database.py\u001b[0m in \u001b[0;36m__init__\u001b[0;34m(self, fasta_file_name, is_uniprot, kmer_size, verbose)\u001b[0m\n\u001b[1;32m     69\u001b[0m         \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mmetadata\u001b[0m\u001b[0;34m:\u001b[0m \u001b[0mdict\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;32mNone\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     70\u001b[0m         \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mverbose\u001b[0m\u001b[0;34m:\u001b[0m \u001b[0mbool\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mverbose\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 71\u001b[0;31m         \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mtree\u001b[0m\u001b[0;34m:\u001b[0m \u001b[0mTree\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m__build_tree\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     72\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     73\u001b[0m     \u001b[0;31m##################### Private Methods #####################\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/Documents/Layer_Research/hypedsearch/src/types/database.py\u001b[0m in \u001b[0;36m__build_tree\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m     89\u001b[0m             \u001b[0mprintskipc\u001b[0m \u001b[0;34m+=\u001b[0m \u001b[0;36m1\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     90\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 91\u001b[0;31m             \u001b[0mt\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0madd\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mkey\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mvalue\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0msequence\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     92\u001b[0m         \u001b[0;32mreturn\u001b[0m \u001b[0mt\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     93\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/site-packages/suffix_tree/tree.py\u001b[0m in \u001b[0;36madd\u001b[0;34m(self, id_, S, builder)\u001b[0m\n\u001b[1;32m     92\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     93\u001b[0m         \u001b[0;32mtry\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 94\u001b[0;31m             \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mbuilder\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mbuild\u001b[0m \u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     95\u001b[0m         \u001b[0;32mexcept\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     96\u001b[0m             \u001b[0;32mif\u001b[0m \u001b[0mis_debug\u001b[0m \u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/site-packages/suffix_tree/ukkonen.py\u001b[0m in \u001b[0;36mbuild\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m    227\u001b[0m             \u001b[0mPath\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0me\u001b[0m  \u001b[0;34m=\u001b[0m \u001b[0mpath\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mend\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    228\u001b[0m             \u001b[0ms\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mpath\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mupdate\u001b[0m   \u001b[0;34m(\u001b[0m\u001b[0ms\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mpath\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 229\u001b[0;31m             \u001b[0ms\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mpath\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mcanonize\u001b[0m \u001b[0;34m(\u001b[0m\u001b[0ms\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mpath\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    230\u001b[0m             \u001b[0mdebug\u001b[0m \u001b[0;34m(\u001b[0m\u001b[0;34m'active point is: s=\"%s\" %s'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0ms\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mpath\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mukko_str\u001b[0m \u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    231\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/site-packages/suffix_tree/ukkonen.py\u001b[0m in \u001b[0;36mcanonize\u001b[0;34m(self, s, path)\u001b[0m\n\u001b[1;32m    129\u001b[0m         \u001b[0;31m# follow transitions starting from s until we get too far down the tree,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    130\u001b[0m         \u001b[0;31m# then return the transition before the one that went too far\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 131\u001b[0;31m         \u001b[0ms_prime\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mpath_prime\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mtransition\u001b[0m \u001b[0;34m(\u001b[0m\u001b[0ms\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mpath\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mk\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    132\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    133\u001b[0m         \u001b[0;32mwhile\u001b[0m \u001b[0mlen\u001b[0m \u001b[0;34m(\u001b[0m\u001b[0mpath_prime\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;34m<=\u001b[0m \u001b[0mlen\u001b[0m \u001b[0;34m(\u001b[0m\u001b[0mpath\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/site-packages/suffix_tree/ukkonen.py\u001b[0m in \u001b[0;36mtransition\u001b[0;34m(self, s, k)\u001b[0m\n\u001b[1;32m     55\u001b[0m         \u001b[0;32massert\u001b[0m \u001b[0mk\u001b[0m \u001b[0;34m>=\u001b[0m \u001b[0;36m0\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     56\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 57\u001b[0;31m         \u001b[0;32mif\u001b[0m \u001b[0ms\u001b[0m \u001b[0;32mis\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0maux\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     58\u001b[0m             \u001b[0;31m# simulates line 2. from *Algorithm 2*.\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     59\u001b[0m             \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mroot\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mPath\u001b[0m \u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mpath\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mS\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;36m0\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;36m1\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m: "
     ]
    }
   ],
   "source": [
    "import os\n",
    "import sys\n",
    "module_path = os.path.abspath(os.path.join('..'))\n",
    "if module_path not in sys.path:\n",
    "    sys.path.append(module_path)\n",
    "module_path = os.path.abspath(os.path.join('../..'))\n",
    "if module_path not in sys.path:\n",
    "    sys.path.append(module_path)\n",
    "\n",
    "from src import runner\n",
    "from time import time\n",
    "\n",
    "test_directory = '../data/testing_output/'\n",
    "fasta_file = '../data/databases/100prots.fasta'\n",
    "\n",
    "args = {\n",
    "    'spectra_folder': test_directory,\n",
    "    'database_file': fasta_file,\n",
    "    'output_dir': test_directory,\n",
    "    'min_peptide_len': 3,\n",
    "    'max_peptide_len': 35,\n",
    "}\n",
    "st = time()\n",
    "runner.run(args)\n",
    "print('\\nTotal runtime: {} seconds'.format(time() - st))\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 5. Load the summary json"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "../data/testing_output/\n"
     ]
    }
   ],
   "source": [
    "import json\n",
    "test_directory = '../data/testing_output/'\n",
    "\n",
    "summary = json.load(open(test_directory + 'summary.json', 'r'))\n",
    "print(test_directory)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### summary format\n",
    "Each entry is of the form\n",
    "```python\n",
    "{\n",
    "    '<file-name>_<scan_number>':{\n",
    "        'spectrum': {...},\n",
    "        'alignments': [{...}],\n",
    "        'b_scores': [{...}],\n",
    "        'y_score': [{...}]\n",
    "    }\n",
    "}\n",
    "```\n",
    "The only attribute we're really interested in is the alignments attribute which has the form\n",
    "```python\n",
    "{\n",
    "    'alignments': [{\n",
    "        'b_alignment': {...},\n",
    "        'y_alignment': {...}, \n",
    "        'spectrum': [...],\n",
    "        'protein': str,\n",
    "        'alignment_score': float,\n",
    "        'sequence': str,\n",
    "        'hybrid': bool,\n",
    "        'hybrid_sequence': str\n",
    "    }, ...\n",
    "    ]\n",
    "}\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 6. Determine which number of alignments were correct\n",
    "This needs to be broken down into hybrid and non hybrid peptides to get some stats on how well its doing"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "n = 5\n",
    "non_hyb_stats = {i: {\n",
    "        'correct': 0,\n",
    "        'correct_parent': 0,\n",
    "        'correct_length': 0,\n",
    "        'correct_start': 0,\n",
    "        'correct_end': 0\n",
    "    } for i in range(n)}\n",
    "non_hyb_stats['count'] = 0\n",
    "\n",
    "hyb_stats ={i: {\n",
    "        'left_correct_parent': 0,\n",
    "        'right_correct_parent': 0, \n",
    "        'correct_start': 0,\n",
    "        'correct_end': 0,\n",
    "        'correct_length': 0,\n",
    "        'correct_sequence': 0,\n",
    "        'correct': 0\n",
    "    } for i in range(n)}\n",
    "hyb_stats['count'] = 0\n",
    "\n",
    "wrong_hybrid_alignemnts = []\n",
    "wrong_nonhybrid_alignments = []"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "def hyb_calc(result, real_pep):\n",
    "    hyb_stats['count'] += 1\n",
    "    is_correct = False\n",
    "    for i in range(min(n, len(result))):\n",
    "        res = result[i]\n",
    "        result_hybrid = bool(res['hybrid'])\n",
    "        if not result_hybrid: \n",
    "            continue\n",
    "        \n",
    "        left_corrparent = real_pep['left_parent_name'] == res['protein'].split('~')[0]\n",
    "        right_corrparent = real_pep['right_parent_name'] == res['protein'].split('~')[1]\n",
    "        corr_start = real_pep['left_parent_starting_position'] == res['b_alignment']['kmer']['start_position']\n",
    "        corr_end = real_pep['right_parent_ending_position'] == res['y_alignment']['kmer']['end_position']\n",
    "        corr_len = len(real_pep['sequence']) == len(res['sequence'])\n",
    "        corr_seq = real_pep['sequence'] == res['sequence']\n",
    "        \n",
    "        if not corr_seq: \n",
    "            print('real pep seq: {} \\t result seq: {}'.format(real_pep['sequence'], res['sequence']))\n",
    "            continue\n",
    "        hyb_stats[i]['left_correct_parent'] += 1 if left_corrparent else 0\n",
    "        hyb_stats[i]['right_correct_parent'] += 1 if right_corrparent else 0\n",
    "        hyb_stats[i]['correct_start'] += 1 if corr_start else 0\n",
    "        hyb_stats[i]['correct_end'] += 1 if corr_end else 0\n",
    "        hyb_stats[i]['correct_length'] += 1 if corr_len else 0\n",
    "        hyb_stats[i]['correct_sequence'] += 1 if corr_seq else 0\n",
    "        hyb_stats[i]['correct'] += 1 if left_corrparent and right_corrparent and corr_start and corr_end and corr_len else 0\n",
    "        \n",
    "        is_correct = corr_seq\n",
    "            \n",
    "    if not is_correct:\n",
    "        wrong_hybrid_alignemnts.append((result, real_pep))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "def non_hyb_calc(result, real_pep):\n",
    "    non_hyb_stats['count'] += 1\n",
    "    iterrange = min(n, len(result))\n",
    "    for i in range(iterrange):\n",
    "        if result[i]['protein'] != real_pep['parent_name']:\n",
    "            continue\n",
    "        corrlen = len(result[i]['sequence']) == len(real_pep['sequence'])\n",
    "        resstartpos = min(result[i]['b_alignment']['kmer']['start_position'], result[i]['y_alignment']['kmer']['start_position'])\n",
    "        resendpos = min(result[i]['b_alignment']['kmer']['end_position'], result[i]['y_alignment']['kmer']['end_position'])\n",
    "        corrstart = resstartpos == real_pep['starting_position']\n",
    "        corrend = resendpos == real_pep['ending_position']\n",
    "        \n",
    "        non_hyb_stats[i]['correct_parent'] += 1 \n",
    "        non_hyb_stats[i]['correct_length'] += 1 if corrlen else 0\n",
    "        non_hyb_stats[i]['correct_start'] += 1 if corrstart else 0\n",
    "        non_hyb_stats[i]['correct_end'] += 1 if corrend else 0\n",
    "        non_hyb_stats[i]['correct'] += 1 if (corrend and corrlen and corrstart) else 0\n",
    "        \n",
    "        if i != 0 and not (corrend and corrlen and corrstart):\n",
    "            wrong_nonhybrid_alignments.append((result, real_pep))\n",
    "        return\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "real pep seq: MDKKHKEE \t result seq: MDKKDKHKEE\n",
      "real pep seq: MDKKHKEE \t result seq: MDKKDKHKEE\n",
      "real pep seq: MDKKHKEE \t result seq: MDKKDKHKEE\n",
      "real pep seq: MDKKHKEED \t result seq: MMEKDKHKEED\n",
      "real pep seq: MDKKHKEED \t result seq: MDKKDKHKEED\n",
      "real pep seq: MDKKHKEED \t result seq: MDKKDKHKEED\n",
      "real pep seq: MDKKHKEEDK \t result seq: MDKKDKHKEEDK\n",
      "real pep seq: MDKKHKEEDK \t result seq: MDKKDKHKEEDK\n",
      "real pep seq: MDKKHKEEDK \t result seq: MDKKDKHKEEDK\n",
      "real pep seq: MDKKHKEEDKG \t result seq: MDKKDKHKEEDKG\n",
      "real pep seq: MDKKHKEEDKG \t result seq: MDKKDKHKEEDKG\n",
      "real pep seq: MDKKHKEEDKG \t result seq: MDKKDKHKEEDKG\n",
      "real pep seq: MDKKHKEEDKGS \t result seq: MDKKDKHKEEDKGS\n",
      "real pep seq: MDKKHKEEDKGS \t result seq: MDKKDKHKEEDKGS\n",
      "real pep seq: MDKKHKEEDKGS \t result seq: MDKKDKHKEEDKGS\n",
      "real pep seq: KMDKKHKE \t result seq: KMKDKHKE\n",
      "real pep seq: KMDKKHKE \t result seq: KMDKKDKHKE\n",
      "real pep seq: KMDKKHKEE \t result seq: KMKDKHKEE\n",
      "real pep seq: KMDKKHKEE \t result seq: KMDKKDKHKEE\n",
      "real pep seq: KMDKKHKEE \t result seq: KMDKDKHKEE\n",
      "real pep seq: KMDKKHKEED \t result seq: KMDKKDKHKEED\n",
      "real pep seq: KMDKKHKEED \t result seq: KMDKDKHKEED\n",
      "real pep seq: KMDKKHKEED \t result seq: KMDKDKHKEED\n",
      "real pep seq: KMDKKHKEEDK \t result seq: KMDKKDKHKEEDK\n",
      "real pep seq: KMDKKHKEEDK \t result seq: KMMKDKHKEEDK\n",
      "real pep seq: KMDKKHKEEDK \t result seq: KMDKDKHKEEDK\n",
      "real pep seq: KMDKKHKEEDKG \t result seq: KMKDKHKEEDKG\n",
      "real pep seq: KMDKKHKEEDKG \t result seq: KMDKKDKHKEEDKG\n",
      "real pep seq: KMDKKHKEEDKG \t result seq: KMDKDKHKEEDKG\n",
      "real pep seq: KMDKKHKEEDKGS \t result seq: KMKDKHKEEDKGS\n",
      "real pep seq: KMDKKHKEEDKGS \t result seq: KMDKKDKHKEEDKGS\n",
      "real pep seq: KMDKKHKEEDKGS \t result seq: KMDKDKHKEEDKGS\n",
      "real pep seq: GKMDKKHKE \t result seq: GKMKDKHKE\n",
      "real pep seq: GKMDKKHKE \t result seq: GKMKDKHKE\n",
      "real pep seq: GKMDKKHKEE \t result seq: GKMKDKHKEE\n",
      "real pep seq: GKMDKKHKEE \t result seq: GKMKDKHKEE\n",
      "real pep seq: GKMDKKHKEE \t result seq: GKMKDKHKEE\n",
      "real pep seq: GKMDKKHKEED \t result seq: GKMKDKHKEED\n",
      "real pep seq: GKMDKKHKEED \t result seq: GKMKDKHKEED\n",
      "real pep seq: GKMDKKHKEED \t result seq: GKMKDKHKEED\n",
      "real pep seq: GKMDKKHKEEDK \t result seq: GKMKDKHKEEDK\n",
      "real pep seq: GKMDKKHKEEDK \t result seq: GKMKDKHKEEDK\n",
      "real pep seq: GKMDKKHKEEDK \t result seq: GKMKDKHKEEDK\n",
      "real pep seq: GKMDKKHKEEDKG \t result seq: GKMKDKHKEEDKG\n",
      "real pep seq: GKMDKKHKEEDKG \t result seq: GKMKDKHKEEDKG\n",
      "real pep seq: GKMDKKHKEEDKG \t result seq: GKMKDKHKEEDKG\n",
      "real pep seq: GKMDKKHKEEDKGS \t result seq: GKMKDKHKEEDKGS\n",
      "real pep seq: GKMDKKHKEEDKGS \t result seq: GKMKDKHKEEDKGS\n",
      "real pep seq: GKMDKKHKEEDKGS \t result seq: GKMKDKHKEEDKGS\n",
      "real pep seq: DGKMDKKHKE \t result seq: DGKMDKKDKHKE\n",
      "real pep seq: DGKMDKKHKE \t result seq: DGKMDKHKE\n",
      "real pep seq: DGKMDKKHKEE \t result seq: DGKMDKKDKHKEE\n",
      "real pep seq: DGKMDKKHKEE \t result seq: DGKMDKKEE\n",
      "real pep seq: DGKMDKKHKEE \t result seq: DGKMDKKEE\n",
      "real pep seq: DGKMDKKHKEED \t result seq: DGKMDKKDKHKEED\n",
      "real pep seq: DGKMDKKHKEED \t result seq: DGKMDKKEED\n",
      "real pep seq: DGKMDKKHKEED \t result seq: DGKMDKEED\n",
      "real pep seq: DGKMDKKHKEEDK \t result seq: DGKMDKKDKHKEEDK\n",
      "real pep seq: DGKMDKKHKEEDK \t result seq: DGKMDKEEDK\n",
      "real pep seq: DGKMDKKHKEEDK \t result seq: DGKMDKEEDK\n",
      "real pep seq: DGKMDKKHKEEDKG \t result seq: DGKMDKKDKHKEEDKG\n",
      "real pep seq: DGKMDKKHKEEDKG \t result seq: DGKKDKHKEEDKG\n",
      "real pep seq: DGKMDKKHKEEDKG \t result seq: DGKKDKHKEEDKG\n",
      "real pep seq: DGKMDKKHKEEDKGS \t result seq: DGKMDKKDKHKEEDKGS\n",
      "real pep seq: DGKMDKKHKEEDKGS \t result seq: DGKKDKHKEEDKGS\n",
      "real pep seq: DGKMDKKHKEEDKGS \t result seq: DGKKDKHKEEDKGS\n",
      "real pep seq: RDGKMDKKHKE \t result seq: RDGKMDKKDKHKE\n",
      "real pep seq: RDGKMDKKHKE \t result seq: RDGKMDKHKE\n",
      "real pep seq: RDGKMDKKHKEE \t result seq: RDGKMDKKDKHKEE\n",
      "real pep seq: RDGKMDKKHKEE \t result seq: RDGKMDKKEE\n",
      "real pep seq: RDGKMDKKHKEE \t result seq: RDGKMDKKEE\n",
      "real pep seq: RDGKMDKKHKEED \t result seq: RDGKMDKKDKHKEED\n",
      "real pep seq: RDGKMDKKHKEED \t result seq: RDGKMDKKEED\n",
      "real pep seq: RDGKMDKKHKEED \t result seq: RDGKMDKEED\n",
      "real pep seq: RDGKMDKKHKEEDK \t result seq: RDGKMDKKDKHKEEDK\n",
      "real pep seq: RDGKMDKKHKEEDK \t result seq: RDGKMDKEEDK\n",
      "real pep seq: RDGKMDKKHKEEDK \t result seq: RDGKMDKEEDK\n",
      "real pep seq: RDGKMDKKHKEEDKG \t result seq: RDGKMDKKDKHKEEDKG\n",
      "real pep seq: RDGKMDKKHKEEDKG \t result seq: RDGKKDKHKEEDKG\n",
      "real pep seq: RDGKMDKKHKEEDKG \t result seq: RDGKDKHKEEDKG\n",
      "real pep seq: RDGKMDKKHKEEDKGS \t result seq: RDGKMDKKDKHKEEDKGS\n",
      "real pep seq: RDGKMDKKHKEEDKGS \t result seq: RDGKKDKHKEEDKGS\n",
      "real pep seq: RDGKMDKKHKEEDKGS \t result seq: RDGKDKHKEEDKGS\n",
      "real pep seq: NRDGKMDKKHKE \t result seq: NRDGKMDKKDKHKE\n",
      "real pep seq: NRDGKMDKKHKE \t result seq: NRDGKMDKHKE\n",
      "real pep seq: NRDGKMDKKHKEE \t result seq: NRDGKMDKKDKHKEE\n",
      "real pep seq: NRDGKMDKKHKEE \t result seq: NRDGKMDKKEE\n",
      "real pep seq: NRDGKMDKKHKEE \t result seq: NRDGKMDKKEE\n",
      "real pep seq: NRDGKMDKKHKEED \t result seq: NRDGKMDKKDKHKEED\n",
      "real pep seq: NRDGKMDKKHKEED \t result seq: NRDGKMDKKEED\n",
      "real pep seq: NRDGKMDKKHKEED \t result seq: NRDGKMDKEED\n",
      "real pep seq: NRDGKMDKKHKEEDK \t result seq: NRDGKMDKKDKHKEEDK\n",
      "real pep seq: NRDGKMDKKHKEEDK \t result seq: NRDGKMDKEEDK\n",
      "real pep seq: NRDGKMDKKHKEEDK \t result seq: NRDGKMDKEEDK\n",
      "real pep seq: NRDGKMDKKHKEEDKG \t result seq: NRDGKMDKKDKHKEEDKG\n",
      "real pep seq: NRDGKMDKKHKEEDKG \t result seq: NRDKDKHKEEDKG\n",
      "real pep seq: NRDGKMDKKHKEEDKG \t result seq: NRDKDKHKEEDKG\n",
      "real pep seq: NRDGKMDKKHKEEDKGS \t result seq: NRDGKMDKKDKHKEEDKGS\n",
      "real pep seq: NRDGKMDKKHKEEDKGS \t result seq: NRDKDKHKEEDKGS\n",
      "real pep seq: NRDGKMDKKHKEEDKGS \t result seq: NRDKDKHKEEDKGS\n",
      "real pep seq: KNRDGKMDKKHKE \t result seq: KNRDGKMDKHKE\n",
      "real pep seq: KNRDGKMDKKHKE \t result seq: KNRDGKMDKHKE\n",
      "real pep seq: KNRDGKMDKKHKEE \t result seq: KNRDGKMDKKDKHKEE\n",
      "real pep seq: KNRDGKMDKKHKEE \t result seq: KNRDGKMDKKRSTQKVPEE\n",
      "real pep seq: KNRDGKMDKKHKEE \t result seq: KNRDGKMDKKEE\n",
      "real pep seq: KNRDGKMDKKHKEED \t result seq: KNRDGKMDKKDKHKEED\n",
      "real pep seq: KNRDGKMDKKHKEED \t result seq: KNRDGKMDKKEED\n",
      "real pep seq: KNRDGKMDKKHKEED \t result seq: KNRDGKMDKEED\n",
      "real pep seq: KNRDGKMDKKHKEEDK \t result seq: KNRDGKMDKKDKHKEEDK\n",
      "real pep seq: KNRDGKMDKKHKEEDK \t result seq: KNRDGKMDKEEDK\n",
      "real pep seq: KNRDGKMDKKHKEEDK \t result seq: KNRDGKMDKEEDK\n",
      "real pep seq: KNRDGKMDKKHKEEDKG \t result seq: KNRDGKMDKKDKHKEEDKG\n",
      "real pep seq: KNRDGKMDKKHKEEDKG \t result seq: KNRDKDKHKEEDKG\n",
      "real pep seq: KNRDGKMDKKHKEEDKG \t result seq: KNRDGKMDKDKG\n",
      "real pep seq: KNRDGKMDKKHKEEDKGS \t result seq: KNRDGKMDKKDKHKEEDKGS\n",
      "real pep seq: KNRDGKMDKKHKEEDKGS \t result seq: KNRDKDKHKEEDKGS\n",
      "real pep seq: KNRDGKMDKKHKEEDKGS \t result seq: KNRDGKMDKKGS\n",
      "real pep seq: ESDAQESLHV \t result seq: ESDMAAQESLHV\n",
      "real pep seq: ESDAQESLHV \t result seq: ESDMAAQESLHV\n",
      "real pep seq: ESDAQESLHV \t result seq: ESDMAAQESLHV\n",
      "real pep seq: ESDAQESLHVK \t result seq: ESDMAAQESLHVK\n",
      "real pep seq: ESDAQESLHVK \t result seq: ESDMAAQESLHVK\n",
      "real pep seq: ESDAQESLHVK \t result seq: ESDMAAQESLHVK\n",
      "real pep seq: ESDAQESLHVKT \t result seq: ESDMAAQESLHVKT\n",
      "real pep seq: ESDAQESLHVKT \t result seq: ESDMAAQESLHVKT\n",
      "real pep seq: ESDAQESLHVKT \t result seq: ESDMAAQESLHVKT\n",
      "real pep seq: SESDAQESLHV \t result seq: SSEDQAMAAQESLHV\n",
      "real pep seq: SESDAQESLHV \t result seq: SESDMAAQESLHV\n",
      "real pep seq: SESDAQESLHV \t result seq: SESDMAAQESLHV\n",
      "real pep seq: SESDAQESLHVK \t result seq: SSEDQAMAAQESLHVK\n",
      "real pep seq: SESDAQESLHVK \t result seq: SESDMAAQESLHVK\n",
      "real pep seq: SESDAQESLHVK \t result seq: SESDMAAQESLHVK\n",
      "real pep seq: SESDAQESLHVKT \t result seq: SSEDQAMAAQESLHVKT\n",
      "real pep seq: SESDAQESLHVKT \t result seq: SESDMAAQESLHVKT\n",
      "real pep seq: SESDAQESLHVKT \t result seq: SESDMAAQESLHVKT\n",
      "real pep seq: ASESDAQES \t result seq: ASESSSAQES\n",
      "real pep seq: ASESDAQESL \t result seq: ASESDFIQAQESL\n",
      "real pep seq: ASESDAQESL \t result seq: ASESDFIQESI\n",
      "real pep seq: ASESDAQESLH \t result seq: ASESAQESLH\n",
      "real pep seq: ASESDAQESLHV \t result seq: ASESDMAAQESLHV\n",
      "real pep seq: ASESDAQESLHV \t result seq: ASESDMAAQESLHV\n",
      "real pep seq: ASESDAQESLHV \t result seq: ASESMAAQESLHV\n",
      "real pep seq: ASESDAQESLHVK \t result seq: ASESDMAAQESLHVK\n",
      "real pep seq: ASESDAQESLHVK \t result seq: ASESDMAAQESLHVK\n",
      "real pep seq: ASESDAQESLHVK \t result seq: ASESMAAQESLHVK\n",
      "real pep seq: ASESDAQESLHVKT \t result seq: ASESDMAAQESLHVKT\n",
      "real pep seq: ASESDAQESLHVKT \t result seq: ASESDMAAQESLHVKT\n",
      "real pep seq: ASESDAQESLHVKT \t result seq: ASESMAAQESLHVKT\n",
      "real pep seq: LASESDAQES \t result seq: LASESDQES\n",
      "real pep seq: LASESDAQES \t result seq: LASESDQES\n",
      "real pep seq: LASESDAQESL \t result seq: LASESAQESL\n",
      "real pep seq: LASESDAQESL \t result seq: LASESDESI\n",
      "real pep seq: LASESDAQESLH \t result seq: LASESAQESLH\n",
      "real pep seq: LASESDAQESLH \t result seq: LASESDESIH\n",
      "real pep seq: LASESDAQESLHV \t result seq: IASEMAAQESLHV\n",
      "real pep seq: LASESDAQESLHV \t result seq: LASEMAAQESLHV\n",
      "real pep seq: LASESDAQESLHV \t result seq: LASEMAAQESLHV\n",
      "real pep seq: LASESDAQESLHVK \t result seq: IASEMAAQESLHVK\n",
      "real pep seq: LASESDAQESLHVK \t result seq: LASEMAAQESLHVK\n",
      "real pep seq: LASESDAQESLHVK \t result seq: LASEMAAQESLHVK\n",
      "real pep seq: LASESDAQESLHVKT \t result seq: IASEMAAQESLHVKT\n",
      "real pep seq: LASESDAQESLHVKT \t result seq: LASEMAAQESLHVKT\n",
      "real pep seq: LASESDAQESLHVKT \t result seq: LASEMAAQESLHVKT\n",
      "real pep seq: DLASESDAQES \t result seq: DLASESDQES\n",
      "real pep seq: DLASESDAQES \t result seq: DLASESDQES\n",
      "real pep seq: DLASESDAQESL \t result seq: DLASESDESI\n",
      "real pep seq: DLASESDAQESL \t result seq: DLASESDESI\n",
      "real pep seq: DLASESDAQESLH \t result seq: DLASESDESIH\n",
      "real pep seq: DLASESDAQESLH \t result seq: DLASESDCSLH\n",
      "real pep seq: DLASESDAQESLHV \t result seq: DLASESDMAAQESLHV\n",
      "real pep seq: DLASESDAQESLHV \t result seq: DLASESDQEVTHV\n",
      "real pep seq: DLASESDAQESLHV \t result seq: DLASMAAQESLHV\n",
      "real pep seq: DLASESDAQESLHVK \t result seq: DLASESDMAAQESLHVK\n",
      "real pep seq: DLASESDAQESLHVK \t result seq: DLASMAAQESLHVK\n",
      "real pep seq: DLASESDAQESLHVK \t result seq: DIAMAAQESLHVK\n",
      "real pep seq: DLASESDAQESLHVKT \t result seq: DLASESDMAAQESLHVKT\n",
      "real pep seq: DLASESDAQESLHVKT \t result seq: DLASMAAQESLHVKT\n",
      "real pep seq: DLASESDAQESLHVKT \t result seq: DIAMAAQESLHVKT\n",
      "real pep seq: QDLASESDAQES \t result seq: QDLASESDQES\n",
      "real pep seq: QDLASESDAQES \t result seq: QDLASESDQES\n",
      "real pep seq: QDLASESDAQESL \t result seq: QDLASESDESI\n",
      "real pep seq: QDLASESDAQESL \t result seq: QDLASESDESI\n",
      "real pep seq: QDLASESDAQESLH \t result seq: QDLASESDQEVTH\n",
      "real pep seq: QDLASESDAQESLH \t result seq: QDLASESDESIH\n",
      "real pep seq: QDLASESDAQESLHV \t result seq: QDLASESDMAAQESLHV\n",
      "real pep seq: QDLASESDAQESLHV \t result seq: QDLASESDQEVTHV\n",
      "real pep seq: QDLASESDAQESLHV \t result seq: QDLAMAAQESLHV\n",
      "real pep seq: QDLASESDAQESLHVK \t result seq: QDLASESDMAAQESLHVK\n",
      "real pep seq: QDLASESDAQESLHVK \t result seq: QDLAMAAQESLHVK\n",
      "real pep seq: QDLASESDAQESLHVK \t result seq: QDIAMAAQESLHVK\n",
      "real pep seq: QDLASESDAQESLHVKT \t result seq: QDLASESDMAAQESLHVKT\n",
      "real pep seq: QDLASESDAQESLHVKT \t result seq: QDLAMAAQESLHVKT\n",
      "real pep seq: QDLASESDAQESLHVKT \t result seq: QDIAMAAQESLHVKT\n",
      "real pep seq: VQDLASESDAQES \t result seq: VQDLASESDQES\n",
      "real pep seq: VQDLASESDAQES \t result seq: VQDLASESDQES\n",
      "real pep seq: VQDLASESDAQESL \t result seq: VQDLASESDESI\n",
      "real pep seq: VQDLASESDAQESL \t result seq: VQDLASESDESI\n",
      "real pep seq: VQDLASESDAQESLH \t result seq: VQDLASESDAEESIH\n",
      "real pep seq: VQDLASESDAQESLH \t result seq: VQDLASESDSLH\n",
      "real pep seq: VQDLASESDAQESLHV \t result seq: VQDLASESDMAAQESLHV\n",
      "real pep seq: VQDLASESDAQESLHV \t result seq: VQDLASESDQEVTHV\n",
      "real pep seq: VQDLASESDAQESLHV \t result seq: VQDLASESDIHV\n",
      "real pep seq: VQDLASESDAQESLHVK \t result seq: VQDLASESDMAAQESLHVK\n",
      "real pep seq: VQDLASESDAQESLHVK \t result seq: VQDLASESDLHVK\n",
      "real pep seq: VQDLASESDAQESLHVK \t result seq: VQDLASESDHVK\n",
      "real pep seq: VQDLASESDAQESLHVKT \t result seq: VQDLASESDMAAQESLHVKT\n",
      "real pep seq: VQDLASESDAQESLHVKT \t result seq: VLRDISEMAAQESLHVKT\n",
      "real pep seq: VQDLASESDAQESLHVKT \t result seq: VQDMAAQESLHVKT\n",
      "real pep seq: NLSGDD \t result seq: NISGDD\n",
      "real pep seq: NLSGDD \t result seq: NISGDD\n",
      "real pep seq: NLSGDD \t result seq: NISGDD\n",
      "real pep seq: NLSGDDP \t result seq: NISGDDP\n",
      "real pep seq: NLSGDDP \t result seq: NISGDDP\n",
      "real pep seq: NLSGDDP \t result seq: NISGDDP\n",
      "real pep seq: NLSGDDPA \t result seq: NISGDDPA\n",
      "real pep seq: NLSGDDPA \t result seq: NISGDDPA\n",
      "real pep seq: NLSGDDPA \t result seq: NISGDDPA\n",
      "real pep seq: NLSGDDPAS \t result seq: NISGDDPAS\n",
      "real pep seq: NLSGDDPAS \t result seq: NISGDDPAS\n",
      "real pep seq: NLSGDDPAS \t result seq: NISGDDPAS\n",
      "real pep seq: NLSGDDPASP \t result seq: NISGDDPASP\n",
      "real pep seq: NLSGDDPASP \t result seq: NISGDDPASP\n",
      "real pep seq: NLSGDDPASP \t result seq: NNDWQSLGDDPASP\n",
      "real pep seq: NLSGDDPASPF \t result seq: NISGDDPASPF\n",
      "real pep seq: NLSGDDPASPF \t result seq: NISGDDPASPF\n",
      "real pep seq: NLSGDDPASPF \t result seq: NISGDDPASPF\n",
      "real pep seq: NLSGDDPASPFE \t result seq: NISGDDPASPFE\n",
      "real pep seq: NLSGDDPASPFE \t result seq: NISGDDPASPFE\n",
      "real pep seq: NLSGDDPASPFE \t result seq: NISGDDPASPFE\n",
      "real pep seq: GNLSGDDP \t result seq: GNISGDDP\n",
      "real pep seq: GNLSGDDP \t result seq: GNISGDDP\n",
      "real pep seq: GNLSGDDP \t result seq: GNISGDDP\n",
      "real pep seq: TGNLSGDDP \t result seq: TGNLSDDP\n",
      "real pep seq: TGNLSGDDP \t result seq: TGNLSDDP\n",
      "real pep seq: TGNLSGDDPA \t result seq: TGNLGDDPA\n",
      "real pep seq: TGNLSGDDPA \t result seq: TGNLSDPA\n",
      "real pep seq: TGNLSGDDPAS \t result seq: TGNLGDDPAS\n",
      "real pep seq: TGNLSGDDPAS \t result seq: TGNGDDPAS\n",
      "real pep seq: TGNLSGDDPASP \t result seq: TGNLGDDPASP\n",
      "real pep seq: TGNLSGDDPASP \t result seq: TGNLSDPASP\n",
      "real pep seq: TGNLSGDDPASPF \t result seq: TGNLGDDPASPF\n",
      "real pep seq: TGNLSGDDPASPF \t result seq: TGNGDDPASPF\n",
      "real pep seq: TGNLSGDDPASPFE \t result seq: TGNLGDDPASPFE\n",
      "real pep seq: TGNLSGDDPASPFE \t result seq: TGNGDDPASPFE\n",
      "real pep seq: LTGNLSGDDP \t result seq: LTGNLSTVNGVP\n",
      "real pep seq: LTGNLSGDDP \t result seq: LTGNLSDDP\n",
      "real pep seq: LTGNLSGDDPA \t result seq: LTGNLSDPA\n",
      "real pep seq: LTGNLSGDDPA \t result seq: LTGNLSDPA\n",
      "real pep seq: LTGNLSGDDPAS \t result seq: LTGNLSDPAS\n",
      "real pep seq: LTGNLSGDDPAS \t result seq: LTGNLSPAS\n",
      "real pep seq: LTGNLSGDDPASP \t result seq: LTGNLSDPASP\n",
      "real pep seq: LTGNLSGDDPASP \t result seq: LTGNLSTVNGVP\n",
      "real pep seq: LTGNLSGDDPASPF \t result seq: LTGGDDPASPF\n",
      "real pep seq: LTGNLSGDDPASPF \t result seq: LTGGDDPASPF\n",
      "real pep seq: LTGNLSGDDPASPFE \t result seq: LTGGDDPASPFE\n",
      "real pep seq: LTGNLSGDDPASPFE \t result seq: LTGGDDPASPFE\n",
      "real pep seq: ELTGNLSGDDP \t result seq: ELTGNLSGIDP\n",
      "real pep seq: ELTGNLSGDDP \t result seq: ELTGNLSIDP\n",
      "real pep seq: ELTGNLSGDDPA \t result seq: ELTGNLSDPA\n",
      "real pep seq: ELTGNLSGDDPA \t result seq: ELTGNLSDPA\n",
      "real pep seq: ELTGNLSGDDPAS \t result seq: ELTGNLSDPAS\n",
      "real pep seq: ELTGNLSGDDPAS \t result seq: ELTGNLSPAS\n",
      "real pep seq: ELTGNLSGDDPASP \t result seq: ELTGNLSDPASP\n",
      "real pep seq: ELTGNLSGDDPASP \t result seq: ELQDGDDPASP\n",
      "real pep seq: ELTGNLSGDDPASPF \t result seq: ELTGGDDPASPF\n",
      "real pep seq: ELTGNLSGDDPASPF \t result seq: EITGDDPASPF\n",
      "real pep seq: ELTGNLSGDDPASPFE \t result seq: ELTGGDDPASPFE\n",
      "real pep seq: ELTGNLSGDDPASPFE \t result seq: EITGDDPASPFE\n",
      "real pep seq: KELTGNLSGDDP \t result seq: KELTGNLSDDP\n",
      "real pep seq: KELTGNLSGDDP \t result seq: KELTGNLSDDP\n",
      "real pep seq: KELTGNLSGDDPA \t result seq: KELTGNLSLAPA\n",
      "real pep seq: KELTGNLSGDDPA \t result seq: KELTGNLSAPA\n",
      "real pep seq: KELTGNLSGDDPAS \t result seq: KELTGNLSDPAS\n",
      "real pep seq: KELTGNLSGDDPAS \t result seq: KELTGNLSPAS\n",
      "real pep seq: KELTGNLSGDDPASP \t result seq: KELTGNLSDPASP\n",
      "real pep seq: KELTGNLSGDDPASP \t result seq: KELTGNLSPASP\n",
      "real pep seq: KELTGNLSGDDPASPF \t result seq: KEITGDDPASPF\n",
      "real pep seq: KELTGNLSGDDPASPF \t result seq: KEITGDDPASPF\n",
      "real pep seq: KELTGNLSGDDPASPFE \t result seq: KEITGDDPASPFE\n",
      "real pep seq: KELTGNLSGDDPASPFE \t result seq: KEITGDDPASPFE\n",
      "real pep seq: IKELTGNLSGDDP \t result seq: IKELTGNLSDDP\n",
      "real pep seq: IKELTGNLSGDDP \t result seq: IKELTGNLSDDP\n",
      "real pep seq: IKELTGNLSGDDPA \t result seq: IKELTGNLSDPA\n",
      "real pep seq: IKELTGNLSGDDPA \t result seq: IKELTGNLSDPA\n",
      "real pep seq: IKELTGNLSGDDPAS \t result seq: IKELTGNLSDPAS\n",
      "real pep seq: IKELTGNLSGDDPAS \t result seq: IKELTGNLSPAS\n",
      "real pep seq: IKELTGNLSGDDPASP \t result seq: IKELTGNLSDPASP\n",
      "real pep seq: IKELTGNLSGDDPASP \t result seq: IKELTGNLSPASP\n",
      "real pep seq: IKELTGNLSGDDPASPF \t result seq: LKEITGDDPASPF\n",
      "real pep seq: IKELTGNLSGDDPASPF \t result seq: IKELTGNLSSPF\n",
      "real pep seq: IKELTGNLSGDDPASPFE \t result seq: LKEITGDDPASPFE\n",
      "real pep seq: IKELTGNLSGDDPASPFE \t result seq: IKELTGNLSPFE\n",
      "real pep seq: QKMGIRT \t result seq: QKMGLRT\n",
      "real pep seq: VQKMGIRT \t result seq: VQKMGLRT\n",
      "real pep seq: VQKMGIRTC \t result seq: VQKGIRTC\n",
      "real pep seq: VQKMGIRTC \t result seq: VQKGIRTC\n",
      "real pep seq: VQKMGIRTCD \t result seq: VQKGIRTCD\n",
      "real pep seq: VQKMGIRTCD \t result seq: VQKGIRTCD\n",
      "real pep seq: VQKMGIRTCDE \t result seq: VQKMGRCEISE\n",
      "real pep seq: VQKMGIRTCDE \t result seq: VQKGIRTCDE\n",
      "real pep seq: VQKMGIRTCDEY \t result seq: VQKGIRTCDEY\n",
      "real pep seq: VQKMGIRTCDEY \t result seq: VQKGIRTCDEY\n",
      "real pep seq: VQKMGIRTCDEYD \t result seq: VQKGIRTCDEYD\n",
      "real pep seq: VQKMGIRTCDEYD \t result seq: VQKGIRTCDEYD\n",
      "real pep seq: EVQKMGIRT \t result seq: EVQKMGLRT\n",
      "real pep seq: EVQKMGIRTC \t result seq: EVQKGIRTC\n",
      "real pep seq: EVQKMGIRTC \t result seq: EVQKGIRTC\n",
      "real pep seq: EVQKMGIRTCD \t result seq: EVQKGIRTCD\n",
      "real pep seq: EVQKMGIRTCD \t result seq: EVQKGIRTCD\n",
      "real pep seq: EVQKMGIRTCDE \t result seq: EVQKMGRCEISE\n",
      "real pep seq: EVQKMGIRTCDE \t result seq: EVQKGIRTCDE\n",
      "real pep seq: EVQKMGIRTCDEY \t result seq: EVQKGIRTCDEY\n",
      "real pep seq: EVQKMGIRTCDEY \t result seq: EVQKGIRTCDEY\n",
      "real pep seq: EVQKMGIRTCDEYD \t result seq: EVQKGIRTCDEYD\n",
      "real pep seq: EVQKMGIRTCDEYD \t result seq: EVQKGIRTCDEYD\n",
      "real pep seq: DEVQKMGIRT \t result seq: DEVQKMGLRT\n",
      "real pep seq: DEVQKMGIRTC \t result seq: DEVQKMGEQPVRC\n",
      "real pep seq: DEVQKMGIRTC \t result seq: DEVQKMRTC\n",
      "real pep seq: DEVQKMGIRTCD \t result seq: DEVQKMTCD\n",
      "real pep seq: DEVQKMGIRTCD \t result seq: DEVQKMTCD\n",
      "real pep seq: DEVQKMGIRTCDE \t result seq: DEVQKMGRCEISE\n",
      "real pep seq: DEVQKMGIRTCDE \t result seq: DEVGIRTCDE\n",
      "real pep seq: DEVQKMGIRTCDEY \t result seq: DEVGIRTCDEY\n",
      "real pep seq: DEVQKMGIRTCDEY \t result seq: DEVGIRTCDEY\n",
      "real pep seq: DEVQKMGIRTCDEYD \t result seq: DEVGIRTCDEYD\n",
      "real pep seq: DEVQKMGIRTCDEYD \t result seq: DEVGIRTCDEYD\n",
      "real pep seq: DDEVQKMGIRT \t result seq: DDEVQKMGLRT\n",
      "real pep seq: DDEVQKMGIRTC \t result seq: DDEVQKMRTC\n",
      "real pep seq: DDEVQKMGIRTC \t result seq: DDEVQKMRTC\n",
      "real pep seq: DDEVQKMGIRTCD \t result seq: DDEVQKMTCD\n",
      "real pep seq: DDEVQKMGIRTCD \t result seq: DDEVQKMTCD\n",
      "real pep seq: DDEVQKMGIRTCDE \t result seq: DDEVQKMGRCEISE\n",
      "real pep seq: DDEVQKMGIRTCDE \t result seq: DDEVQKMCDE\n",
      "real pep seq: DDEVQKMGIRTCDEY \t result seq: DDEGIRTCDEY\n",
      "real pep seq: DDEVQKMGIRTCDEY \t result seq: DDEGIRTCDEY\n",
      "real pep seq: DDEVQKMGIRTCDEYD \t result seq: DDEGIRTCDEYD\n",
      "real pep seq: DDEVQKMGIRTCDEYD \t result seq: DDEGIRTCDEYD\n",
      "real pep seq: EDDEVQKMGIRT \t result seq: EDDEVQKMGLRT\n",
      "real pep seq: EDDEVQKMGIRTC \t result seq: EDDEVQKMRTC\n",
      "real pep seq: EDDEVQKMGIRTC \t result seq: EDDEVQKMRTC\n",
      "real pep seq: EDDEVQKMGIRTCD \t result seq: EDDEVQKMTLVTCD\n",
      "real pep seq: EDDEVQKMGIRTCD \t result seq: EDDEVQKMTCD\n",
      "real pep seq: EDDEVQKMGIRTCDE \t result seq: EDDEVQKMGRCEISE\n",
      "real pep seq: EDDEVQKMGIRTCDE \t result seq: EDDEVQKMGLFADE\n",
      "real pep seq: EDDEVQKMGIRTCDEY \t result seq: EDDEGIRTCDEY\n",
      "real pep seq: EDDEVQKMGIRTCDEY \t result seq: EDDEGIRTCDEY\n",
      "real pep seq: EDDEVQKMGIRTCDEYD \t result seq: EDDEGIRTCDEYD\n",
      "real pep seq: EDDEVQKMGIRTCDEYD \t result seq: EDDEGIRTCDEYD\n",
      "real pep seq: LEDDEVQKMGIRT \t result seq: LEDDEVQKMGLRT\n",
      "real pep seq: LEDDEVQKMGIRTC \t result seq: LEDDEVQKMKQESEIHS\n",
      "real pep seq: LEDDEVQKMGIRTC \t result seq: LEDDEVQKMRTC\n",
      "real pep seq: LEDDEVQKMGIRTCD \t result seq: LEDDEVQKMGLMIQC\n",
      "real pep seq: LEDDEVQKMGIRTCD \t result seq: LEDDEVQKMTCD\n",
      "real pep seq: LEDDEVQKMGIRTCDE \t result seq: LEDDEVQKMGRCEISE\n",
      "real pep seq: LEDDEVQKMGIRTCDE \t result seq: LEDDEVQKMKQESEIHS\n",
      "real pep seq: LEDDEVQKMGIRTCDEY \t result seq: LEDDEVQKMKQESEIHS\n",
      "real pep seq: LEDDEVQKMGIRTCDEY \t result seq: IEDDLDGIRTCDEY\n",
      "real pep seq: LEDDEVQKMGIRTCDEYD \t result seq: IEDDLDGIRTCDEYD\n",
      "real pep seq: LEDDEVQKMGIRTCDEYD \t result seq: LEDGIRTCDEYD\n",
      "real pep seq: ARRKRELTI \t result seq: ARRKRELTL\n",
      "real pep seq: VARRKRELT \t result seq: VARKRELT\n",
      "real pep seq: VARRKRELTI \t result seq: VARKRELTI\n",
      "real pep seq: VARRKRELTIS \t result seq: VARKRELTIS\n",
      "real pep seq: VARRKRELTISP \t result seq: TTLRRKRELTISP\n",
      "real pep seq: VARRKRELTISPV \t result seq: VARKRELTISPV\n",
      "real pep seq: QVARRKRELT \t result seq: QVARRRELT\n",
      "real pep seq: QVARRKRELT \t result seq: QVARRRELT\n",
      "real pep seq: QVARRKRELTI \t result seq: QVARRRELTL\n",
      "real pep seq: QVARRKRELTI \t result seq: QVARRELTL\n",
      "real pep seq: QVARRKRELTIS \t result seq: QVAKRELTIS\n",
      "real pep seq: QVARRKRELTIS \t result seq: QVAKRELTIS\n",
      "real pep seq: QVARRKRELTISP \t result seq: QVAKRELTISP\n",
      "real pep seq: QVARRKRELTISP \t result seq: QVAKRELTISP\n",
      "real pep seq: QVARRKRELTISPV \t result seq: QVAKRELTISPV\n",
      "real pep seq: QVARRKRELTISPV \t result seq: QVAKRELTISPV\n",
      "real pep seq: LQVARRKRELT \t result seq: IQVAKALKRELT\n",
      "real pep seq: LQVARRKRELT \t result seq: LQVARRRELT\n",
      "real pep seq: LQVARRKRELTI \t result seq: IQVAKALKRELTI\n",
      "real pep seq: LQVARRKRELTI \t result seq: LQVARRRELTL\n",
      "real pep seq: LQVARRKRELTIS \t result seq: IQVAKALKRELTIS\n",
      "real pep seq: LQVARRKRELTIS \t result seq: LQVAKRELTIS\n",
      "real pep seq: LQVARRKRELTISP \t result seq: IQVAKALKRELTISP\n",
      "real pep seq: LQVARRKRELTISP \t result seq: LQVAKRELTISP\n",
      "real pep seq: LQVARRKRELTISPV \t result seq: ITSLKRELTISPV\n",
      "real pep seq: LQVARRKRELTISPV \t result seq: ITSLKRELTISPV\n",
      "real pep seq: PLQVARRKRELT \t result seq: PLQVARRRELT\n",
      "real pep seq: PLQVARRKRELT \t result seq: PLQVARRRELT\n",
      "real pep seq: PLQVARRKRELTI \t result seq: PLQVARRRELTL\n",
      "real pep seq: PLQVARRKRELTI \t result seq: PLQVARRELTL\n",
      "real pep seq: PLQVARRKRELTIS \t result seq: PLQVARRERITLS\n",
      "real pep seq: PLQVARRKRELTIS \t result seq: PLQVARRITIS\n",
      "real pep seq: PLQVARRKRELTISP \t result seq: TTLRRKRELTISP\n",
      "real pep seq: PLQVARRKRELTISP \t result seq: PLQVARRERITLSP\n",
      "real pep seq: PLQVARRKRELTISPV \t result seq: PIQKRELTISPV\n",
      "real pep seq: PLQVARRKRELTISPV \t result seq: PIQKRELTISPV\n",
      "real pep seq: SPLQVARRKRELT \t result seq: SPLQVARRRELT\n",
      "real pep seq: SPLQVARRKRELT \t result seq: SPLQVARRRELT\n",
      "real pep seq: SPLQVARRKRELTI \t result seq: SPLQVARRRELTL\n",
      "real pep seq: SPLQVARRKRELTI \t result seq: SPLQVARRELTL\n",
      "real pep seq: SPLQVARRKRELTIS \t result seq: SPLQVARRERITLS\n",
      "real pep seq: SPLQVARRKRELTIS \t result seq: SPLQVARRITIS\n",
      "real pep seq: SPLQVARRKRELTISP \t result seq: SPLQVARRERITLSP\n",
      "real pep seq: SPLQVARRKRELTISP \t result seq: SPLQVARRRRLSP\n",
      "real pep seq: SPLQVARRKRELTISPV \t result seq: SPLQKRELTISPV\n",
      "real pep seq: SPLQVARRKRELTISPV \t result seq: SPLKRELTISPV\n",
      "real pep seq: SSPLQVARRKRELT \t result seq: SSPLQVARRRELT\n",
      "real pep seq: SSPLQVARRKRELT \t result seq: SSPLQVARRRELT\n",
      "real pep seq: SSPLQVARRKRELTI \t result seq: SSPLQVARRRELTL\n",
      "real pep seq: SSPLQVARRKRELTI \t result seq: SSPLQVARRELTL\n",
      "real pep seq: SSPLQVARRKRELTIS \t result seq: SSPLQVARRERITLS\n",
      "real pep seq: SSPLQVARRKRELTIS \t result seq: SSPLQVARRNLTITIS\n",
      "real pep seq: SSPLQVARRKRELTISP \t result seq: SSPLQVARRERITLSP\n",
      "real pep seq: SSPLQVARRKRELTISP \t result seq: TTLRRKRELTISP\n",
      "real pep seq: SSPLQVARRKRELTISPV \t result seq: SSPIIKRELTISPV\n",
      "real pep seq: SSPLQVARRKRELTISPV \t result seq: SSPLKRELTISPV\n"
     ]
    }
   ],
   "source": [
    "expfile = '../data/testing_output/experiment_info.json'\n",
    "exp = json.load(open(expfile, 'r'))\n",
    "\n",
    "scan_no_keyed_results = {x['spectrum']['scan_number']: x for _, x in summary.items()}\n",
    "sorted_keys = [int(c) for c in exp['peptides'].keys()]\n",
    "\n",
    "for k in sorted_keys:\n",
    "    pep = exp['peptides'][str(k)]\n",
    "    if k not in scan_no_keyed_results:\n",
    "        continue\n",
    "    if 'hybrid' in pep['peptide_name'].lower():\n",
    "        hyb_calc(scan_no_keyed_results[k]['alignments'], pep)\n",
    "    else:\n",
    "        non_hyb_calc(scan_no_keyed_results[k]['alignments'], pep)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "NON HYBRID STATS\n",
      "------------------------------------------------------------\n",
      "number of peptides.......................................100\n",
      "Top alignment\n",
      "\n",
      "correct alignment.........................................98\n",
      "%.........................................................98\n",
      "correct protein...........................................98\n",
      "%.........................................................98\n",
      "correct starting position.................................98\n",
      "%.........................................................98\n",
      "correct ending position...................................98\n",
      "%.........................................................98\n",
      "correct length............................................98\n",
      "%.........................................................98\n",
      "\n",
      "2 to 5 alignment\n",
      "\n",
      "number of peptides.......................................100\n",
      "correct alignment..........................................1\n",
      "%..........................................................1\n",
      "correct protein............................................1\n",
      "%..........................................................1\n",
      "correct starting position..................................1\n",
      "%..........................................................1\n",
      "correct ending position....................................1\n",
      "%..........................................................1\n",
      "correct length.............................................1\n",
      "%..........................................................1\n",
      "\n",
      "============================================================\n",
      "\n",
      "HYBRID STATS\n",
      "------------------------------------------------------------\n",
      "number of peptides.......................................245\n",
      "Top alignment\n",
      "\n",
      "correct alignment........................................114\n",
      "%.........................................................46\n",
      "correct sequence.........................................173\n",
      "%.........................................................70\n",
      "correct left parent......................................146\n",
      "%.........................................................59\n",
      "correct right parent.....................................137\n",
      "%.........................................................55\n",
      "correct starting position................................142\n",
      "%.........................................................57\n",
      "correct ending position..................................137\n",
      "%.........................................................55\n",
      "correct length...........................................173\n",
      "%.........................................................70\n",
      "\n",
      "2 to 5 alignment\n",
      "\n",
      "correct alignment.........................................24\n",
      "%..........................................................9\n",
      "correct sequence.........................................147\n",
      "%.........................................................60\n",
      "correct left parent.......................................94\n",
      "%.........................................................38\n",
      "correct right parent......................................62\n",
      "%.........................................................25\n",
      "correct starting position.................................92\n",
      "%.........................................................37\n",
      "correct ending position...................................62\n",
      "%.........................................................25\n",
      "correct length.............................................1\n",
      "%.........................................................60\n",
      "\n"
     ]
    }
   ],
   "source": [
    "percent = lambda a, b: (a* 100 // b)\n",
    "\n",
    "printstat = lambda name, stat: '{}{}\\n'.format(name, str(stat).rjust(60-len(name), '.'))\n",
    "\n",
    "secbreak = ''.join(['=' for _ in range(60)])\n",
    "headbreak = ''.join(['-' for _ in range(60)])\n",
    "nhcount = non_hyb_stats['count']\n",
    "topalign = non_hyb_stats[0]\n",
    "otheralign = {}\n",
    "for i in range(1, n):\n",
    "    for stat in topalign.keys():\n",
    "        if stat not in otheralign:\n",
    "            otheralign[stat] = 0\n",
    "        otheralign[stat] += non_hyb_stats[i][stat]\n",
    "\n",
    "######################## NON HYBRID PRETTY PRINTING ############################\n",
    "\n",
    "nonhybsum = 'NON HYBRID STATS\\n' + headbreak + '\\n'\n",
    "nonhybsum += printstat('number of peptides', nhcount) \n",
    "nonhybsum += 'Top alignment\\n\\n'\n",
    "nonhybsum += printstat('correct alignment', topalign['correct'])\n",
    "nonhybsum += printstat('%', percent(topalign['correct'], nhcount)) \n",
    "nonhybsum += printstat('correct protein', topalign['correct_parent']) \n",
    "nonhybsum += printstat('%', percent(topalign['correct_parent'], nhcount))\n",
    "nonhybsum += printstat('correct starting position', topalign['correct_start'])\n",
    "nonhybsum += printstat('%', percent(topalign['correct_start'], nhcount))\n",
    "nonhybsum += printstat('correct ending position', topalign['correct_end'])\n",
    "nonhybsum += printstat('%', percent(topalign['correct_end'], nhcount))\n",
    "nonhybsum += printstat('correct length', topalign['correct_length'])\n",
    "nonhybsum += printstat('%', percent(topalign['correct_length'], nhcount))\n",
    "\n",
    "nonhybsum += '\\n2 to {} alignment\\n\\n'.format(n)\n",
    "nonhybsum += printstat('number of peptides', nhcount) \n",
    "nonhybsum += printstat('correct alignment', otheralign['correct'])\n",
    "nonhybsum += printstat('%', percent(otheralign['correct'], nhcount)) \n",
    "nonhybsum += printstat('correct protein', otheralign['correct_parent']) \n",
    "nonhybsum += printstat('%', percent(otheralign['correct_parent'], nhcount))\n",
    "nonhybsum += printstat('correct starting position', otheralign['correct_start'])\n",
    "nonhybsum += printstat('%', percent(otheralign['correct_start'], nhcount))\n",
    "nonhybsum += printstat('correct ending position', otheralign['correct_end'])\n",
    "nonhybsum += printstat('%', percent(otheralign['correct_end'], nhcount))\n",
    "nonhybsum += printstat('correct length', otheralign['correct_length'])\n",
    "nonhybsum += printstat('%', percent(otheralign['correct_length'], nhcount))\n",
    "nonhybsum += '\\n' + secbreak + '\\n\\n'\n",
    "\n",
    "############################ HYBRID PRETYY PRINTING ##############################\n",
    "hcount = hyb_stats['count']\n",
    "topalignh = hyb_stats[0]\n",
    "otheralignh = {}\n",
    "for i in range(1, n):\n",
    "    for stat in topalignh.keys():\n",
    "        if stat not in otheralignh:\n",
    "            otheralignh[stat] = 0\n",
    "        otheralignh[stat] += hyb_stats[i][stat]\n",
    "\n",
    "hybsum = 'HYBRID STATS\\n'+ headbreak + '\\n'\n",
    "hybsum += printstat('number of peptides', hcount)\n",
    "hybsum += 'Top alignment\\n\\n'\n",
    "hybsum += printstat('correct alignment', topalignh['correct'])\n",
    "hybsum += printstat('%', percent(topalignh['correct'], hcount))\n",
    "hybsum += printstat('correct sequence', topalignh['correct_sequence'])\n",
    "hybsum += printstat('%', percent(topalignh['correct_sequence'], hcount))\n",
    "hybsum += printstat('correct left parent', topalignh['left_correct_parent'])\n",
    "hybsum += printstat('%', percent(topalignh['left_correct_parent'], hcount))\n",
    "hybsum += printstat('correct right parent', topalignh['right_correct_parent'])\n",
    "hybsum += printstat('%', percent(topalignh['right_correct_parent'], hcount))\n",
    "hybsum += printstat('correct starting position', topalignh['correct_start'])\n",
    "hybsum += printstat('%', percent(topalignh['correct_start'], hcount))\n",
    "hybsum += printstat('correct ending position', topalignh['correct_end'])\n",
    "hybsum += printstat('%', percent(topalignh['correct_end'], hcount))\n",
    "hybsum += printstat('correct length', topalignh['correct_length'])\n",
    "hybsum += printstat('%', percent(topalignh['correct_length'], hcount))\n",
    "\n",
    "hybsum += '\\n2 to {} alignment\\n\\n'.format(n)\n",
    "hybsum += printstat('correct alignment', otheralignh['correct'])\n",
    "hybsum += printstat('%', percent(otheralignh['correct'], hcount))\n",
    "hybsum += printstat('correct sequence', otheralignh['correct_sequence'])\n",
    "hybsum += printstat('%', percent(otheralignh['correct_sequence'], hcount))\n",
    "hybsum += printstat('correct left parent', otheralignh['left_correct_parent'])\n",
    "hybsum += printstat('%', percent(otheralignh['left_correct_parent'], hcount))\n",
    "hybsum += printstat('correct right parent', otheralignh['right_correct_parent'])\n",
    "hybsum += printstat('%', percent(otheralignh['right_correct_parent'], hcount))\n",
    "hybsum += printstat('correct starting position', otheralignh['correct_start'])\n",
    "hybsum += printstat('%', percent(otheralignh['correct_start'], hcount))\n",
    "hybsum += printstat('correct ending position', otheralignh['correct_end'])\n",
    "hybsum += printstat('%', percent(otheralignh['correct_end'], hcount))\n",
    "hybsum += printstat('correct length', otheralign['correct_length'])\n",
    "hybsum += printstat('%', percent(otheralignh['correct_length'], hcount))\n",
    "print(nonhybsum + hybsum)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## For each of the hybrid missed alignments, see what was chosen instead"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Correct sequence: \t MDKKHKEE \t MDK-KHKEE\n",
      "Attempted alignments\n",
      "\tMDKKDKHKEE\tMDK-KDKHKEE\n",
      "\tMDKKDKHKEE\tMDK-KDKHKEE\n",
      "\tMDKKDKHKEE\tMDK-KDKHKEE\n",
      "\n",
      "Correct sequence: \t MDKKHKEED \t MDK-KHKEED\n",
      "Attempted alignments\n",
      "\tMMEKDKHKEED\tMME-KDKHKEED\n",
      "\tMDKKDKHKEED\tMDK-KDKHKEED\n",
      "\tMDKKDKHKEED\tMDK-KDKHKEED\n",
      "\n",
      "Correct sequence: \t MDKKHKEEDK \t MDK-KHKEEDK\n",
      "Attempted alignments\n",
      "\tMDKKDKHKEEDK\tMDK-KDKHKEEDK\n",
      "\tMDKKDKHKEEDK\tMDK-KDKHKEEDK\n",
      "\tMDKKDKHKEEDK\tMDK-KDKHKEEDK\n",
      "\n",
      "Correct sequence: \t MDKKHKEEDKG \t MDK-KHKEEDKG\n",
      "Attempted alignments\n",
      "\tMDKKDKHKEEDKG\tMDK-KDKHKEEDKG\n",
      "\tMDKKDKHKEEDKG\tMDK-KDKHKEEDKG\n",
      "\tMDKKDKHKEEDKG\tMDK-KDKHKEEDKG\n",
      "\n",
      "Correct sequence: \t MDKKHKEEDKGS \t MDK-KHKEEDKGS\n",
      "Attempted alignments\n",
      "\tMDKKDKHKEEDKGS\tMDK-KDKHKEEDKGS\n",
      "\tMDKKDKHKEEDKGS\tMDK-KDKHKEEDKGS\n",
      "\tMDKKDKHKEEDKGS\tMDK-KDKHKEEDKGS\n",
      "\n",
      "Correct sequence: \t KMDKKHKEE \t KMDK-KHKEE\n",
      "Attempted alignments\n",
      "\tKMKDKHKEE\tKM-KDKHKEE\n",
      "\tKMDKKDKHKEE\tKMDK-KDKHKEE\n",
      "\tKMDKDKHKEE\tKMD-KDKHKEE\n",
      "\n",
      "Correct sequence: \t KMDKKHKEED \t KMDK-KHKEED\n",
      "Attempted alignments\n",
      "\tKMDKKDKHKEED\tKMDK-KDKHKEED\n",
      "\tKMDKDKHKEED\tKMD-KDKHKEED\n",
      "\tKMDKDKHKEED\tKMD-KDKHKEED\n",
      "\n",
      "Correct sequence: \t KMDKKHKEEDK \t KMDK-KHKEEDK\n",
      "Attempted alignments\n",
      "\tKMDKKDKHKEEDK\tKMDK-KDKHKEEDK\n",
      "\tKMMKDKHKEEDK\tKMM-KDKHKEEDK\n",
      "\tKMDKDKHKEEDK\tKMD-KDKHKEEDK\n",
      "\n",
      "Correct sequence: \t KMDKKHKEEDKG \t KMDK-KHKEEDKG\n",
      "Attempted alignments\n",
      "\tKMKDKHKEEDKG\tKM-KDKHKEEDKG\n",
      "\tKMDKKDKHKEEDKG\tKMDK-KDKHKEEDKG\n",
      "\tKMDKDKHKEEDKG\tKMD-KDKHKEEDKG\n",
      "\n",
      "Correct sequence: \t KMDKKHKEEDKGS \t KMDK-KHKEEDKGS\n",
      "Attempted alignments\n",
      "\tKMKDKHKEEDKGS\tKM-KDKHKEEDKGS\n",
      "\tKMDKKDKHKEEDKGS\tKMDK-KDKHKEEDKGS\n",
      "\tKMDKDKHKEEDKGS\tKMD-KDKHKEEDKGS\n",
      "\n",
      "Correct sequence: \t GKMDKKHKEE \t GKMDK-KHKEE\n",
      "Attempted alignments\n",
      "\tGKMKDKHKEE\tGKM-KDKHKEE\n",
      "\tGKMKDKHKEE\tGKM-KDKHKEE\n",
      "\tGKMKDKHKEE\tGKM-KDKHKEE\n",
      "\n",
      "Correct sequence: \t GKMDKKHKEED \t GKMDK-KHKEED\n",
      "Attempted alignments\n",
      "\tGKMKDKHKEED\tGKM-KDKHKEED\n",
      "\tGKMKDKHKEED\tGKM-KDKHKEED\n",
      "\tGKMKDKHKEED\tGKM-KDKHKEED\n",
      "\n",
      "Correct sequence: \t GKMDKKHKEEDK \t GKMDK-KHKEEDK\n",
      "Attempted alignments\n",
      "\tGKMKDKHKEEDK\tGKM-KDKHKEEDK\n",
      "\tGKMKDKHKEEDK\tGKM-KDKHKEEDK\n",
      "\tGKMKDKHKEEDK\tGKM-KDKHKEEDK\n",
      "\n",
      "Correct sequence: \t GKMDKKHKEEDKG \t GKMDK-KHKEEDKG\n",
      "Attempted alignments\n",
      "\tGKMKDKHKEEDKG\tGKM-KDKHKEEDKG\n",
      "\tGKMKDKHKEEDKG\tGKM-KDKHKEEDKG\n",
      "\tGKMKDKHKEEDKG\tGKM-KDKHKEEDKG\n",
      "\n",
      "Correct sequence: \t GKMDKKHKEEDKGS \t GKMDK-KHKEEDKGS\n",
      "Attempted alignments\n",
      "\tGKMKDKHKEEDKGS\tGKM-KDKHKEEDKGS\n",
      "\tGKMKDKHKEEDKGS\tGKM-KDKHKEEDKGS\n",
      "\tGKMKDKHKEEDKGS\tGKM-KDKHKEEDKGS\n",
      "\n",
      "Correct sequence: \t DGKMDKKHKEE \t DGKMDK-KHKEE\n",
      "Attempted alignments\n",
      "\tDGKMDKKDKHKEE\tDGKMDK-KDKHKEE\n",
      "\tDGKMDKKEE\tDGKMDK-KEE\n",
      "\tDGKMDKKEE\tDGKMDK-KEE\n",
      "\n",
      "Correct sequence: \t DGKMDKKHKEED \t DGKMDK-KHKEED\n",
      "Attempted alignments\n",
      "\tDGKMDKKDKHKEED\tDGKMDK-KDKHKEED\n",
      "\tDGKMDKKEED\tDGKMDK-KEED\n",
      "\tDGKMDKEED\tDGKMDK-EED\n",
      "\n",
      "Correct sequence: \t DGKMDKKHKEEDK \t DGKMDK-KHKEEDK\n",
      "Attempted alignments\n",
      "\tDGKMDKKDKHKEEDK\tDGKMDK-KDKHKEEDK\n",
      "\tDGKMDKEEDK\tDGKMDK-EEDK\n",
      "\tDGKMDKEEDK\tDGKMDK-EEDK\n",
      "\n",
      "Correct sequence: \t DGKMDKKHKEEDKG \t DGKMDK-KHKEEDKG\n",
      "Attempted alignments\n",
      "\tDGKMDKKDKHKEEDKG\tDGKMDK-KDKHKEEDKG\n",
      "\tDGKKDKHKEEDKG\tDGK-KDKHKEEDKG\n",
      "\tDGKKDKHKEEDKG\tDGK-KDKHKEEDKG\n",
      "\n",
      "Correct sequence: \t DGKMDKKHKEEDKGS \t DGKMDK-KHKEEDKGS\n",
      "Attempted alignments\n",
      "\tDGKMDKKDKHKEEDKGS\tDGKMDK-KDKHKEEDKGS\n",
      "\tDGKKDKHKEEDKGS\tDGK-KDKHKEEDKGS\n",
      "\tDGKKDKHKEEDKGS\tDGK-KDKHKEEDKGS\n",
      "\n",
      "Correct sequence: \t RDGKMDKKHKEE \t RDGKMDK-KHKEE\n",
      "Attempted alignments\n",
      "\tRDGKMDKKDKHKEE\tRDGKMDK-KDKHKEE\n",
      "\tRDGKMDKKEE\tRDGKMDK-KEE\n",
      "\tRDGKMDKKEE\tRDGKMDK-KEE\n",
      "\n",
      "Correct sequence: \t RDGKMDKKHKEED \t RDGKMDK-KHKEED\n",
      "Attempted alignments\n",
      "\tRDGKMDKKDKHKEED\tRDGKMDK-KDKHKEED\n",
      "\tRDGKMDKKEED\tRDGKMDK-KEED\n",
      "\tRDGKMDKEED\tRDGKMDK-EED\n",
      "\n",
      "Correct sequence: \t RDGKMDKKHKEEDK \t RDGKMDK-KHKEEDK\n",
      "Attempted alignments\n",
      "\tRDGKMDKKDKHKEEDK\tRDGKMDK-KDKHKEEDK\n",
      "\tRDGKMDKEEDK\tRDGKMDK-EEDK\n",
      "\tRDGKMDKEEDK\tRDGKMDK-EEDK\n",
      "\n",
      "Correct sequence: \t RDGKMDKKHKEEDKG \t RDGKMDK-KHKEEDKG\n",
      "Attempted alignments\n",
      "\tRDGKMDKKDKHKEEDKG\tRDGKMDK-KDKHKEEDKG\n",
      "\tRDGKKDKHKEEDKG\tRDGK-KDKHKEEDKG\n",
      "\tRDGKDKHKEEDKG\tRDG-KDKHKEEDKG\n",
      "\n",
      "Correct sequence: \t RDGKMDKKHKEEDKGS \t RDGKMDK-KHKEEDKGS\n",
      "Attempted alignments\n",
      "\tRDGKMDKKDKHKEEDKGS\tRDGKMDK-KDKHKEEDKGS\n",
      "\tRDGKKDKHKEEDKGS\tRDGK-KDKHKEEDKGS\n",
      "\tRDGKDKHKEEDKGS\tRDG-KDKHKEEDKGS\n",
      "\n",
      "Correct sequence: \t NRDGKMDKKHKEE \t NRDGKMDK-KHKEE\n",
      "Attempted alignments\n",
      "\tNRDGKMDKKDKHKEE\tNRDGKMDK-KDKHKEE\n",
      "\tNRDGKMDKKEE\tNRDGKMDK-KEE\n",
      "\tNRDGKMDKKEE\tNRDGKMDK-KEE\n",
      "\n",
      "Correct sequence: \t NRDGKMDKKHKEED \t NRDGKMDK-KHKEED\n",
      "Attempted alignments\n",
      "\tNRDGKMDKKDKHKEED\tNRDGKMDK-KDKHKEED\n",
      "\tNRDGKMDKKEED\tNRDGKMDK-KEED\n",
      "\tNRDGKMDKEED\tNRDGKMDK-EED\n",
      "\n",
      "Correct sequence: \t NRDGKMDKKHKEEDK \t NRDGKMDK-KHKEEDK\n",
      "Attempted alignments\n",
      "\tNRDGKMDKKDKHKEEDK\tNRDGKMDK-KDKHKEEDK\n",
      "\tNRDGKMDKEEDK\tNRDGKMDK-EEDK\n",
      "\tNRDGKMDKEEDK\tNRDGKMDK-EEDK\n",
      "\n",
      "Correct sequence: \t NRDGKMDKKHKEEDKG \t NRDGKMDK-KHKEEDKG\n",
      "Attempted alignments\n",
      "\tNRDGKMDKKDKHKEEDKG\tNRDGKMDK-KDKHKEEDKG\n",
      "\tNRDKDKHKEEDKG\tNRD-KDKHKEEDKG\n",
      "\tNRDKDKHKEEDKG\tNRD-KDKHKEEDKG\n",
      "\n",
      "Correct sequence: \t NRDGKMDKKHKEEDKGS \t NRDGKMDK-KHKEEDKGS\n",
      "Attempted alignments\n",
      "\tNRDGKMDKKDKHKEEDKGS\tNRDGKMDK-KDKHKEEDKGS\n",
      "\tNRDKDKHKEEDKGS\tNRD-KDKHKEEDKGS\n",
      "\tNRDKDKHKEEDKGS\tNRD-KDKHKEEDKGS\n",
      "\n",
      "Correct sequence: \t KNRDGKMDKKHKEE \t KNRDGKMDK-KHKEE\n",
      "Attempted alignments\n",
      "\tKNRDGKMDKKDKHKEE\tKNRDGKMDK-KDKHKEE\n",
      "\tKNRDGKMDKKRSTQKVPEE\tKNRDGKMDK-KRSTQKVPEE\n",
      "\tKNRDGKMDKKEE\tKNRDGKMDK-KEE\n",
      "\n",
      "Correct sequence: \t KNRDGKMDKKHKEED \t KNRDGKMDK-KHKEED\n",
      "Attempted alignments\n",
      "\tKNRDGKMDKKDKHKEED\tKNRDGKMDK-KDKHKEED\n",
      "\tKNRDGKMDKKEED\tKNRDGKMDK-KEED\n",
      "\tKNRDGKMDKEED\tKNRDGKMDK-EED\n",
      "\n",
      "Correct sequence: \t KNRDGKMDKKHKEEDK \t KNRDGKMDK-KHKEEDK\n",
      "Attempted alignments\n",
      "\tKNRDGKMDKKDKHKEEDK\tKNRDGKMDK-KDKHKEEDK\n",
      "\tKNRDGKMDKEEDK\tKNRDGKMDK-EEDK\n",
      "\tKNRDGKMDKEEDK\tKNRDGKMDK-EEDK\n",
      "\n",
      "Correct sequence: \t KNRDGKMDKKHKEEDKG \t KNRDGKMDK-KHKEEDKG\n",
      "Attempted alignments\n",
      "\tKNRDGKMDKKDKHKEEDKG\tKNRDGKMDK-KDKHKEEDKG\n",
      "\tKNRDKDKHKEEDKG\tKNRD-KDKHKEEDKG\n",
      "\tKNRDGKMDKDKG\tKNRDGKMDK-DKG\n",
      "\n",
      "Correct sequence: \t KNRDGKMDKKHKEEDKGS \t KNRDGKMDK-KHKEEDKGS\n",
      "Attempted alignments\n",
      "\tKNRDGKMDKKDKHKEEDKGS\tKNRDGKMDK-KDKHKEEDKGS\n",
      "\tKNRDKDKHKEEDKGS\tKNRD-KDKHKEEDKGS\n",
      "\tKNRDGKMDKKGS\tKNRDGKMDK-KGS\n",
      "\n",
      "Correct sequence: \t ESDAQESLHV \t ESD-AQESLHV\n",
      "Attempted alignments\n",
      "\tESDMAAQESLHV\tESD-MAAQESLHV\n",
      "\tESDMAAQESLHV\tESD-MAAQESLHV\n",
      "\tESDMAAQESLHV\tESD-MAAQESLHV\n",
      "\n",
      "Correct sequence: \t ESDAQESLHVK \t ESD-AQESLHVK\n",
      "Attempted alignments\n",
      "\tESDMAAQESLHVK\tESD-MAAQESLHVK\n",
      "\tESDMAAQESLHVK\tESD-MAAQESLHVK\n",
      "\tESDMAAQESLHVK\tESD-MAAQESLHVK\n",
      "\n",
      "Correct sequence: \t ESDAQESLHVKT \t ESD-AQESLHVKT\n",
      "Attempted alignments\n",
      "\tESDMAAQESLHVKT\tESD-MAAQESLHVKT\n",
      "\tESDMAAQESLHVKT\tESD-MAAQESLHVKT\n",
      "\tESDMAAQESLHVKT\tESD-MAAQESLHVKT\n",
      "\n",
      "Correct sequence: \t SESDAQESLHV \t SESD-AQESLHV\n",
      "Attempted alignments\n",
      "\tSSEDQAMAAQESLHV\tSSEDQA-MAAQESLHV\n",
      "\tSESDMAAQESLHV\tSESD-MAAQESLHV\n",
      "\tSESDMAAQESLHV\tSESD-MAAQESLHV\n",
      "\n",
      "Correct sequence: \t SESDAQESLHVK \t SESD-AQESLHVK\n",
      "Attempted alignments\n",
      "\tSSEDQAMAAQESLHVK\tSSEDQA-MAAQESLHVK\n",
      "\tSESDMAAQESLHVK\tSESD-MAAQESLHVK\n",
      "\tSESDMAAQESLHVK\tSESD-MAAQESLHVK\n",
      "\n",
      "Correct sequence: \t SESDAQESLHVKT \t SESD-AQESLHVKT\n",
      "Attempted alignments\n",
      "\tSSEDQAMAAQESLHVKT\tSSEDQA-MAAQESLHVKT\n",
      "\tSESDMAAQESLHVKT\tSESD-MAAQESLHVKT\n",
      "\tSESDMAAQESLHVKT\tSESD-MAAQESLHVKT\n",
      "\n",
      "Correct sequence: \t ASESDAQESLHV \t ASESD-AQESLHV\n",
      "Attempted alignments\n",
      "\tASESDMAAQESLHV\tASESD-MAAQESLHV\n",
      "\tASESDMAAQESLHV\tASESD-MAAQESLHV\n",
      "\tASESMAAQESLHV\tASES-MAAQESLHV\n",
      "\n",
      "Correct sequence: \t ASESDAQESLHVK \t ASESD-AQESLHVK\n",
      "Attempted alignments\n",
      "\tASESDMAAQESLHVK\tASESD-MAAQESLHVK\n",
      "\tASESDMAAQESLHVK\tASESD-MAAQESLHVK\n",
      "\tASESMAAQESLHVK\tASES-MAAQESLHVK\n",
      "\n",
      "Correct sequence: \t ASESDAQESLHVKT \t ASESD-AQESLHVKT\n",
      "Attempted alignments\n",
      "\tASESDMAAQESLHVKT\tASESD-MAAQESLHVKT\n",
      "\tASESDMAAQESLHVKT\tASESD-MAAQESLHVKT\n",
      "\tASESMAAQESLHVKT\tASES-MAAQESLHVKT\n",
      "\n",
      "Correct sequence: \t LASESDAQESLHV \t LASESD-AQESLHV\n",
      "Attempted alignments\n",
      "\tIASEMAAQESLHV\tIASE-MAAQESLHV\n",
      "\tLASEMAAQESLHV\tLASE-MAAQESLHV\n",
      "\tLASEMAAQESLHV\tLASE-MAAQESLHV\n",
      "\n",
      "Correct sequence: \t LASESDAQESLHVK \t LASESD-AQESLHVK\n",
      "Attempted alignments\n",
      "\tIASEMAAQESLHVK\tIASE-MAAQESLHVK\n",
      "\tLASEMAAQESLHVK\tLASE-MAAQESLHVK\n",
      "\tLASEMAAQESLHVK\tLASE-MAAQESLHVK\n",
      "\n",
      "Correct sequence: \t LASESDAQESLHVKT \t LASESD-AQESLHVKT\n",
      "Attempted alignments\n",
      "\tIASEMAAQESLHVKT\tIASE-MAAQESLHVKT\n",
      "\tLASEMAAQESLHVKT\tLASE-MAAQESLHVKT\n",
      "\tLASEMAAQESLHVKT\tLASE-MAAQESLHVKT\n",
      "\n",
      "Correct sequence: \t DLASESDAQESLHV \t DLASESD-AQESLHV\n",
      "Attempted alignments\n",
      "\tDLASESDMAAQESLHV\tDLASESD-MAAQESLHV\n",
      "\tDLASESDQEVTHV\tDLASESD-QEVTHV\n",
      "\tDLASMAAQESLHV\tDLAS-MAAQESLHV\n",
      "\n",
      "Correct sequence: \t DLASESDAQESLHVK \t DLASESD-AQESLHVK\n",
      "Attempted alignments\n",
      "\tDLASESDMAAQESLHVK\tDLASESD-MAAQESLHVK\n",
      "\tDLASMAAQESLHVK\tDLAS-MAAQESLHVK\n",
      "\tDIAMAAQESLHVK\tDIA-MAAQESLHVK\n",
      "\n",
      "Correct sequence: \t DLASESDAQESLHVKT \t DLASESD-AQESLHVKT\n",
      "Attempted alignments\n",
      "\tDLASESDMAAQESLHVKT\tDLASESD-MAAQESLHVKT\n",
      "\tDLASMAAQESLHVKT\tDLAS-MAAQESLHVKT\n",
      "\tDIAMAAQESLHVKT\tDIA-MAAQESLHVKT\n",
      "\n",
      "Correct sequence: \t QDLASESDAQESLHV \t QDLASESD-AQESLHV\n",
      "Attempted alignments\n",
      "\tQDLASESDMAAQESLHV\tQDLASESD-MAAQESLHV\n",
      "\tQDLASESDQEVTHV\tQDLASESD-QEVTHV\n",
      "\tQDLAMAAQESLHV\tQDLA-MAAQESLHV\n",
      "\n",
      "Correct sequence: \t QDLASESDAQESLHVK \t QDLASESD-AQESLHVK\n",
      "Attempted alignments\n",
      "\tQDLASESDMAAQESLHVK\tQDLASESD-MAAQESLHVK\n",
      "\tQDLAMAAQESLHVK\tQDLA-MAAQESLHVK\n",
      "\tQDIAMAAQESLHVK\tQDIA-MAAQESLHVK\n",
      "\n",
      "Correct sequence: \t QDLASESDAQESLHVKT \t QDLASESD-AQESLHVKT\n",
      "Attempted alignments\n",
      "\tQDLASESDMAAQESLHVKT\tQDLASESD-MAAQESLHVKT\n",
      "\tQDLAMAAQESLHVKT\tQDLA-MAAQESLHVKT\n",
      "\tQDIAMAAQESLHVKT\tQDIA-MAAQESLHVKT\n",
      "\n",
      "Correct sequence: \t VQDLASESDAQESLHV \t VQDLASESD-AQESLHV\n",
      "Attempted alignments\n",
      "\tVQDLASESDMAAQESLHV\tVQDLASESD-MAAQESLHV\n",
      "\tVQDLASESDQEVTHV\tVQDLASESD-QEVTHV\n",
      "\tVQDLASESDIHV\tVQDLASESD-IHV\n",
      "\n",
      "Correct sequence: \t VQDLASESDAQESLHVK \t VQDLASESD-AQESLHVK\n",
      "Attempted alignments\n",
      "\tVQDLASESDMAAQESLHVK\tVQDLASESD-MAAQESLHVK\n",
      "\tVQDLASESDLHVK\tVQDLASESD-LHVK\n",
      "\tVQDLASESDHVK\tVQDLASESD-HVK\n",
      "\n",
      "Correct sequence: \t VQDLASESDAQESLHVKT \t VQDLASESD-AQESLHVKT\n",
      "Attempted alignments\n",
      "\tVQDLASESDMAAQESLHVKT\tVQDLASESD-MAAQESLHVKT\n",
      "\tVLRDISEMAAQESLHVKT\tVLRDISE-MAAQESLHVKT\n",
      "\tVQDMAAQESLHVKT\tVQD-MAAQESLHVKT\n",
      "\n",
      "Correct sequence: \t NLSGDD \t NLS-GDD\n",
      "Attempted alignments\n",
      "\tNISGDD\tNIS-GDD\n",
      "\tNISGDD\tNIS-GDD\n",
      "\tNISGDD\tNIS-GDD\n",
      "\n",
      "Correct sequence: \t NLSGDDP \t NLS-GDDP\n",
      "Attempted alignments\n",
      "\tNISGDDP\tNISG-DDP\n",
      "\tNISGDDP\tNISG-DDP\n",
      "\tNISGDDP\tNISG-DDP\n",
      "\n",
      "Correct sequence: \t NLSGDDPA \t NLS-GDDPA\n",
      "Attempted alignments\n",
      "\tNISGDDPA\tNIS-GDDPA\n",
      "\tNISGDDPA\tNIS-GDDPA\n",
      "\tNISGDDPA\tNIS-GDDPA\n",
      "\n",
      "Correct sequence: \t NLSGDDPAS \t NLS-GDDPAS\n",
      "Attempted alignments\n",
      "\tNISGDDPAS\tNIS-GDDPAS\n",
      "\tNISGDDPAS\tNIS-GDDPAS\n",
      "\tNISGDDPAS\tNIS-GDDPAS\n",
      "\n",
      "Correct sequence: \t NLSGDDPASP \t NLS-GDDPASP\n",
      "Attempted alignments\n",
      "\tNISGDDPASP\tNIS-GDDPASP\n",
      "\tNISGDDPASP\tNIS-GDDPASP\n",
      "\tNNDWQSLGDDPASP\tNNDWQSL-GDDPASP\n",
      "\n",
      "Correct sequence: \t NLSGDDPASPF \t NLS-GDDPASPF\n",
      "Attempted alignments\n",
      "\tNISGDDPASPF\tNIS-GDDPASPF\n",
      "\tNISGDDPASPF\tNIS-GDDPASPF\n",
      "\tNISGDDPASPF\tNIS-GDDPASPF\n",
      "\n",
      "Correct sequence: \t NLSGDDPASPFE \t NLS-GDDPASPFE\n",
      "Attempted alignments\n",
      "\tNISGDDPASPFE\tNIS-GDDPASPFE\n",
      "\tNISGDDPASPFE\tNIS-GDDPASPFE\n",
      "\tNISGDDPASPFE\tNIS-GDDPASPFE\n",
      "\n",
      "Correct sequence: \t GNLSGDDP \t GNLS-GDDP\n",
      "Attempted alignments\n",
      "\tGNISGDDP\tGNISG-DDP\n",
      "\tGNISGDDP\tGNISG-DDP\n",
      "\tGNISGDDP\tGNISG-DDP\n"
     ]
    }
   ],
   "source": [
    "for i, bad in enumerate(wrong_hybrid_alignemnts):\n",
    "    print('\\nCorrect sequence: \\t {} \\t {}'.format(bad[1]['sequence'], bad[1]['hybrid_sequence']))\n",
    "    print('Attempted alignments')\n",
    "    for b in bad[0]:\n",
    "        print('\\t{}\\t{}'.format(b['sequence'], b['hybrid_sequence']))\n",
    "\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "119\n"
     ]
    }
   ],
   "source": [
    "print(len(wrong_hybrid_alignemnts))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
