{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Debug the real spectra\n",
    "\n",
    "So it doesn't work well with real data, so having taken out one example of real data, the sequence is\n",
    "\n",
    "Non hybrid:\n",
    "```\n",
    "DLPVNSPMTKG\n",
    "```\n",
    "\n",
    "Non hybrid: \n",
    "```\n",
    "DAAIVGYK\n",
    "```\n",
    "\n",
    "Non hybrid:\n",
    "```\n",
    "DEAPNFEANTTIGRIRFH\n",
    "```\n",
    "\n",
    "Non hybrid:\n",
    "``` \n",
    "DPQVEQLEL\n",
    "```\n",
    "\n",
    "Hybrid 1: \n",
    "```\n",
    "DLQTLALWSRM\n",
    "```\n",
    "\n",
    "Non hybrid:\n",
    "```\n",
    "DLTEYLSRFGEVV\n",
    "```\n",
    "\n",
    "Hybrid 2:\n",
    "```\n",
    "DLQTLEVE\n",
    "```\n",
    "\n",
    "Non Hybrid:\n",
    "```\n",
    "DLSSASAIMGNAKVKAHGKKVITAFN\n",
    "```\n",
    "\n",
    "Non hybrid:\n",
    "```\n",
    "EEILKAFKLF\n",
    "```\n",
    "\n",
    "so toy with the tool until we find out whats wrong"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Load params and run"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import sys\n",
    "module_path = os.path.abspath(os.path.join('..'))\n",
    "if module_path not in sys.path:\n",
    "    sys.path.append(module_path)\n",
    "module_path = os.path.abspath(os.path.join('../..'))\n",
    "if module_path not in sys.path:\n",
    "    sys.path.append(module_path)\n",
    "    \n",
    "from src import runner\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "specPath = '/Users/zacharymcgrath/Desktop/nod2 data/single/'\n",
    "fastaPath = '/Users/zacharymcgrath/Desktop/nod2 data/filteredNOD2.fasta'\n",
    "outputDir = '/Users/zacharymcgrath/Desktop/Experiment output/singleReal/'\n",
    "minPep = 3\n",
    "maxPep = 30\n",
    "tolerance = 20\n",
    "relative_abundance_filter = 0.0\n",
    "precursor_tolerance = 3\n",
    "peak_filter = 25\n",
    "verbose = True\n",
    "scoringAlg = 'iip'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "params = {\n",
    "    'spectra_folder': specPath,\n",
    "    'database_file': fastaPath,\n",
    "    'output_dir': outputDir,\n",
    "    'min_peptide_len': minPep,\n",
    "    'max_peptide_len': maxPep,\n",
    "    'tolerance': tolerance,\n",
    "    'precursor_tolerance': precursor_tolerance,\n",
    "    'peak_filter': peak_filter, \n",
    "    'relative_abundance_filter': relative_abundance_filter,\n",
    "    'verbose': verbose, \n",
    "    'scoring_alg': scoringAlg, \n",
    "    'DEBUG': True\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Loading database...\n",
      "Adding protein 279/279 to tree\n",
      "Done.\n",
      "Building hashes for kmers...\n",
      "Looking at protein 279/279\n",
      "Done.\n",
      "Analyzing spectra file 1/1[0%]\n",
      "\n",
      "B results after filtering:\n",
      "[('LSISADIET', 19), ('MRLFVGEP', 18), ('EVDICTVGL', 16), ('KKLDLNCD', 16), ('KKDRESGE', 16), ('KAAGWSELS', 14), ('LSEGLTTLD', 14), ('KTFSHELS', 14), ('MLLNELLC', 13), ('LLENMTEV', 13), ('ITEKETVE', 13), ('EDQPIYLA', 13), ('FVRDLQGN', 13), ('LVQSNALTCHVCEA', 13)]\n",
      "Y results after filtering:\n",
      "[('PVNSPMTKG', 14), ('SNQPEGVSI', 14), ('DLTPKDIE', 13), ('SSMAYPNLVAMASQ', 13), ('DTLPGSEVL', 13), ('ECGHLRAQLEEQG', 13), ('CATTFIKF', 13), ('ESLRPSNQ', 13), ('TYSIGRSF', 12), ('FHLPSRSS', 12), ('TNEAPQATV', 12), ('MVLHCQLS', 12)]\n",
      "\n",
      "Filtering time took 0.5902979373931885 resulting in 14 b sequences and 12 y sequences\n",
      "First alignment round took 0.018114089965820312 time resulting in 194 alignments\n",
      "Getting rid of ambiguous time took 0.09794998168945312\n",
      "Filling in precursor took 0.10555529594421387 for 194 sequences\n",
      "Time to make into objects took 0.008098125457763672\n",
      "B results after filtering:]\n",
      "[('SATG', 12)]\n",
      "Y results after filtering:\n",
      "[('TITTSPN', 14), ('PGGS', 10), ('SPGG', 10), ('RLVRLLNTTDV', 10), ('VVKRDIDS', 8), ('VVDSLQLTG', 8)]\n",
      "\n",
      "Filtering time took 0.2381300926208496 resulting in 1 b sequences and 6 y sequences\n",
      "First alignment round took 0.00189208984375 time resulting in 13 alignments\n",
      "Getting rid of ambiguous time took 0.00893712043762207\n",
      "Filling in precursor took 0.014477968215942383 for 13 sequences\n",
      "Time to make into objects took 0.020541906356811523\n",
      "B results after filtering:]\n",
      "[('TTDDPSMLARVVK', 13), ('TTALLKIDITDTE', 10), ('SSPNGILLFREA', 9), ('TTWLQWASLLFVDN', 9), ('TTTTTFKG', 8), ('EDED', 8), ('DEED', 8), ('KANKTVDDTKLAE', 8)]\n",
      "Y results after filtering:\n",
      "[('APVPSPVAPAPVPS', 11), ('NFEANTTIGRIRFH', 11), ('GAGRDTLLGAKATPS', 9), ('EAERKLQEQQR', 8), ('LQVQAKDRGSPS', 8), ('AKATESPAAGVRK', 8), ('EATDNQAIKALLQ', 8), ('SLQVQAKDRGSP', 8)]\n",
      "\n",
      "Filtering time took 0.6241111755371094 resulting in 8 b sequences and 8 y sequences\n",
      "First alignment round took 0.008461236953735352 time resulting in 80 alignments\n",
      "Getting rid of ambiguous time took 0.06611824035644531\n",
      "Filling in precursor took 0.09209179878234863 for 80 sequences\n",
      "Time to make into objects took 0.3014841079711914\n",
      "B results after filtering:]\n",
      "[('TTPATSTTCTAT', 10), ('TTSPNLGTREN', 10), ('FSLHDP', 9), ('KRCPTCHNIIDN', 9), ('LHAYDQHLNMILGD', 8), ('HISFDP', 8), ('LQGWNKTMDGKADVSFV', 8), ('QVWADP', 8)]\n",
      "Y results after filtering:\n",
      "[('YEEDIPSVTRQVPS', 11)]\n",
      "\n",
      "Filtering time took 0.5639500617980957 resulting in 8 b sequences and 1 y sequences\n",
      "First alignment round took 0.0032389163970947266 time resulting in 17 alignments\n",
      "Getting rid of ambiguous time took 0.004602909088134766\n",
      "Filling in precursor took 0.002183198928833008 for 17 sequences\n",
      "Time to make into objects took 0.0041849613189697266\n",
      "B results after filtering:]\n",
      "[('EVSNNHVLIYLDKVSNQ', 11), ('ESLQLAGKWDLGNLLL', 11), ('VEDILNVLGMENVV', 11), ('GVRREASDGEVRRGYHK', 10), ('YTAAALSGGLMVFTAA', 10), ('LYVDDPNWPGIGVV', 10), ('EIDVNGILRVTAED', 10), ('EWLLTRDGVPSEAA', 10), ('QLDCALDLMRRLP', 9), ('LQSKNGSFTPRTAHIL', 9), ('YKTAFKSLQEQNS', 9), ('LSSLEKSSPTPRATAPQ', 9), ('QGPVAGTGDFGPLKEA', 9), ('RPIEDNEKSKDVN', 9), ('DENFLLFFRLET', 9), ('NTDQLALSQAQLEI', 9), ('TTFQISLNISYIGS', 9), ('APKGMLARGTYHNK', 9), ('GEDAGGGVGKELVDLKII', 9), ('RAKDREISALQEQLL', 9), ('LSQELDFVSHNVR', 8), ('DIVNEEDKDVLI', 8), ('SLSLHCTIDEKR', 8), ('VTTVTTVQSRPRVDGKLD', 8), ('TVVTEAGNLLKDNAT', 8), ('TVISNWIKY', 8), ('YHRYPELREALV', 8), ('VEVERDNLAEDI', 8), ('SSCGRTGPLRRGSLP', 8), ('LFLTGESYAGIYIP', 8)]\n",
      "Y results after filtering:\n",
      "[('GGQIASIGQQANIPTA', 12), ('LLNLGLITESQAHTLE', 12), ('IIAASAQECIVEPVVPI', 12), ('AAAAAAADISARKMAH', 11), ('GFLLLLFQGTRCDIQMT', 11), ('ILVFPNDDLHVKDLN', 10), ('PRVEYPAHQAMNL', 10), ('NQNSVPDIRGVETP', 10), ('VRLTYSTGESNTVV', 10), ('KAASNLRDNYRFA', 10), ('VRLFTEEELARY', 10), ('IIAFGDRIEEFKSIN', 10), ('ELREHLSQSKVAE', 9), ('LLLLPPPNAYAADTPGE', 9), ('PLKDFWPKVPDSP', 9), ('ASKPSVNEQNQVQP', 9), ('LLTKAEPEAIPERRE', 9), ('LITASQFFEIWLHF', 9), ('ILLRHCDVNKDGKIQ', 9), ('GIAVPDVYSACKRF', 8), ('CPSAYKLLLGTEKC', 8), ('QTAVKTNSSFIQGF', 8), ('VPYARSEAHLTELLEEV', 8), ('LICTLNAANVPAGTEVVCAP', 8)]\n",
      "\n",
      "Filtering time took 1.306732177734375 resulting in 30 b sequences and 24 y sequences\n",
      "First alignment round took 0.08436727523803711 time resulting in 774 alignments\n",
      "Getting rid of ambiguous time took 4.658141851425171\n",
      "Filling in precursor took 0.032549142837524414 for 774 sequences\n",
      "Time to make into objects took 0.018150806427001953\n",
      "B results after filtering:]\n",
      "[('VEEVRKNKESKDPADET', 8), ('VEDLTKETLHKQY', 7), ('LSSLRAPLDIPIPDPPPKDD', 7), ('LSTVEE', 7), ('SLSIEE', 7), ('ELVDSVLDVVRKEC', 7), ('DVLEGTLQCPESGRLFPI', 7), ('VTTVTTVQSRPRVDGKLDSP', 7), ('TVVTEAGNLLKDNATQEEIL', 7)]\n",
      "Y results after filtering:\n",
      "[('LADPRTWRNVPEKKECVV', 7), ('LNPPGGKTSDIFGSPVTATAPL', 6), ('GYSVVDVTSFEGSKAPIPTAL', 6), ('LEVAIQCLETAFGV', 6), ('QEAEAARVELQKKAQALQE', 6)]\n",
      "\n",
      "Filtering time took 1.238652229309082 resulting in 9 b sequences and 5 y sequences\n",
      "First alignment round took 0.004745006561279297 time resulting in 59 alignments\n",
      "Getting rid of ambiguous time took 0.12044382095336914\n",
      "Filling in precursor took 0.006750822067260742 for 59 sequences\n",
      "Time to make into objects took 0.002131938934326172\n",
      "B results after filtering:]\n",
      "[('EEVKHMELFQK', 8), ('TTILSSPDGELVDA', 8), ('TFPFLHTPSLSNG', 8), ('EEIAQVATISANGD', 7), ('EELCRQR', 7), ('EEGIVLGGGC', 7), ('EEEDKKR', 7), ('EEFMKTWRKY', 7)]\n",
      "Y results after filtering:\n",
      "[('VVPS', 7), ('AQIGSGAVALCPESP', 7), ('EQSAPPS', 6), ('WGLPGPS', 6), ('AEWHSHL', 6)]\n",
      "\n",
      "Filtering time took 0.5827760696411133 resulting in 8 b sequences and 5 y sequences\n",
      "First alignment round took 0.008127927780151367 time resulting in 53 alignments\n",
      "Getting rid of ambiguous time took 0.027981042861938477\n",
      "Filling in precursor took 0.009266138076782227 for 53 sequences\n",
      "Time to make into objects took 0.0008189678192138672\n",
      "B results after filtering:]\n",
      "[('SSLL', 14), ('SSIL', 14)]\n",
      "Y results after filtering:\n",
      "[('PHKRPAAATATARPSTLPA', 12), ('SLQALPT', 11), ('EKIADPTLA', 11)]\n",
      "\n",
      "Filtering time took 0.0973212718963623 resulting in 2 b sequences and 3 y sequences\n",
      "First alignment round took 0.0007469654083251953 time resulting in 11 alignments\n",
      "Getting rid of ambiguous time took 0.005053997039794922\n",
      "Filling in precursor took 0.007562160491943359 for 11 sequences\n",
      "Time to make into objects took 0.02227306365966797\n",
      "B results after filtering:]\n",
      "[('TGLTILQSGIRP', 11), ('KLLDFIKHNQ', 11), ('TVDLKRKPADL', 11), ('ISIRSKSAPLPS', 11)]\n",
      "Y results after filtering:\n",
      "[('EEILKAFKLF', 11), ('TKDRSLRKSF', 10)]\n",
      "\n",
      "Filtering time took 0.45171213150024414 resulting in 4 b sequences and 2 y sequences\n",
      "First alignment round took 0.0008668899536132812 time resulting in 14 alignments\n",
      "Getting rid of ambiguous time took 0.019744873046875\n",
      "Filling in precursor took 0.0002942085266113281 for 14 sequences\n",
      "Time to make into objects took 0.0018470287322998047\n",
      "\n",
      "Finished search. Writting results to /Users/zacharymcgrath/Desktop/Experiment output/singleReal/...\n",
      "Could not make an alignment for 0/9 spectra (0%)\n"
     ]
    }
   ],
   "source": [
    "import time\n",
    "st = time.time()\n",
    "runner.run(params)\n",
    "ttime = time.time() - st"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "TIME: 81.1438581943512s\n"
     ]
    }
   ],
   "source": [
    "print(f'TIME: {ttime}s')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Load and read results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "\n",
    "outputFile = '/Users/zacharymcgrath/Desktop/Experiment output/singleReal/summary.tsv'\n",
    "hybridOutputFile = '/Users/zacharymcgrath/Desktop/Experiment output/singleReal/hybrid_summary.tsv'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "df = pd.read_csv(outputFile, sep='\\t')\n",
    "hdf = pd.read_csv(hybridOutputFile, sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th>left_proteins</th>\n",
       "      <th>right_proteins</th>\n",
       "      <th>sequence</th>\n",
       "      <th>hybrid_sequence</th>\n",
       "      <th>b_score</th>\n",
       "      <th>y_score</th>\n",
       "      <th>total_score</th>\n",
       "      <th>precursor_distance</th>\n",
       "      <th>entry name</th>\n",
       "      <th>id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>['CELR3_MOUSE Cadherin EGF LAG seven-pass G-ty...</td>\n",
       "      <td>['TNFL9_MOUSE Tumor necrosis factor ligand sup...</td>\n",
       "      <td>SATGTITTS</td>\n",
       "      <td>SATG-TITTS</td>\n",
       "      <td>4</td>\n",
       "      <td>0</td>\n",
       "      <td>50</td>\n",
       "      <td>0.983231</td>\n",
       "      <td>/Users/zacharymcgrath/Desktop/nod2 data/single...</td>\n",
       "      <td>scan=2470</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>['EF1B_MOUSE Elongation factor 1-beta', 'EF1D_...</td>\n",
       "      <td>['MAP4_MOUSE Microtubule-associated protein 4']</td>\n",
       "      <td>SSILAPHKRPAAATATARPSTLPARD</td>\n",
       "      <td>SSIL-APHKRPAAATATARPSTLPARD</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>12</td>\n",
       "      <td>2.502045</td>\n",
       "      <td>/Users/zacharymcgrath/Desktop/nod2 data/single...</td>\n",
       "      <td>NOD2_E3.7258.7258.5.pkl</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   Unnamed: 0                                      left_proteins  \\\n",
       "0           0  ['CELR3_MOUSE Cadherin EGF LAG seven-pass G-ty...   \n",
       "1           1  ['EF1B_MOUSE Elongation factor 1-beta', 'EF1D_...   \n",
       "\n",
       "                                      right_proteins  \\\n",
       "0  ['TNFL9_MOUSE Tumor necrosis factor ligand sup...   \n",
       "1    ['MAP4_MOUSE Microtubule-associated protein 4']   \n",
       "\n",
       "                     sequence              hybrid_sequence  b_score  y_score  \\\n",
       "0                   SATGTITTS                   SATG-TITTS        4        0   \n",
       "1  SSILAPHKRPAAATATARPSTLPARD  SSIL-APHKRPAAATATARPSTLPARD        3        0   \n",
       "\n",
       "   total_score  precursor_distance  \\\n",
       "0           50            0.983231   \n",
       "1           12            2.502045   \n",
       "\n",
       "                                          entry name                       id  \n",
       "0  /Users/zacharymcgrath/Desktop/nod2 data/single...                scan=2470  \n",
       "1  /Users/zacharymcgrath/Desktop/nod2 data/single...  NOD2_E3.7258.7258.5.pkl  "
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "hdf.head(5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th>proteins</th>\n",
       "      <th>sequence</th>\n",
       "      <th>b_score</th>\n",
       "      <th>y_score</th>\n",
       "      <th>total_score</th>\n",
       "      <th>precursor_distance</th>\n",
       "      <th>entry name</th>\n",
       "      <th>id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>['CBPE_MOUSE Carboxypeptidase E']</td>\n",
       "      <td>FVRDLQGNPI</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>22</td>\n",
       "      <td>0.021672</td>\n",
       "      <td>/Users/zacharymcgrath/Desktop/nod2 data/single...</td>\n",
       "      <td>NOD2_E3.7065.7065.2.pkl</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>['PRDX6_MOUSE Peroxiredoxin-6']</td>\n",
       "      <td>DEAPNFEANTTIGRIRFH</td>\n",
       "      <td>2</td>\n",
       "      <td>5</td>\n",
       "      <td>41</td>\n",
       "      <td>1.010352</td>\n",
       "      <td>/Users/zacharymcgrath/Desktop/nod2 data/single...</td>\n",
       "      <td>NOD2_E3.9670.9702.4.pkl</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>[\"5NT3A_MOUSE Cytosolic 5'-nucleotidase 3A\"]</td>\n",
       "      <td>KRCPTCHNI</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>37</td>\n",
       "      <td>0.490427</td>\n",
       "      <td>/Users/zacharymcgrath/Desktop/nod2 data/single...</td>\n",
       "      <td>NOD2_E3.13446.13477.2.pkl</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>['SEGN_MOUSE Secretagogin']</td>\n",
       "      <td>ENFLLFFRLE</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>11</td>\n",
       "      <td>2.994200</td>\n",
       "      <td>/Users/zacharymcgrath/Desktop/nod2 data/single...</td>\n",
       "      <td>scan=10130</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td>['TR112_MOUSE Multifunctional methyltransferas...</td>\n",
       "      <td>EGTLQCPESGRLFP</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>7</td>\n",
       "      <td>2.983264</td>\n",
       "      <td>/Users/zacharymcgrath/Desktop/nod2 data/single...</td>\n",
       "      <td>scan=11410</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>5</td>\n",
       "      <td>5</td>\n",
       "      <td>['NUCB1_MOUSE Nucleobindin-1']</td>\n",
       "      <td>EQSAPPSKP</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>12</td>\n",
       "      <td>2.999958</td>\n",
       "      <td>/Users/zacharymcgrath/Desktop/nod2 data/single...</td>\n",
       "      <td>NOD2_E3.11427.11516.2.pkl</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   Unnamed: 0                                           proteins  \\\n",
       "0           0                  ['CBPE_MOUSE Carboxypeptidase E']   \n",
       "1           1                    ['PRDX6_MOUSE Peroxiredoxin-6']   \n",
       "2           2       [\"5NT3A_MOUSE Cytosolic 5'-nucleotidase 3A\"]   \n",
       "3           3                        ['SEGN_MOUSE Secretagogin']   \n",
       "4           4  ['TR112_MOUSE Multifunctional methyltransferas...   \n",
       "5           5                     ['NUCB1_MOUSE Nucleobindin-1']   \n",
       "\n",
       "             sequence  b_score  y_score  total_score  precursor_distance  \\\n",
       "0          FVRDLQGNPI        2        0           22            0.021672   \n",
       "1  DEAPNFEANTTIGRIRFH        2        5           41            1.010352   \n",
       "2           KRCPTCHNI        2        1           37            0.490427   \n",
       "3          ENFLLFFRLE        1        0           11            2.994200   \n",
       "4      EGTLQCPESGRLFP        1        0            7            2.983264   \n",
       "5           EQSAPPSKP        1        0           12            2.999958   \n",
       "\n",
       "                                          entry name  \\\n",
       "0  /Users/zacharymcgrath/Desktop/nod2 data/single...   \n",
       "1  /Users/zacharymcgrath/Desktop/nod2 data/single...   \n",
       "2  /Users/zacharymcgrath/Desktop/nod2 data/single...   \n",
       "3  /Users/zacharymcgrath/Desktop/nod2 data/single...   \n",
       "4  /Users/zacharymcgrath/Desktop/nod2 data/single...   \n",
       "5  /Users/zacharymcgrath/Desktop/nod2 data/single...   \n",
       "\n",
       "                          id  \n",
       "0    NOD2_E3.7065.7065.2.pkl  \n",
       "1    NOD2_E3.9670.9702.4.pkl  \n",
       "2  NOD2_E3.13446.13477.2.pkl  \n",
       "3                 scan=10130  \n",
       "4                 scan=11410  \n",
       "5  NOD2_E3.11427.11516.2.pkl  "
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.head(6)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "import json\n",
    "\n",
    "s = json.load(open('/Users/zacharymcgrath/Desktop/Experiment output/singleReal/summary.json', 'r'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "FVRDLQGNPI: 22, 2, 0\n",
      "AYPNLVAMASQ: 20, 1, 1\n",
      "DCATTFIKFL: 11, 0, 1\n",
      "SATGTITTS: 50, 4, 0\n",
      "SATGAEPGGS: 44, 3, 1\n",
      "SATGVKRD: 42, 3, 0\n",
      "DEAPNFEANTTIGRIRFH: 41, 2, 5\n",
      "DEEDGKEAERKLQEQQR: 31, 2, 3\n",
      "SSPNGILLFREAERKLQE: 29, 5, 0\n",
      "KRCPTCHNI: 37, 2, 1\n",
      "TTSPNLGTRE: 22, 2, 0\n",
      "LMIQVWADP: 12, 0, 1\n",
      "ENFLLFFRLE: 11, 1, 0\n",
      "YHRYPELREA: 11, 1, 0\n",
      "QLDCALDLMRR: 10, 1, 1\n",
      "EGTLQCPESGRLFP: 7, 1, 0\n",
      "EEVRKNKESKDPA: 0, 0, 0\n",
      "EQSAPPSKP: 12, 1, 0\n",
      "SSILAPHKRPAAATATARPSTLPARD: 12, 3, 0\n",
      "DILSSILPHKRPAAATATARPSTLPA: 12, 1, 2\n",
      "SSLLAPHKRPAAATATARPSTLPARD: 12, 3, 0\n",
      "EEILKAFKLF: 44, 2, 2\n",
      "TKDRSLRKSF: 11, 0, 1\n"
     ]
    }
   ],
   "source": [
    "for k, v in s.items():\n",
    "    for a in v['alignments']:\n",
    "        print(f'{a[\"sequence\"]}: {a[\"total_score\"]}, {a[\"b_score\"]}, {a[\"y_score\"]}')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
